<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Orchestrating Large-Scale Single-Cell Analysis with Bioconductor: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="intro-sce.html">1. Introduction to Bioconductor and the SingleCellExperiment class</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="eda_qc.html">2. Exploratory data analysis and quality control</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="cell_type_annotation.html">3. Cell type annotation</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="multi-sample.html">4. Multi-sample analyses</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="large_data.html">5. Working with large data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="hca.html">6. Accessing data from the Human Cell Atlas (HCA)</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-intro-sce"><p>Content from <a href="intro-sce.html">Introduction to Bioconductor and the SingleCellExperiment class</a></p>
<hr>
<p>Last updated on 2025-02-07 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/intro-sce.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is Bioconductor?</li>
<li>How is single-cell data stored in the Bioconductor ecosystem?</li>
<li>What is a <code>SingleCellExperiment</code> object?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Install and update Bioconductor packages.</li>
<li>Load data generated with common single-cell technologies as
<code>SingleCellExperiment</code> objects.</li>
<li>Inspect and manipulate <code>SingleCellExperiment</code>
objects.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="bioconductor">Bioconductor<a class="anchor" aria-label="anchor" href="#bioconductor"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>Within the <a href="https://www.r-project.org/" class="external-link">R</a> ecosystem, the
<a href="https://bioconductor.org/" class="external-link">Bioconductor</a> project provides
tools for the analysis and comprehension of high-throughput genomics
data. The scope of the project covers microarray data, various forms of
sequencing (RNA-seq, ChIP-seq, bisulfite, genotyping, etc.), proteomics,
flow cytometry and more. One of Bioconductor’s main selling points is
the use of common data structures to promote interoperability between
packages, allowing code written by different people (from different
organizations, in different countries) to work together seamlessly in
complex analyses.</p>
</div>
<div class="section level3">
<h3 id="installing-bioconductor-packages">Installing Bioconductor Packages<a class="anchor" aria-label="anchor" href="#installing-bioconductor-packages"></a>
</h3>
<p>The default repository for R packages is the <a href="https://cran.r-project.org/mirrors.html" class="external-link">Comprehensive R Archive
Network</a> (CRAN), which is home to over 13,000 different R packages.
We can easily install packages from CRAN - say, the popular <em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em>
package for data visualization - by opening up R and typing in:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span></code></pre>
</div>
<p>In our case, however, we want to install Bioconductor packages. These
packages are located in a separate repository hosted by Bioconductor, so
we first install the <em><a href="https://CRAN.R-project.org/package=BiocManager" class="external-link">BiocManager</a></em>
package to easily connect to the Bioconductor servers.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p>After that, we can use <em><a href="https://CRAN.R-project.org/package=BiocManager" class="external-link">BiocManager</a></em>’s
<code>install()</code> function to install any package from
Bioconductor. For example, the code chunk below uses this approach to
install the <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
package.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span></code></pre>
</div>
<p>Should we forget, the same instructions are present on the landing
page of any Bioconductor package. For example, looking at the <a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link"><code>scater</code></a>
package page on Bioconductor, we can see the following copy-pasteable
instructions:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span></span></code></pre>
</div>
<p>Packages only need to be installed once, and then they are available
for all subsequent uses of a particular R installation. There is no need
to repeat the installation every time we start R.</p>
</div>
<div class="section level3">
<h3 id="finding-relevant-packages">Finding relevant packages<a class="anchor" aria-label="anchor" href="#finding-relevant-packages"></a>
</h3>
<p>To find relevant Bioconductor packages, one useful resource is the <a href="https://bioconductor.org/packages/release/BiocViews.html" class="external-link">BiocViews</a>
page. This provides a hierarchically organized view of annotations
associated with each Bioconductor package. For example, under the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">“Software”</a>
label, we might be interested in a particular <a href="https://bioconductor.org/packages/release/BiocViews.html#___Technology" class="external-link">“Technology”</a>
such as… say, <a href="https://bioconductor.org/packages/release/BiocViews.html#___SingleCell" class="external-link">“SingleCell”</a>.
This gives us a listing of all Bioconductor packages that might be
useful for our single-cell data analyses. CRAN uses the similar concept
of <a href="https://cran.r-project.org/web/views/" class="external-link">“Task views”</a>,
though this is understandably more general than genomics. For example,
the <a href="https://cran.r-project.org/web/views/Cluster.html" class="external-link">Cluster
task view page</a> lists an assortment of packages that are relevant to
cluster analyses.</p>
</div>
<div class="section level3">
<h3 id="staying-up-to-date">Staying up to date<a class="anchor" aria-label="anchor" href="#staying-up-to-date"></a>
</h3>
<p>Updating all R/Bioconductor packages is as simple as running
<code>BiocManager::install()</code> without any arguments. This will
check for more recent versions of each package (within a Bioconductor
release) and prompt the user to update if any are available.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>This might take some time if many packages need to be updated, but is
typically recommended to avoid issues resulting from outdated package
versions.</p>
</div>
</section><section><h2 class="section-heading" id="the-singlecellexperiment-class">The <code>SingleCellExperiment</code> class<a class="anchor" aria-label="anchor" href="#the-singlecellexperiment-class"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>We start by loading some libraries we’ll be using:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span></code></pre>
</div>
<p>It is normal to see lot of startup messages when loading these
packages.</p>
</div>
<div class="section level3">
<h3 id="motivation-and-overview">Motivation and overview<a class="anchor" aria-label="anchor" href="#motivation-and-overview"></a>
</h3>
<p>One of the main strengths of the Bioconductor project lies in the use
of a common data infrastructure that powers interoperability across
packages.</p>
<p>Users should be able to analyze their data using functions from
different Bioconductor packages without the need to convert between
formats. To this end, the <code>SingleCellExperiment</code> class (from
the <em>SingleCellExperiment</em> package) serves as the common currency
for data exchange across 70+ single-cell-related Bioconductor
packages.</p>
<p>This class implements a data structure that stores all aspects of our
single-cell data - gene-by-cell expression data, cell-wise metadata, and
gene-wise annotation - and lets us manipulate them in an organized
manner.</p>
<figure><img src="http://bioconductor.org/books/release/OSCA.intro/images/SingleCellExperiment.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> Benefits of using the integrated SingleCellExperiment data container </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" aria-labelledby="headingSpoiler1" data-bs-parent="#accordionSpoiler1">
<div class="accordion-body">
<p>The complexity of the <code>SingleCellExperiment</code> container
might be a little bit intimidating in the beginning. One might be
tempted to use a simpler approach by just keeping all of these
components in separate objects, e.g. a <code>matrix</code> of counts, a
<code>data.frame</code> of sample metadata, a <code>data.frame</code> of
gene annotations, and so on.</p>
<p>There are two main disadvantages to this type of “from scratch”
approach:</p>
<ol style="list-style-type: decimal">
<li>It requires a substantial amount of manual bookkeeping to keep the
different data components in sync. If you performed a QC step that
removed dead cells from the count matrix, you also had to remember to
remove that same set of cells from the cell-wise metadata. Did you
filter out genes that did not display sufficient expression levels to be
retained for further analysis? Then you also need to remember to filter
the gene metadata table too.</li>
<li>All the downstream steps have to be “from scratch” as well. All the
data munging, analysis, and visualization code will need to be
customized to the idiosyncrasies of a given input set.</li>
</ol>
</div>
</div>
</div>
</div>
<p>Let’s look at an example dataset. <code>WTChimeraData</code> comes
from a study on mouse development <a href="https://www.nature.com/articles/s41586-019-0933-9" class="external-link">Pijuan-Sala et
al.</a>. The study profiles the effect of a transcription factor TAL1
and its influence on mouse development. Because mutations in this gene
can cause severe developmental issues, Tal1-/- cells (positive for
tdTomato, a fluorescent protein) were injected into wild-type
blastocysts (tdTomato-), forming chimeric embryos.</p>
<p>We can assign one sample to a <code>SingleCellExperiment</code>
object named <code>sce</code> like so:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 2411
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(2411): cell_9769 cell_9770 ... cell_12178 cell_12179
colData names(11): cell barcode ... doub.density sizeFactor
reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>We can think of this (and other) class as a <em>container</em>, that
contains several different pieces of data in so-called <em>slots</em>.
SingleCellExperiment objects come with dedicated methods for
<em>getting</em> and <em>setting</em> the data in their slots.</p>
<p>Depending on the object, slots can contain different types of data
(e.g., numeric matrices, lists, etc.). Here we’ll review the main slots
of the SingleCellExperiment class as well as their getter/setter
methods.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Get the data for a different sample from <code>WTChimeraData</code>
(other than the fifth one).</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Here we obtain the sixth sample and assign it to
<code>sce6</code>:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce6</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce6</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="assays">
<code>assays</code><a class="anchor" aria-label="anchor" href="#assays"></a>
</h3>
<p>This is arguably the most fundamental part of the object that
contains the count matrix, and potentially other matrices with
transformed data. We can access the <em>list</em> of matrices with the
<code>assays</code> function and individual matrices with the
<code>assay</code> function. If one of these matrices is called
“counts”, we can use the special <code>counts</code> getter (likewise
for <code>logcounts</code>).</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>3 x 3 sparse Matrix of class "dgCMatrix"
                   cell_9769 cell_9770 cell_9771
ENSMUSG00000051951         .         .         .
ENSMUSG00000089699         .         .         .
ENSMUSG00000102343         .         .         .</code></pre>
</div>
<p>You will notice that in this case we have a sparse matrix of class
<code>dgTMatrix</code> inside the object. More generally, any
“matrix-like” object can be used, e.g., dense matrices or HDF5-backed
matrices (as we will explore later in the <a href="https://carpentries-incubator.github.io/bioc-scrnaseq/large_data.html" class="external-link">Working
with large data</a> episode).</p>
</div>
<div class="section level3">
<h3 id="coldata-and-rowdata">
<code>colData</code> and <code>rowData</code>
<a class="anchor" aria-label="anchor" href="#coldata-and-rowdata"></a>
</h3>
<p>Conceptually, these are two data frames that annotate the columns and
the rows of your assay, respectively.</p>
<p>One can interact with them as usual, e.g., by extracting columns or
adding additional variables as columns.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 4 columns
                 cell          barcode    sample       stage
          &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt;
cell_9769   cell_9769 AAACCTGAGACTGTAA         5        E8.5
cell_9770   cell_9770 AAACCTGAGATGCCTT         5        E8.5
cell_9771   cell_9771 AAACCTGAGCAGCCTC         5        E8.5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 2 columns
                              ENSEMBL      SYMBOL
                          &lt;character&gt; &lt;character&gt;
ENSMUSG00000051951 ENSMUSG00000051951        Xkr4
ENSMUSG00000089699 ENSMUSG00000089699      Gm1992
ENSMUSG00000102343 ENSMUSG00000102343     Gm37381</code></pre>
</div>
<p>You can access columns of the colData with the <code>$</code>
accessor to quickly add cell-wise metadata to the
<code>colData</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">my_sum</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 12 columns
                 cell          barcode    sample       stage    tomato
          &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;
cell_9769   cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE
cell_9770   cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE
cell_9771   cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE
               pool stage.mapped celltype.mapped closest.cell doub.density
          &lt;integer&gt;  &lt;character&gt;     &lt;character&gt;  &lt;character&gt;    &lt;numeric&gt;
cell_9769         3        E8.25      Mesenchyme   cell_24159   0.02985045
cell_9770         3         E8.5     Endothelium   cell_96660   0.00172753
cell_9771         3         E8.5       Allantois  cell_134982   0.01338013
          sizeFactor    my_sum
           &lt;numeric&gt; &lt;numeric&gt;
cell_9769    1.41243     27577
cell_9770    1.22757     29309
cell_9771    1.15439     28795</code></pre>
</div>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Add a column of gene-wise metadata to the <code>rowData</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Here, we add a column of named <code>conservation</code> that could
represent an evolutionary conservation score.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">conservation</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fu">nrow</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>These are just random numbers for demonstration purposes, but in
practice storing gene-wise data in the <code>rowData</code> is
convenient and simplifies data management.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="the-reduceddims">The <code>reducedDims</code>
<a class="anchor" aria-label="anchor" href="#the-reduceddims"></a>
</h3>
<p>Everything that we have described so far (except for the
<code>counts</code> getter) is part of the
<code>SummarizedExperiment</code> class that
<code>SingleCellExperiment</code> extends. You can find a complete
lesson on the <code>SummarizedExperiment</code> class in <a href="https://carpentries-incubator.github.io/bioc-intro/60-next-steps.html" class="external-link">Introduction
to data analysis with R and Bioconductor</a> course.</p>
<p>One peculiarity of <code>SingleCellExperiment</code> is its ability
to store reduced dimension matrices within the object. These may include
PCA, t-SNE, UMAP, etc.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reducedDims</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of length 2
names(2): pca.corrected.E7.5 pca.corrected.E8.5</code></pre>
</div>
<p>As for the other slots, we have the usual setter/getter, but it is
somewhat rare to interact directly with these functions.</p>
<p>It is more common for other functions to <em>store</em> this
information in the object, e.g., the <code>runPCA</code> function from
the <code>scater</code> package.</p>
<p>Here, we use <code>scater</code>’s <code>plotReducedDim</code>
function as an example of how to extract this information
<em>indirectly</em> from the objects. Note that one could obtain the
same results by manually extracting the corresponding
<code>reducedDim</code> matrix and cell type labels then passing them to
<code>ggplot</code> in a data frame.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"pca.corrected.E8.5"</span>, colour_by <span class="op">=</span> <span class="st">"stage.mapped"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/intro-sce-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Create a <code>SingleCellExperiment</code> object “from scratch”.
That means: start from a <code>matrix</code> (either randomly generated
or with some fake data in it) and add one or more columns as
<code>colData</code>.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>The <code>SingleCellExperiment</code> constructor function can be
used to create a new <code>SingleCellExperiment</code> object.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">runif</span><span class="op">(</span><span class="fl">30</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="op">(</span>assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>logcounts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_sce</span><span class="op">$</span><span class="va">my_col_info</span> <span class="op">=</span> <span class="fu">runif</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 6 5
metadata(0):
assays(1): logcounts
rownames: NULL
rowData names(0):
colnames: NULL
colData names(1): my_col_info
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Combine two <code>SingleCellExperiment</code> objects. The
<code>MouseGastrulationData</code> package contains several datasets.
Download sample 6 of the chimera experiment. Use the <code>cbind</code>
function to combine the new data with the <code>sce</code> object
created before.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span>  <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce6</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="va">combined_sce</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">sce</span>, <span class="va">sce6</span><span class="op">)</span></span>
<span></span>
<span><span class="va">combined_sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 3458
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(3458): cell_9769 cell_9770 ... cell_13225 cell_13226
colData names(11): cell barcode ... doub.density sizeFactor
reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="further-reading" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further Reading</h3>
<div class="callout-content">
<ul>
<li>OSCA book, <a href="https://bioconductor.org/books/release/OSCA.intro" class="external-link">Introduction</a>
</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The Bioconductor project provides open-source software packages for
the comprehension of high-throughput biological data.</li>
<li>A <code>SingleCellExperiment</code> object is an extension of the
<code>SummarizedExperiment</code> object.</li>
<li>
<code>SingleCellExperiment</code> objects contain specialized data
fields for storing data unique to single-cell analyses, such as the
<code>reducedDims</code> field.</li>
</ul>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width">
<ol style="list-style-type: decimal">
<li>Pijuan-Sala B, Griffiths JA, Guibentif C et al. (2019). A
single-cell molecular map of mouse gastrulation and early organogenesis.
Nature 566, 7745:490-495.</li>
</ol></section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] scater_1.32.1                ggplot2_3.5.1
 [3] scuttle_1.14.0               MouseGastrulationData_1.18.0
 [5] SpatialExperiment_1.14.0     SingleCellExperiment_1.26.0
 [7] SummarizedExperiment_1.34.0  Biobase_2.64.0
 [9] GenomicRanges_1.56.2         GenomeInfoDb_1.40.1
[11] IRanges_2.38.1               S4Vectors_0.42.1
[13] BiocGenerics_0.50.0          MatrixGenerics_1.16.0
[15] matrixStats_1.5.0            BiocStyle_2.32.1

loaded via a namespace (and not attached):
 [1] DBI_1.2.3                 formatR_1.14
 [3] gridExtra_2.3             rlang_1.1.5
 [5] magrittr_2.0.3            compiler_4.4.2
 [7] RSQLite_2.3.9             DelayedMatrixStats_1.26.0
 [9] png_0.1-8                 vctrs_0.6.5
[11] pkgconfig_2.0.3           crayon_1.5.3
[13] fastmap_1.2.0             dbplyr_2.5.0
[15] magick_2.8.5              XVector_0.44.0
[17] labeling_0.4.3            rmarkdown_2.29
[19] UCSC.utils_1.0.0          ggbeeswarm_0.7.2
[21] purrr_1.0.2               bit_4.5.0.1
[23] xfun_0.50                 zlibbioc_1.50.0
[25] cachem_1.1.0              beachmat_2.20.0
[27] jsonlite_1.8.9            blob_1.2.4
[29] DelayedArray_0.30.1       BiocParallel_1.38.0
[31] irlba_2.3.5.1             parallel_4.4.2
[33] R6_2.5.1                  Rcpp_1.0.14
[35] knitr_1.49                Matrix_1.7-2
[37] tidyselect_1.2.1          abind_1.4-8
[39] yaml_2.3.10               viridis_0.6.5
[41] codetools_0.2-20          curl_6.2.0
[43] lattice_0.22-6            tibble_3.2.1
[45] withr_3.0.2               KEGGREST_1.44.1
[47] BumpyMatrix_1.12.0        evaluate_1.0.3
[49] BiocFileCache_2.12.0      ExperimentHub_2.12.0
[51] Biostrings_2.72.1         pillar_1.10.1
[53] BiocManager_1.30.25       filelock_1.0.3
[55] renv_1.1.0                generics_0.1.3
[57] BiocVersion_3.19.1        sparseMatrixStats_1.16.0
[59] munsell_0.5.1             scales_1.3.0
[61] glue_1.8.0                tools_4.4.2
[63] AnnotationHub_3.12.0      BiocNeighbors_1.22.0
[65] ScaledMatrix_1.12.0       cowplot_1.1.3
[67] grid_4.4.2                AnnotationDbi_1.66.0
[69] colorspace_2.1-1          GenomeInfoDbData_1.2.12
[71] beeswarm_0.4.0            BiocSingular_1.20.0
[73] vipor_0.4.7               cli_3.6.3
[75] rsvd_1.0.5                rappdirs_0.3.3
[77] viridisLite_0.4.2         S4Arrays_1.4.1
[79] dplyr_1.1.4               gtable_0.3.6
[81] digest_0.6.37             SparseArray_1.4.8
[83] ggrepel_0.9.6             farver_2.1.2
[85] rjson_0.2.23              memoise_2.0.1
[87] htmltools_0.5.8.1         lifecycle_1.0.4
[89] httr_1.4.7                mime_0.12
[91] bit64_4.6.0-1            </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-eda_qc"><p>Content from <a href="eda_qc.html">Exploratory data analysis and quality control</a></p>
<hr>
<p>Last updated on 2025-02-07 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/eda_qc.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I examine the quality of single-cell data?</li>
<li>What data visualizations should I use during quality control in a
single-cell analysis?</li>
<li>How do I prepare single-cell data for analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Determine and communicate the quality of single-cell data.</li>
<li>Identify and filter empty droplets and doublets.</li>
<li>Perform normalization, feature selection, and dimensionality
reduction as parts of a typical single-cell analysis pipeline.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="setup-and-experimental-design">Setup and experimental design<a class="anchor" aria-label="anchor" href="#setup-and-experimental-design"></a>
</h2>
<hr class="half-width">
<p>As mentioned in the introduction, in this tutorial we will use the
wild-type data from the Tal1 chimera experiment. These data are
available through the <a href="https://bioconductor.org/packages/release/data/experiment/html/MouseGastrulationData.html" class="external-link">MouseGastrulationData</a>
Bioconductor package, which contains several datasets.</p>
<p>In particular, the package contains the following samples that we
will use for the tutorial:</p>
<ul>
<li>Sample 5: E8.5 injected cells (tomato positive), pool 3</li>
<li>Sample 6: E8.5 host cells (tomato negative), pool 3</li>
<li>Sample 7: E8.5 injected cells (tomato positive), pool 4</li>
<li>Sample 8: E8.5 host cells (tomato negative), pool 4</li>
<li>Sample 9: E8.5 injected cells (tomato positive), pool 5</li>
<li>Sample 10: E8.5 host cells (tomato negative), pool 5</li>
</ul>
<p>We start our analysis by selecting only sample 5, which contains the
injected cells in one biological replicate. We download the “raw” data
that contains all the droplets for which we have sequenced reads.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">EnsDb.Mmusculus.v79</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scDblFinder</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 522554
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(522554): AAACCTGAGAAACCAT AAACCTGAGAAACCGC ...
  TTTGTCATCTTTACGT TTTGTCATCTTTCCTC
colData names(0):
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>This is the same data we examined in the previous lesson.</p>
</section><section><h2 class="section-heading" id="droplet-processing">Droplet processing<a class="anchor" aria-label="anchor" href="#droplet-processing"></a>
</h2>
<hr class="half-width">
<p>From the experiment, we expect to have only a few thousand cells,
while we can see that we have data for more than 500,000 droplets. It is
likely that most of these droplets are empty and are capturing only
ambient or background RNA.</p>
<div id="callout1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout</h3>
<div class="callout-content">
<p>Depending on your data source, identifying and discarding empty
droplets may not be necessary. Some academic institutions have research
cores dedicated to single cell work that perform the sample preparation
and sequencing. Many of these cores will also perform empty droplet
filtering and other initial QC steps. Specific details on the steps in
common pipelines like <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/tutorials" class="external-link">10x
Genomics’ CellRanger</a> can usually be found in the documentation that
came with the sequencing material.</p>
<p>The main point is: if the sequencing outputs were provided to you by
someone else, make sure to communicate with them about what
pre-processing steps have been performed, if any.</p>
</div>
</div>
</div>
<p>We can visualize barcode read totals to visualize the distinction
between empty droplets and properly profiled single cells in a so-called
“knee plot”:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcrank</span> <span class="op">&lt;-</span> <span class="fu">barcodeRanks</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Only showing unique points for plotting speed.</span></span>
<span><span class="va">uniq</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu">duplicated</span><span class="op">(</span><span class="va">bcrank</span><span class="op">$</span><span class="va">rank</span><span class="op">)</span></span>
<span></span>
<span><span class="va">line_df</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>cutoff <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">bcrank</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      value  <span class="op">=</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">bcrank</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">bcrank</span><span class="op">[</span><span class="va">uniq</span>,<span class="op">]</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">rank</span>, <span class="va">total</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_hline</span><span class="op">(</span>data <span class="op">=</span> <span class="va">line_df</span>,</span>
<span>               <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">cutoff</span>,</span>
<span>                   yintercept <span class="op">=</span> <span class="va">value</span><span class="op">)</span>,</span>
<span>               lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Total UMI count"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The distribution of total counts (called the unique molecular
identifier or UMI count) exhibits a sharp transition between barcodes
with large and small total counts, probably corresponding to
cell-containing and empty droplets respectively.</p>
<p>A simple approach would be to apply a threshold on the total count to
only retain those barcodes with large totals. However, this may
unnecessarily discard libraries derived from cell types with low RNA
content.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>What is the median number of total counts in the raw data?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">median</span><span class="op">(</span><span class="va">bcrank</span><span class="op">$</span><span class="va">total</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2</code></pre>
</div>
<p>Just 2! Clearly many barcodes produce practically no output.</p>
<!-- This is a direct application challenge of of  "It is likely that most of these droplets are empty and are capturing only ambient or background RNA" and a synthesis challenge of the previous episode where they learned to access columns of DataFrames. -->
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="testing-for-empty-droplets">Testing for empty droplets<a class="anchor" aria-label="anchor" href="#testing-for-empty-droplets"></a>
</h3>
<p>A better approach is to test whether the expression profile for each
cell barcode is significantly different from the ambient RNA pool<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Any
significant deviation indicates that the barcode corresponds to a
cell-containing droplet. This allows us to discriminate between
well-sequenced empty droplets and droplets derived from cells with
little RNA, both of which would have similar total counts.</p>
<p>We call cells at a false discovery rate (FDR) of 0.1%, meaning that
no more than 0.1% of our called barcodes should be empty droplets on
average.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># emptyDrops performs Monte Carlo simulations to compute p-values,</span></span>
<span><span class="co"># so we need to set the seed to obtain reproducible results.</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># this may take a few minutes</span></span>
<span><span class="va">e.out</span> <span class="op">&lt;-</span> <span class="fu">emptyDrops</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">e.out</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.001</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Mode   FALSE    TRUE    NA's
logical    6184    3131  513239 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu">which</span><span class="op">(</span><span class="va">e.out</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 3131
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(3131): AAACCTGAGACTGTAA AAACCTGAGATGCCTT ... TTTGTCAGTCTGATTG
  TTTGTCATCTGAGTGT
colData names(0):
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>The result confirms our expectation: only 3,131 droplets contain a
cell, while the large majority of droplets are empty.</p>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> Setting the Random Seed </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<p>Whenever your code involves the generation of random numbers, it’s a
good practice to set the random seed in R with
<code>set.seed()</code>.</p>
<p>Setting the seed to a specific value (in the above example to 100)
will cause the pseudo-random number generator to return the same
pseudo-random numbers in the same order.</p>
<p>This allows us to write code with reproducible results, despite
technically involving the generation of (pseudo-)random numbers.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<hr class="half-width">
<p>While we have removed empty droplets, this does not necessarily imply
that all the cell-containing droplets should be kept for downstream
analysis. In fact, some droplets could contain low-quality samples, due
to cell damage or failure in library preparation.</p>
<p>Retaining these low-quality samples in the analysis could be
problematic as they could:</p>
<ul>
<li>form their own cluster, complicating the interpretation of the
results</li>
<li>interfere with variance estimation and principal component
analysis</li>
<li>contain contaminating transcripts from ambient RNA</li>
</ul>
<p>To mitigate these problems, we can check a few quality control (QC)
metrics and, if needed, remove low-quality samples.</p>
<div class="section level3">
<h3 id="choice-of-quality-control-metrics">Choice of quality control metrics<a class="anchor" aria-label="anchor" href="#choice-of-quality-control-metrics"></a>
</h3>
<p>There are many possible ways to define a set of quality control
metrics, see for instance <a href="reference.html#litref">Cole 2019</a>.
Here, we keep it simple and consider only:</p>
<ul>
<li>the <em>library size</em>, defined as the total sum of counts across
all relevant features <em>for each cell</em>;</li>
<li>the number of expressed features in each cell, defined as the number
of endogenous genes with non-zero counts for that cell;</li>
<li>the proportion of reads mapped to genes in the mitochondrial
genome.</li>
</ul>
<p>In particular, high proportions of mitochondrial genes are indicative
of poor-quality cells, presumably because of loss of cytoplasmic RNA
from perforated cells. The reasoning is that, in the presence of modest
damage, the holes in the cell membrane permit efflux of individual
transcript molecules but are too small to allow mitochondria to escape,
leading to a relative enrichment of mitochondrial transcripts. For
single-nucleus RNA-seq experiments, high proportions are also useful as
they can mark cells where the cytoplasm has not been successfully
stripped.</p>
<p>First, we need to identify mitochondrial genes. We use the available
<code>EnsDb</code> mouse package available in Bioconductor, but a more
updated version of Ensembl can be used through the
<code>AnnotationHub</code> or <code>biomaRt</code> packages.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr.loc</span> <span class="op">&lt;-</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">EnsDb.Mmusculus.v79</span>,</span>
<span>                  keys    <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>                  keytype <span class="op">=</span> <span class="st">"GENEID"</span>, </span>
<span>                  column  <span class="op">=</span> <span class="st">"SEQNAME"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu">which</span><span class="op">(</span><span class="va">chr.loc</span> <span class="op">==</span> <span class="st">"MT"</span><span class="op">)</span></span></code></pre>
</div>
<p>We can use the <code>scuttle</code> package to compute a set of
quality control metrics, specifying that we want to use the
mitochondrial genes as a special set of features.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">perCellQCMetrics</span><span class="op">(</span><span class="va">sce</span>, subsets <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>Mito <span class="op">=</span> <span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3131 rows and 6 columns
                       sum  detected subsets_Mito_sum subsets_Mito_detected
                 &lt;numeric&gt; &lt;integer&gt;        &lt;numeric&gt;             &lt;integer&gt;
AAACCTGAGACTGTAA     27577      5418              471                    10
AAACCTGAGATGCCTT     29309      5405              679                    10
AAACCTGAGCAGCCTC     28795      5218              480                    12
AAACCTGCATACTCTT     34794      4781              496                    12
AAACCTGGTGGTACAG       262       229                0                     0
...                    ...       ...              ...                   ...
TTTGGTTTCGCCATAA     38398      6020              252                    12
TTTGTCACACCCTATC      3013      1451              123                     9
TTTGTCACATTCTCAT      1472       675              599                    11
TTTGTCAGTCTGATTG       361       293                0                     0
TTTGTCATCTGAGTGT       267       233               16                     6
                 subsets_Mito_percent     total
                            &lt;numeric&gt; &lt;numeric&gt;
AAACCTGAGACTGTAA              1.70795     27577
AAACCTGAGATGCCTT              2.31669     29309
AAACCTGAGCAGCCTC              1.66696     28795
AAACCTGCATACTCTT              1.42553     34794
AAACCTGGTGGTACAG              0.00000       262
...                               ...       ...
TTTGGTTTCGCCATAA             0.656284     38398
TTTGTCACACCCTATC             4.082310      3013
TTTGTCACATTCTCAT            40.692935      1472
TTTGTCAGTCTGATTG             0.000000       361
TTTGTCATCTGAGTGT             5.992509       267</code></pre>
</div>
<p>Now that we have computed the metrics, we have to decide on
thresholds to define high- and low-quality samples. We could check how
many cells are above/below a certain fixed threshold. For instance,</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
FALSE  TRUE
 2477   654 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">subsets_Mito_percent</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
FALSE  TRUE
 2761   370 </code></pre>
</div>
<p>or we could look at the distribution of such metrics and use a data
adaptive threshold.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">detected</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
     98    4126    5168    4455    5670    7908 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">subsets_Mito_percent</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
  0.000   1.155   1.608   5.079   2.182  66.968 </code></pre>
</div>
<p>We can use the <code>perCellQCFilters</code> function to apply a set
of common adaptive filters to identify low-quality cells. By default, we
consider a value to be an outlier if it is more than 3 median absolute
deviations (MADs) from the median in the “problematic” direction. This
is loosely motivated by the fact that such a filter will retain 99% of
non-outlier values that follow a normal distribution.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reasons</span> <span class="op">&lt;-</span> <span class="fu">perCellQCFilters</span><span class="op">(</span><span class="va">df</span>, sub.fields <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">reasons</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3131 rows and 4 columns
         low_lib_size   low_n_features high_subsets_Mito_percent   discard
     &lt;outlier.filter&gt; &lt;outlier.filter&gt;          &lt;outlier.filter&gt; &lt;logical&gt;
1               FALSE            FALSE                     FALSE     FALSE
2               FALSE            FALSE                     FALSE     FALSE
3               FALSE            FALSE                     FALSE     FALSE
4               FALSE            FALSE                     FALSE     FALSE
5                TRUE             TRUE                     FALSE      TRUE
...               ...              ...                       ...       ...
3127            FALSE            FALSE                     FALSE     FALSE
3128             TRUE             TRUE                      TRUE      TRUE
3129             TRUE             TRUE                      TRUE      TRUE
3130             TRUE             TRUE                     FALSE      TRUE
3131             TRUE             TRUE                      TRUE      TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">discard</span> <span class="op">&lt;-</span> <span class="va">reasons</span><span class="op">$</span><span class="va">discard</span></span></code></pre>
</div>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Maybe our sample preparation was poor and we want the QC to be more
strict. How could we change the set the QC filtering to use 2.5 MADs as
the threshold for outlier calling?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>You set <code>nmads = 2.5</code> like so:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reasons_strict</span> <span class="op">&lt;-</span> <span class="fu">perCellQCFilters</span><span class="op">(</span><span class="va">df</span>, sub.fields <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span>, nmads <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre>
</div>
<p>You would then need to reassign the <code>discard</code> column as
well, but we’ll stick with the 3 MADs default for now.
<!-- This is a direct application of what was just shown --></p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="diagnostic-plots">Diagnostic plots<a class="anchor" aria-label="anchor" href="#diagnostic-plots"></a>
</h3>
<p>It is always a good idea to check the distribution of the QC metrics
and to visualize the cells that were removed, to identify possible
problems with the procedure. In particular, we expect to have few
outliers and with a marked difference from “regular” cells (e.g., a
bimodal distribution or a long tail). Moreover, if there are too many
discarded cells, further exploration might be needed.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>, y <span class="op">=</span> <span class="st">"sum"</span>, colour_by <span class="op">=</span> <span class="st">"discard"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Total count"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>, y <span class="op">=</span> <span class="st">"detected"</span>, colour_by <span class="op">=</span> <span class="st">"discard"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Detected features"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-11-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>, y <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span>, colour_by <span class="op">=</span> <span class="st">"discard"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Mito percent"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-11-3.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>While the univariate distribution of QC metrics can give some insight
on the quality of the sample, often looking at the bivariate
distribution of QC metrics is useful, e.g., to confirm that there are no
cells with both large total counts and large mitochondrial counts, to
ensure that we are not inadvertently removing high-quality cells that
happen to be highly metabolically active.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>,  x <span class="op">=</span><span class="st">"sum"</span>, y <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span>, colour_by <span class="op">=</span> <span class="st">"discard"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It could also be a good idea to perform a differential expression
analysis between retained and discarded cells to check wether we are
removing an unusual cell population rather than low-quality libraries
(see <a href="https://bioconductor.org/books/release/OSCA.advanced/quality-control-redux.html#qc-discard-cell-types" class="external-link">Section
1.5 of OSCA advanced</a>).</p>
<p>Once we are happy with the results, we can discard the low-quality
cells by subsetting the original object.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">sce</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 2437
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(2437): AAACCTGAGACTGTAA AAACCTGAGATGCCTT ... TTTGGTTTCAGTCAGT
  TTTGGTTTCGCCATAA
colData names(7): sum detected ... total discard
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<hr class="half-width">
<p>Systematic differences in sequencing coverage between libraries are
often observed in single-cell RNA sequencing data. They typically arise
from technical differences in cDNA capture or PCR amplification
efficiency across cells, attributable to the difficulty of achieving
consistent library preparation with minimal starting material<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.
Normalization aims to remove these differences such that they do not
interfere with comparisons of the expression profiles between cells. The
hope is that the observed heterogeneity or differential expression
within the cell population are driven by biology and not technical
biases.</p>
<p>We will mostly focus our attention on scaling normalization, which is
the simplest and most commonly used class of normalization strategies.
This involves dividing all counts for each cell by a cell-specific
scaling factor, often called a <em>size factor</em>. The assumption here
is that any cell-specific bias (e.g., in capture or amplification
efficiency) affects all genes equally via scaling of the expected mean
count for that cell. The size factor for each cell represents the
estimate of the relative bias in that cell, so division of its counts by
its size factor should remove that bias. The resulting “normalized
expression values” can then be used for downstream analyses such as
clustering and dimensionality reduction.</p>
<p>The simplest and most natural strategy would be to normalize by the
total sum of counts across all genes for each cell. This is often called
the <em>library size normalization</em>.</p>
<p>The <em>library size factor</em> for each cell is directly
proportional to its library size where the proportionality constant is
defined such that the mean size factor across all cells is equal to 1.
This ensures that the normalized expression values are on the same scale
as the original counts.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lib.sf</span> <span class="op">&lt;-</span> <span class="fu">librarySizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">lib.sf</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.2730  0.7879  0.9600  1.0000  1.1730  2.5598 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sf_df</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>size_factor <span class="op">=</span> <span class="va">lib.sf</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">sf_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">size_factor</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="normalization-by-deconvolution">Normalization by deconvolution<a class="anchor" aria-label="anchor" href="#normalization-by-deconvolution"></a>
</h3>
<p>Library size normalization is not optimal, as it assumes that the
total sum of UMI counts differ between cells only for technical and not
biological reasons. This can be a problem if a highly-expressed subset
of genes is differentially expressed between cells or cell types.</p>
<p>Several robust normalization methods have been proposed for bulk
RNA-seq. However, these methods may underperform in single-cell data due
to the dominance of low and zero counts. To overcome this, one solution
is to <em>pool</em> counts from many cells to increase the size of the
counts for accurate size factor estimation<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. Pool-based size
factors are then <em>deconvolved</em> into cell-based factors for
normalization of each cell’s expression profile.</p>
<p>We use a pre-clustering step: cells in each cluster are normalized
separately and the size factors are rescaled to be comparable across
clusters. This avoids the assumption that most genes are non-DE across
the entire population – only a non-DE majority is required between pairs
of clusters, which is a weaker assumption for highly heterogeneous
populations.</p>
<p>Note that while we focus on normalization by deconvolution here, many
other methods have been proposed and lead to similar performance (see <a href="reference.html#litref">Borella 2022</a> for a comparative
review).</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">quickCluster</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>clust
  1   2   3   4   5   6   7   8   9  10  11  12  13
273 159 250 122 187 201 154 252 152 169 199 215 104 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconv.sf</span> <span class="op">&lt;-</span> <span class="fu">pooledSizeFactors</span><span class="op">(</span><span class="va">sce</span>, cluster <span class="op">=</span> <span class="va">clust</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">deconv.sf</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.3100  0.8028  0.9626  1.0000  1.1736  2.7858 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sf_df</span><span class="op">$</span><span class="va">deconv_sf</span> <span class="op">&lt;-</span> <span class="va">deconv.sf</span></span>
<span></span>
<span><span class="va">sf_df</span><span class="op">$</span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="va">clust</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">sf_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">size_factor</span>, <span class="va">deconv_sf</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_abline</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">clust</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Once we have computed the size factors, we compute the normalized
expression values for each cell by dividing the count for each gene with
the appropriate size factor for that cell. Since we are typically going
to work with log-transformed counts, the function
<code>logNormCounts</code> also log-transforms the normalized values,
creating a new assay called <code>logcounts</code>.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">deconv.sf</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 2437
metadata(0):
assays(2): counts logcounts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(2437): AAACCTGAGACTGTAA AAACCTGAGATGCCTT ... TTTGGTTTCAGTCAGT
  TTTGGTTTCGCCATAA
colData names(8): sum detected ... discard sizeFactor
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Fill in the blanks for normalization that uses simpler library size
factors instead of deconvolution.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a>____ <span class="ot">&lt;-</span> <span class="fu">____SizeFactors</span>(sce)</span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" tabindex="-1"></a><span class="fu">sizeFactors</span>(sce) <span class="ot">&lt;-</span> ____</span>
<span id="cb41-4"><a href="#cb41-4" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">____</span>(sce)</span>
<span id="cb41-6"><a href="#cb41-6" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" tabindex="-1"></a>sce</span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lib.sf</span> <span class="op">&lt;-</span> <span class="fu">librarySizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">lib.sf</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<p>If you run this chunk, make sure to go back and re-run the
normalization with deconvolution normalization if you want your work to
align with the rest of this episode.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="feature-selection">Feature Selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h2>
<hr class="half-width">
<p>The typical next steps in the analysis of single-cell data are
dimensionality reduction and clustering, which involve measuring the
similarity between cells.</p>
<p>The choice of genes to use in this calculation has a major impact on
the results. We want to select genes that contain useful information
about the biology of the system while removing genes that contain only
random noise. This aims to preserve interesting biological structure
without the variance that obscures that structure, and to reduce the
size of the data to improve computational efficiency of later steps.</p>
<div class="section level3">
<h3 id="quantifying-per-gene-variation">Quantifying per-gene variation<a class="anchor" aria-label="anchor" href="#quantifying-per-gene-variation"></a>
</h3>
<p>The simplest approach to feature selection is to select the most
variable genes based on their log-normalized expression across the
population. This is motivated by practical idea that if we’re going to
try to explain variation in gene expression by biological factors, those
genes need to have variance to explain.</p>
<p>Calculation of the per-gene variance is simple but feature selection
requires modeling of the mean-variance relationship. The
log-transformation is not a variance stabilizing transformation in most
cases, which means that the total variance of a gene is driven more by
its abundance than its underlying biological heterogeneity. To account
for this, the <code>modelGeneVar</code> function fits a trend to the
variance with respect to abundance across all genes.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec.sce</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit.sce</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">dec.sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mean_var_df</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>mean <span class="op">=</span> <span class="va">fit.sce</span><span class="op">$</span><span class="va">mean</span>,</span>
<span>                          var  <span class="op">=</span> <span class="va">fit.sce</span><span class="op">$</span><span class="va">var</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">mean_var_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">mean</span>, <span class="va">var</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_function</span><span class="op">(</span>fun <span class="op">=</span> <span class="va">fit.sce</span><span class="op">$</span><span class="va">trend</span>, </span>
<span>                  color <span class="op">=</span> <span class="st">"dodgerblue"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Mean of log-expression"</span>,</span>
<span>         y <span class="op">=</span> <span class="st">"Variance of log-expression"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-20-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The blue line represents the uninteresting “technical” variance for
any given gene abundance. The genes with a lot of additional variance
exhibit interesting “biological” variation.</p>
</div>
<div class="section level3">
<h3 id="selecting-highly-variable-genes">Selecting highly variable genes<a class="anchor" aria-label="anchor" href="#selecting-highly-variable-genes"></a>
</h3>
<p>The next step is to select the subset of HVGs to use in downstream
analyses. A larger set will assure that we do not remove important
genes, at the cost of potentially increasing noise. Typically, we
restrict ourselves to the top <span class="math inline">\(n\)</span>
genes, here we chose <span class="math inline">\(n = 1000\)</span>, but
this choice should be guided by prior biological knowledge; for
instance, we may expect that only about 10% of genes to be
differentially expressed across our cell populations and hence select
10% of genes as highly variable (e.g., by setting
<code>prop = 0.1</code>).</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvg.sce.var</span> <span class="op">&lt;-</span> <span class="fu">getTopHVGs</span><span class="op">(</span><span class="va">dec.sce</span>, n <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">hvg.sce.var</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919"
[4] "ENSMUSG00000052187" "ENSMUSG00000048583" "ENSMUSG00000051855"</code></pre>
</div>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Imagine you have data that were prepared by three people with varying
level of experience, which leads to varying technical noise. How can you
account for this blocking structure when selecting HVGs?</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p><code>modelGeneVar()</code> can take a <code>block</code>
argument.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>Use the <code>block</code> argument in the call to
<code>modelGeneVar()</code> like so. We don’t have experimenter
information in this dataset, so in order to have some names to work with
we assign them randomly from a set of names.</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">experimenter</span> <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"Perry"</span>, <span class="st">"Merry"</span>, <span class="st">"Gary"</span><span class="op">)</span>,</span>
<span>                          replace <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                          size <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">blocked_variance_df</span> <span class="op">=</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, </span>
<span>                                   block <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">experimenter</span><span class="op">)</span></span></code></pre>
</div>
<p>Blocked models are evaluated on each block separately then combined.
If the experimental groups are related in some structured way, it may be
preferable to use the <code>design</code> argument. See
<code>?modelGeneVar</code> for more detail.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="dimensionality-reduction">Dimensionality Reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h2>
<hr class="half-width">
<p>Many scRNA-seq analysis procedures involve comparing cells based on
their expression values across multiple genes. For example, clustering
aims to identify cells with similar transcriptomic profiles by computing
Euclidean distances across genes. In these applications, each individual
gene represents a dimension of the data, hence we can think of the data
as “living” in a ten-thousand-dimensional space.</p>
<p>As the name suggests, dimensionality reduction aims to reduce the
number of dimensions, while preserving as much as possible of the
original information. This obviously reduces the computational work
(e.g., it is easier to compute distance in lower-dimensional spaces),
and more importantly leads to less noisy and more interpretable results
(cf. the <em>curse of dimensionality</em>).</p>
<div class="section level3">
<h3 id="principal-component-analysis-pca">Principal Component Analysis (PCA)<a class="anchor" aria-label="anchor" href="#principal-component-analysis-pca"></a>
</h3>
<p>Principal component analysis (PCA) is a dimensionality reduction
technique that provides a parsimonious summarization of the data by
replacing the original variables (genes) by fewer linear combinations of
these variables, that are orthogonal and have successively maximal
variance. Such linear combinations seek to “separate out” the
observations (cells), while losing as little information as
possible.</p>
<p>Without getting into the technical details, one nice feature of PCA
is that the principal components (PCs) are ordered by how much variance
of the original data they “explain”. Furthermore, by focusing on the top
<span class="math inline">\(k\)</span> PC we are focusing on the most
important directions of variability, which hopefully correspond to
biological rather than technical variance. (It is however good practice
to check this by e.g. looking at correlation between technical QC
metrics and PCs).</p>
<p>One simple way to maximize our chance of capturing biological
variation is by computing the PCs starting from the highly variable
genes identified before.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">hvg.sce.var</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 2437
metadata(0):
assays(2): counts logcounts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(2437): AAACCTGAGACTGTAA AAACCTGAGATGCCTT ... TTTGGTTTCAGTCAGT
  TTTGGTTTCGCCATAA
colData names(8): sum detected ... discard sizeFactor
reducedDimNames(1): PCA
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>By default, <code>runPCA</code> computes the first 50 principal
components. We can check how much original variability they explain.
These values are stored in the attributes of the <code>percentVar</code>
reducedDim:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pct_var_df</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>PC <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span>,</span>
<span>                         pct_var <span class="op">=</span> <span class="fu">attr</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pct_var_df</span>,</span>
<span>       <span class="fu">aes</span><span class="op">(</span><span class="va">PC</span>, <span class="va">pct_var</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Variance explained (%)"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-24-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You can see the first two PCs capture the largest amount of
variation, but in this case you have to take the first 8 PCs before
you’ve captured 50% of the total.</p>
<p>And we can of course visualize the first 2-3 components, perhaps
color-coding each point by an interesting feature, in this case the
total number of UMIs per cell.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-25-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It can be helpful to compare pairs of PCs. This can be done with the
<code>ncomponents</code> argument to <code>plotReducedDim()</code>. For
example if one batch or cell type splits off on a particular PC, this
can help visualize the effect of that.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-26-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level3">
<h3 id="non-linear-methods">Non-linear methods<a class="anchor" aria-label="anchor" href="#non-linear-methods"></a>
</h3>
<p>While PCA is a simple and effective way to visualize (and interpret!)
scRNA-seq data, non-linear methods such as t-SNE (<em>t-stochastic
neighbor embedding</em>) and UMAP (<em>uniform manifold approximation
and projection</em>) have gained much popularity in the literature.</p>
<p>These methods attempt to find a low-dimensional representation of the
data that attempt to preserve pair-wise distance and structure in
high-dimensional gene space as best as possible.</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-27-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">111</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-28-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It is easy to over-interpret t-SNE and UMAP plots. We note that the
relative sizes and positions of the visual clusters may be misleading,
as they tend to inflate dense clusters and compress sparse ones, such
that we cannot use the size as a measure of subpopulation
heterogeneity.</p>
<p>In addition, these methods are not guaranteed to preserve the global
structure of the data (e.g., the relative locations of non-neighboring
clusters), such that we cannot use their positions to determine
relationships between distant clusters.</p>
<p>Note that the <code>sce</code> object now includes all the computed
dimensionality reduced representations of the data for ease of reusing
and replotting without the need for recomputing. Note the added
<code>reducedDimNames</code> row when printing <code>sce</code>
here:</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 2437
metadata(0):
assays(2): counts logcounts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(2437): AAACCTGAGACTGTAA AAACCTGAGATGCCTT ... TTTGGTTTCAGTCAGT
  TTTGGTTTCGCCATAA
colData names(8): sum detected ... discard sizeFactor
reducedDimNames(3): PCA TSNE UMAP
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>Despite their shortcomings, t-SNE and UMAP can be useful
visualization techniques. When using them, it is important to consider
that they are stochastic methods that involve a random component (each
run will lead to different plots) and that there are key parameters to
be set that change the results substantially (e.g., the “perplexity”
parameter of t-SNE).</p>
<div id="challenge5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Re-run the UMAP for the same sample starting from the pre-processed
data (i.e. not <code>type = "raw"</code>). What looks the same? What
looks different?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">111</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce5</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">sce5</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce5</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce5</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">sce5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">sce5</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-30-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Given that it’s the same cells processed through a very similar
pipeline, the result should look very similar. There’s a slight
difference in the total number of cells, probably because the official
processing pipeline didn’t use the exact same random seed / QC arguments
as us.</p>
<p>But you’ll notice that even though the shape of the structures are
similar, they look slightly distorted. If the upstream QC parameters
change, the downstream output visualizations will also change.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="doublet-identification">Doublet identification<a class="anchor" aria-label="anchor" href="#doublet-identification"></a>
</h2>
<hr class="half-width">
<p><em>Doublets</em> are artifactual libraries generated from two cells.
They typically arise due to errors in cell sorting or capture.
Specifically, in droplet-based protocols, it may happen that two cells
are captured in the same droplet.</p>
<p>Doublets are obviously undesirable when the aim is to characterize
populations at the single-cell level. In particular, doublets can be
mistaken for intermediate populations or transitory states that do not
actually exist. Thus, it is desirable to identify and remove doublet
libraries so that they do not compromise interpretation of the
results.</p>
<p>It is not easy to computationally identify doublets as they can be
hard to distinguish from transient states and/or cell populations with
high RNA content. When possible, it is good to rely on experimental
strategies to minimize doublets, e.g., by using genetic variation (e.g.,
pooling multiple donors in one run) or antibody tagging (e.g.,
CITE-seq).</p>
<p>There are several computational methods to identify doublets; we
describe only one here based on in-silico simulation of doublets.</p>
<div class="section level3">
<h3 id="computing-doublet-densities">Computing doublet densities<a class="anchor" aria-label="anchor" href="#computing-doublet-densities"></a>
</h3>
<p>At a high level, the algorithm can be defined by the following
steps:</p>
<ol style="list-style-type: decimal">
<li>Simulate thousands of doublets by adding together two randomly
chosen single-cell profiles.</li>
<li>For each original cell, compute the density of simulated doublets in
the surrounding neighborhood.</li>
<li>For each original cell, compute the density of other observed cells
in the neighborhood.</li>
<li>Return the ratio between the two densities as a “doublet score” for
each cell.</li>
</ol>
<p>Intuitively, if a “cell” is surrounded only by simulated doublets is
very likely to be a doublet itself.</p>
<p>This approach is implemented below using the <code>scDblFinder</code>
library. We then visualize the scores in a t-SNE plot.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dbl.dens</span> <span class="op">&lt;-</span> <span class="fu">computeDoubletDensity</span><span class="op">(</span><span class="va">sce</span>, subset.row <span class="op">=</span> <span class="va">hvg.sce.var</span>,</span>
<span>                                  d <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">dbl.dens</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max.
 0.04874  0.28757  0.46790  0.65614  0.82371 14.88032 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">DoubletScore</span> <span class="op">&lt;-</span> <span class="va">dbl.dens</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"DoubletScore"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-31-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can explicitly convert this into doublet calls by identifying
large outliers for the score within each sample. Here we use the
“griffiths” method to do so.</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dbl.calls</span> <span class="op">&lt;-</span> <span class="fu">doubletThresholding</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span>score <span class="op">=</span> <span class="va">dbl.dens</span><span class="op">)</span>,</span>
<span>                                 method <span class="op">=</span> <span class="st">"griffiths"</span>,</span>
<span>                                 returnType <span class="op">=</span> <span class="st">"call"</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">dbl.calls</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>singlet doublet
   2124     313 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">doublet</span> <span class="op">&lt;-</span> <span class="va">dbl.calls</span></span>
<span></span>
<span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>, y <span class="op">=</span> <span class="st">"DoubletScore"</span>, colour_by <span class="op">=</span> <span class="st">"doublet"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-32-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"doublet"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-32-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>One way to determine whether a cell is in a real transient state or
it is a doublet is to check the number of detected genes and total UMI
counts.</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"detected"</span>, <span class="st">"sum"</span>, colour_by <span class="op">=</span> <span class="st">"DoubletScore"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-33-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotColData</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"detected"</span>, <span class="st">"sum"</span>, colour_by <span class="op">=</span> <span class="st">"doublet"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-33-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In this case, we only have a few doublets at the periphery of some
clusters. It could be fine to keep the doublets in the analysis, but it
may be safer to discard them to avoid biases in downstream analysis
(e.g., differential expression).</p>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<div id="exercise-1-normalization" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-normalization" class="callout-inner">
<h3 class="callout-title">Exercise 1: Normalization</h3>
<div class="callout-content">
<p>Here we used the deconvolution method implemented in
<code>scran</code> based on a previous clustering step. Use the
<code>pooledSizeFactors</code> to compute the size factors without
considering a preliminary clustering. Compare the resulting size factors
via a scatter plot. How do the results change? What are the risks of not
including clustering information?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconv.sf2</span> <span class="op">&lt;-</span> <span class="fu">pooledSizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="co"># dropped `cluster = clust` here</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">deconv.sf2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.2985  0.8142  0.9583  1.0000  1.1582  2.7054 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sf_df</span><span class="op">$</span><span class="va">deconv_sf2</span> <span class="op">&lt;-</span> <span class="va">deconv.sf2</span></span>
<span></span>
<span><span class="va">sf_df</span><span class="op">$</span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="va">clust</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">sf_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">deconv_sf</span>, <span class="va">deconv_sf2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_abline</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">clust</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">facet_wrap</span><span class="op">(</span><span class="fu">vars</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/eda_qc-rendered-unnamed-chunk-34-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You can see that we get largely similar results, though for clusters
3 and 9 there’s a slight deviation from the <code>y=x</code>
relationship because these clusters (which are fairly distinct from the
other clusters) are now being pooled with cells from other clusters.
This slightly violates the “majority non-DE genes” assumption.</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-pbmc-data" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-pbmc-data" class="callout-inner">
<h3 class="callout-title">Exercise 2: PBMC Data</h3>
<div class="callout-content">
<p>The <a href="https://www.bioconductor.org/packages/release/data/experiment/html/DropletTestFiles.html" class="external-link"><code>DropletTestFiles</code>
package</a> includes the raw output from Cell Ranger of the peripheral
blood mononuclear cell (PBMC) dataset from 10X Genomics, publicly
available from the 10X Genomics website. Repeat the analysis of this
vignette using those data.</p>
<p>The hint demonstrates how to identify, download, extract, and read
the data starting from the help documentation of
<code>?DropletTestFiles::listTestFiles</code>, but try working through
those steps on your own for extra challenge (they’re useful skills to
develop in practice).</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">DropletTestFiles</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">listTestFiles</span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"tenx-3.1.0-5k_pbmc_protein_v3"</span><span class="op">)</span> <span class="co"># look up the remote data path of the raw data</span></span>
<span></span>
<span><span class="va">raw_rdatapath</span> <span class="op">&lt;-</span> <span class="st">"DropletTestFiles/tenx-3.1.0-5k_pbmc_protein_v3/1.0.0/raw.tar.gz"</span></span>
<span></span>
<span><span class="va">local_path</span> <span class="op">&lt;-</span> <span class="fu">getTestFile</span><span class="op">(</span><span class="va">raw_rdatapath</span>, prefix <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">file.copy</span><span class="op">(</span><span class="va">local_path</span>, </span>
<span>          <span class="fu">paste0</span><span class="op">(</span><span class="va">local_path</span>, <span class="st">".tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">untar</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="va">local_path</span>, <span class="st">".tar.gz"</span><span class="op">)</span>,</span>
<span>      exdir <span class="op">=</span> <span class="fu">dirname</span><span class="op">(</span><span class="va">local_path</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">read10xCounts</span><span class="op">(</span><span class="fu">file.path</span><span class="op">(</span><span class="fu">dirname</span><span class="op">(</span><span class="va">local_path</span><span class="op">)</span>, <span class="st">"raw_feature_bc_matrix/"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<p>After getting the data, the steps are largely copy-pasted from above.
For the sake of simplicity the mitochondrial genes are identified by
gene symbols in the row data. Otherwise the steps are the same:</p>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e.out</span> <span class="op">&lt;-</span> <span class="fu">emptyDrops</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu">which</span><span class="op">(</span><span class="va">e.out</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Thankfully the data come with gene symbols, which we can use to identify mitochondrial genes:</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># QC metrics ----</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">perCellQCMetrics</span><span class="op">(</span><span class="va">sce</span>, subsets <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>Mito <span class="op">=</span> <span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">reasons</span> <span class="op">&lt;-</span> <span class="fu">perCellQCFilters</span><span class="op">(</span><span class="va">df</span>, sub.fields <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">reasons</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">discard</span> <span class="op">&lt;-</span> <span class="va">reasons</span><span class="op">$</span><span class="va">discard</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">sce</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Normalization ----</span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">quickCluster</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span>
<span></span>
<span><span class="va">deconv.sf</span> <span class="op">&lt;-</span> <span class="fu">pooledSizeFactors</span><span class="op">(</span><span class="va">sce</span>, cluster <span class="op">=</span> <span class="va">clust</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sizeFactors</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">deconv.sf</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection ----</span></span>
<span><span class="va">dec.sce</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hvg.sce.var</span> <span class="op">&lt;-</span> <span class="fu">getTopHVGs</span><span class="op">(</span><span class="va">dec.sce</span>, n <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Dimensionality reduction ----</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">hvg.sce.var</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Doublet finding ----</span></span>
<span><span class="va">dbl.dens</span> <span class="op">&lt;-</span> <span class="fu">computeDoubletDensity</span><span class="op">(</span><span class="va">sce</span>, subset.row <span class="op">=</span> <span class="va">hvg.sce.var</span>,</span>
<span>                                  d <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">DoubletScore</span> <span class="op">&lt;-</span> <span class="va">dbl.dens</span></span>
<span></span>
<span><span class="va">dbl.calls</span> <span class="op">&lt;-</span> <span class="fu">doubletThresholding</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span>score <span class="op">=</span> <span class="va">dbl.dens</span><span class="op">)</span>,</span>
<span>                                 method <span class="op">=</span> <span class="st">"griffiths"</span>,</span>
<span>                                 returnType <span class="op">=</span> <span class="st">"call"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">doublet</span> <span class="op">&lt;-</span> <span class="va">dbl.calls</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="extension-challenge-1-spike-ins" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-1-spike-ins" class="callout-inner">
<h3 class="callout-title">Extension challenge 1: Spike-ins</h3>
<div class="callout-content">
<p>Some sophisticated experiments perform additional steps so that they
can estimate size factors from so-called “spike-ins”. Judging by the
name, what do you think “spike-ins” are, and what additional steps are
required to use them?</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<p>Spike-ins are deliberately-introduced exogeneous RNA from an exotic
or synthetic source at a known concentration. This provides a known
signal to normalize against. Exotic (e.g. soil bacteria RNA in a study
of human cells) or synthetic RNA is used in order to avoid confusing
spike-in RNA with sample RNA. This has the obvious advantage of
accounting for cell-wise variation, but can substantially increase the
amount of sample-preparation work.</p>
</div>
</div>
</div>
</div>
<div id="extension-challenge-2-background-research" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="extension-challenge-2-background-research" class="callout-inner">
<h3 class="callout-title">Extension challenge 2: Background research</h3>
<div class="callout-content">
<p>Run an internet search for some of the most highly variable genes we
identified in the feature selection section. See if you can identify the
type of protein they produce or what sort of process they’re involved
in. Do they make biological sense to you?</p>
</div>
</div>
</div>
<div id="extension-challenge-3-reduced-dimensionality-representations" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-3-reduced-dimensionality-representations" class="callout-inner">
<h3 class="callout-title">Extension challenge 3: Reduced dimensionality representations</h3>
<div class="callout-content">
<p>Can dimensionality reduction techniques provide a perfectly accurate
representation of the data?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" data-bs-parent="#accordionSolution9" aria-labelledby="headingSolution9">
<div class="accordion-body">
<p>Mathematically, this would require the data to fall on a
two-dimensional plane (for linear methods like PCA) or a smooth 2D
manifold (for methods like UMAP). You can be confident that this will
never happen in real-world data, so the reduction from ~2500-dimensional
gene space to two-dimensional plot space always involves some degree of
information loss.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Empty droplets, i.e. droplets that do not contain intact cells and
that capture only ambient or background RNA, should be removed prior to
an analysis. The <code>emptyDrops</code> function from the <a href="https://bioconductor.org/packages/DropletUtils" class="external-link">DropletUtils</a>
package can be used to identify empty droplets.</li>
<li>Doublets, i.e. instances where two cells are captured in the same
droplet, should also be removed prior to an analysis. The
<code>computeDoubletDensity</code> and <code>doubletThresholding</code>
functions from the <a href="https://bioconductor.org/packages/scDblFinder" class="external-link">scDblFinder</a>
package can be used to identify doublets.</li>
<li>Quality control (QC) uses metrics such as library size, number of
expressed features, and mitochondrial read proportion, based on which
low-quality cells can be detected and filtered out. Diagnostic plots of
the chosen QC metrics are important to identify possible issues.</li>
<li>Normalization is required to account for systematic differences in
sequencing coverage between libraries and to make measurements
comparable between cells. Library size normalization is the most
commonly used normalization strategy, and involves dividing all counts
for each cell by a cell-specific scaling factor.</li>
<li>Feature selection aims at selecting genes that contain useful
information about the biology of the system while removing genes that
contain only random noise. Calculate per-gene variance with the
<code>modelGeneVar</code> function and select highly-variable genes with
<code>getTopHVGs</code>.</li>
<li>Dimensionality reduction aims at reducing the computational work and
at obtaining less noisy and more interpretable results. PCA is a simple
and effective linear dimensionality reduction technique that provides
interpretable results for further analysis such as clustering of cells.
Non-linear approaches such as UMAP and t-SNE can be useful for
visualization, but the resulting representations should not be used in
downstream analysis.</li>
</ul>
</div>
</div>
</div>
<div id="further-reading" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further Reading</h3>
<div class="callout-content">
<ul>
<li>OSCA book, Basics, <a href="https://bioconductor.org/books/release/OSCA.basic" class="external-link">Chapters
1-4</a>
</li>
<li>OSCA book, Advanced, <a href="https://bioconductor.org/books/release/OSCA.advanced/" class="external-link">Chapters
7-8</a>
</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] scDblFinder_1.18.0           scran_1.32.0
 [3] scater_1.32.1                scuttle_1.14.0
 [5] EnsDb.Mmusculus.v79_2.99.0   ensembldb_2.28.1
 [7] AnnotationFilter_1.28.0      GenomicFeatures_1.56.0
 [9] AnnotationDbi_1.66.0         ggplot2_3.5.1
[11] DropletUtils_1.24.0          MouseGastrulationData_1.18.0
[13] SpatialExperiment_1.14.0     SingleCellExperiment_1.26.0
[15] SummarizedExperiment_1.34.0  Biobase_2.64.0
[17] GenomicRanges_1.56.2         GenomeInfoDb_1.40.1
[19] IRanges_2.38.1               S4Vectors_0.42.1
[21] BiocGenerics_0.50.0          MatrixGenerics_1.16.0
[23] matrixStats_1.5.0            BiocStyle_2.32.1

loaded via a namespace (and not attached):
  [1] jsonlite_1.8.9            magrittr_2.0.3
  [3] ggbeeswarm_0.7.2          magick_2.8.5
  [5] farver_2.1.2              rmarkdown_2.29
  [7] BiocIO_1.14.0             zlibbioc_1.50.0
  [9] vctrs_0.6.5               memoise_2.0.1
 [11] Rsamtools_2.20.0          DelayedMatrixStats_1.26.0
 [13] RCurl_1.98-1.16           htmltools_0.5.8.1
 [15] S4Arrays_1.4.1            AnnotationHub_3.12.0
 [17] curl_6.2.0                BiocNeighbors_1.22.0
 [19] xgboost_1.7.8.1           Rhdf5lib_1.26.0
 [21] SparseArray_1.4.8         rhdf5_2.48.0
 [23] cachem_1.1.0              GenomicAlignments_1.40.0
 [25] igraph_2.1.4              mime_0.12
 [27] lifecycle_1.0.4           pkgconfig_2.0.3
 [29] rsvd_1.0.5                Matrix_1.7-2
 [31] R6_2.5.1                  fastmap_1.2.0
 [33] GenomeInfoDbData_1.2.12   digest_0.6.37
 [35] colorspace_2.1-1          dqrng_0.4.1
 [37] irlba_2.3.5.1             ExperimentHub_2.12.0
 [39] RSQLite_2.3.9             beachmat_2.20.0
 [41] labeling_0.4.3            filelock_1.0.3
 [43] httr_1.4.7                abind_1.4-8
 [45] compiler_4.4.2            bit64_4.6.0-1
 [47] withr_3.0.2               BiocParallel_1.38.0
 [49] viridis_0.6.5             DBI_1.2.3
 [51] HDF5Array_1.32.1          R.utils_2.12.3
 [53] MASS_7.3-64               rappdirs_0.3.3
 [55] DelayedArray_0.30.1       bluster_1.14.0
 [57] rjson_0.2.23              tools_4.4.2
 [59] vipor_0.4.7               beeswarm_0.4.0
 [61] R.oo_1.27.0               glue_1.8.0
 [63] restfulr_0.0.15           rhdf5filters_1.16.0
 [65] grid_4.4.2                Rtsne_0.17
 [67] cluster_2.1.8             generics_0.1.3
 [69] gtable_0.3.6              R.methodsS3_1.8.2
 [71] data.table_1.16.4         metapod_1.12.0
 [73] BiocSingular_1.20.0       ScaledMatrix_1.12.0
 [75] XVector_0.44.0            ggrepel_0.9.6
 [77] BiocVersion_3.19.1        pillar_1.10.1
 [79] limma_3.60.6              BumpyMatrix_1.12.0
 [81] dplyr_1.1.4               BiocFileCache_2.12.0
 [83] lattice_0.22-6            FNN_1.1.4.1
 [85] renv_1.1.0                rtracklayer_1.64.0
 [87] bit_4.5.0.1               tidyselect_1.2.1
 [89] locfit_1.5-9.11           Biostrings_2.72.1
 [91] knitr_1.49                gridExtra_2.3
 [93] ProtGenerics_1.36.0       edgeR_4.2.2
 [95] xfun_0.50                 statmod_1.5.0
 [97] UCSC.utils_1.0.0          lazyeval_0.2.2
 [99] yaml_2.3.10               evaluate_1.0.3
[101] codetools_0.2-20          tibble_3.2.1
[103] BiocManager_1.30.25       cli_3.6.3
[105] uwot_0.2.2                munsell_0.5.1
[107] Rcpp_1.0.14               dbplyr_2.5.0
[109] png_0.1-8                 XML_3.99-0.18
[111] parallel_4.4.2            blob_1.2.4
[113] sparseMatrixStats_1.16.0  bitops_1.0-9
[115] viridisLite_0.4.2         scales_1.3.0
[117] purrr_1.0.2               crayon_1.5.3
[119] rlang_1.1.5               formatR_1.14
[121] cowplot_1.1.3             KEGGREST_1.44.1          </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="reference.html#litref">Lun (2019)</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p><a href="reference.html#litref">Vallejos (2017)</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p><a href="reference.html#litref">Lun (2016)</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-cell_type_annotation"><p>Content from <a href="cell_type_annotation.html">Cell type annotation</a></p>
<hr>
<p>Last updated on 2025-02-07 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/cell_type_annotation.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we identify groups of cells with similar expression
profiles?</li>
<li>How can we identify genes that drive separation between these groups
of cells?</li>
<li>How to leverage reference datasets and known marker genes for the
cell type annotation of new datasets?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Identify groups of cells by clustering cells based on gene
expression patterns.</li>
<li>Identify marker genes through testing for differential expression
between clusters.</li>
<li>Annotate cell types through annotation transfer from reference
datasets.</li>
<li>Annotate cell types through marker gene set enrichment testing.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<hr class="half-width">
<p>Again we’ll start by loading the libraries we’ll be using:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">AUCell</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">SingleR</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="data-retrieval">Data retrieval<a class="anchor" aria-label="anchor" href="#data-retrieval"></a>
</h2>
<hr class="half-width">
<p>We’ll be using the fifth processed sample from the WT chimeric mouse
embryo data:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 2411
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(2411): cell_9769 cell_9770 ... cell_12178 cell_12179
colData names(11): cell barcode ... doub.density sizeFactor
reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>To speed up the computations, we take a random subset of 1,000
cells.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">ncol</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">ind</span><span class="op">]</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<hr class="half-width">
<p>The SCE object needs to contain log-normalized expression counts as
well as PCA coordinates in the reduced dimensions, so we compute those
here:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<hr class="half-width">
<p>Clustering is an unsupervised learning procedure that is used to
empirically define groups of cells with similar expression profiles. Its
primary purpose is to summarize complex scRNA-seq data into a digestible
format for human interpretation. This allows us to describe population
heterogeneity in terms of discrete labels that are easily understood,
rather than attempting to comprehend the high-dimensional manifold on
which the cells truly reside. After annotation based on marker genes,
the clusters can be treated as proxies for more abstract biological
concepts such as cell types or states.</p>
<p>Graph-based clustering is a flexible and scalable technique for
identifying coherent groups of cells in large scRNA-seq datasets. We
first build a graph where each node is a cell that is connected to its
nearest neighbors in the high-dimensional space. Edges are weighted
based on the similarity between the cells involved, with higher weight
given to cells that are more closely related. We then apply algorithms
to identify “communities” of cells that are more connected to cells in
the same community than they are to cells of different communities. Each
community represents a cluster that we can use for downstream
interpretation.</p>
<p>Here, we use the <code>clusterCells()</code> function from the <a href="https://bioconductor.org/packages/scran" class="external-link">scran</a> package to
perform graph-based clustering using the <a href="https://doi.org/10.1088/1742-5468/2008/10/P10008" class="external-link">Louvain
algorithm</a> for community detection. All calculations are performed
using the top PCs to take advantage of data compression and denoising.
This function returns a vector containing cluster assignments for each
cell in our <code>SingleCellExperiment</code> object. We use the
<code>colLabels()</code> function to assign the cluster labels as a
factor in the column data.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                               BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
  1   2   3   4   5   6   7   8   9  10  11  12
100 160  60 141  63  93  60 108  44  91  41  39 </code></pre>
</div>
<p>You can see we ended up with 12 clusters of varying sizes.</p>
<p>We can now overlay the cluster labels as color on a UMAP plot:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span>, color_by <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-cluster-viz-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Our clusters look semi-reasonable, but what if we wanted to make them
less granular? Look at the help documentation for
<code>?clusterCells</code> and <code>?NNGraphParam</code> to find out
what we’d need to change to get fewer, larger clusters.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>We see in the help documentation for <code>?clusterCells</code> that
all of the clustering algorithm details are handled through the
<code>BLUSPARAM</code> argument, which needs to provide a
<code>BlusterParam</code> object (of which <code>NNGraphParam</code> is
a sub-class). Each type of clustering algorithm will have some sort of
hyper-parameter that controls the granularity of the output clusters.
Looking at <code>?NNGraphParam</code> specifically, we see an argument
called <code>k</code> which is described as “An integer scalar
specifying the number of nearest neighbors to consider during graph
construction.” If the clustering process has to connect larger sets of
neighbors, the graph will tend to be cut into larger groups, resulting
in less granular clusters. Try the two code blocks above once more with
<code>k = 30</code>. Given their visual differences, do you think one
set of clusters is “right” and the other is “wrong”?</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">clust2</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                           BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>                                                    k <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span>, color_by <span class="op">=</span> <span class="st">"clust2"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-unnamed-chunk-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="marker-gene-detection">Marker gene detection<a class="anchor" aria-label="anchor" href="#marker-gene-detection"></a>
</h2>
<hr class="half-width">
<p>To interpret clustering results as obtained in the previous section,
we identify the genes that drive separation between clusters. These
marker genes allow us to assign biological meaning to each cluster based
on their functional annotation. In the simplest case, we have <em>a
priori</em> knowledge of the marker genes associated with particular
cell types, allowing us to treat the clustering as a proxy for cell type
identity.</p>
<p>The most straightforward approach to marker gene detection involves
testing for differential expression between clusters. If a gene is
strongly DE between clusters, it is likely to have driven the separation
of cells in the clustering algorithm.</p>
<p>Here, we use <code>scoreMarkers()</code> to perform pairwise
comparisons of gene expression, focusing on up-regulated (positive)
markers in one cluster when compared to another cluster.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">scoreMarkers</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of length 12
names(12): 1 2 3 4 5 6 7 8 9 10 11 12</code></pre>
</div>
<p>The resulting object contains a sorted marker gene list for each
cluster, in which the top genes are those that contribute the most to
the separation of that cluster from all other clusters.</p>
<p>Here, we inspect the ranked marker gene list for the first
cluster.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 29453 rows and 19 columns
               self.average other.average self.detected other.detected
                  &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;      &lt;numeric&gt;
Xkr4              0.0000000    0.00332819          0.00     0.00339804
Gm1992            0.0000000    0.00000000          0.00     0.00000000
Gm37381           0.0000000    0.00000000          0.00     0.00000000
Rp1               0.0000000    0.00000000          0.00     0.00000000
Sox17             0.0279547    0.25553459          0.02     0.12352446
...                     ...           ...           ...            ...
AC149090.1        0.3852624   0.347993521          0.33    0.281869817
DHRSX             0.4108022   0.495896614          0.35    0.395677179
Vmn2r122          0.0000000   0.000000000          0.00    0.000000000
CAAA01147332.1    0.0164546   0.000729715          0.01    0.000999001
tomato-td         0.6341678   0.636134988          0.51    0.491105734
               mean.logFC.cohen min.logFC.cohen median.logFC.cohen
                      &lt;numeric&gt;       &lt;numeric&gt;          &lt;numeric&gt;
Xkr4                  -0.035152       -0.208498          0.0000000
Gm1992                 0.000000        0.000000          0.0000000
Gm37381                0.000000        0.000000          0.0000000
Rp1                    0.000000        0.000000          0.0000000
Sox17                 -0.239182       -1.992655          0.0272241
...                         ...             ...                ...
AC149090.1           0.07484762      -0.1263241          0.0303362
DHRSX               -0.12286973      -0.4619613         -0.1249748
Vmn2r122             0.00000000       0.0000000          0.0000000
CAAA01147332.1       0.13453495       0.0656709          0.1414214
tomato-td            0.00484753      -0.2535145         -0.0173963
               max.logFC.cohen rank.logFC.cohen  mean.AUC   min.AUC median.AUC
                     &lt;numeric&gt;        &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
Xkr4                  0.000000             6949  0.498301  0.489247   0.500000
Gm1992                0.000000             6554  0.500000  0.500000   0.500000
Gm37381               0.000000             6554  0.500000  0.500000   0.500000
Rp1                   0.000000             6554  0.500000  0.500000   0.500000
Sox17                 0.200319             1482  0.447693  0.105167   0.498409
...                        ...              ...       ...       ...        ...
AC149090.1            0.427051             1685  0.519966  0.475000   0.507234
DHRSX                 0.130189             3431  0.472563  0.389878   0.470000
Vmn2r122              0.000000             6554  0.500000  0.500000   0.500000
CAAA01147332.1        0.141421             2438  0.504505  0.499560   0.505000
tomato-td             0.318068             2675  0.498386  0.427083   0.487582
                 max.AUC  rank.AUC mean.logFC.detected min.logFC.detected
               &lt;numeric&gt; &lt;integer&gt;           &lt;numeric&gt;          &lt;numeric&gt;
Xkr4                0.50      6882        -2.34997e-01       -1.58496e+00
Gm1992              0.50      6513        -2.91221e-17       -3.20343e-16
Gm37381             0.50      6513        -2.91221e-17       -3.20343e-16
Rp1                 0.50      6513        -2.91221e-17       -3.20343e-16
Sox17               0.51      3957        -6.27326e-01       -4.47721e+00
...                  ...       ...                 ...                ...
AC149090.1      0.588462      1932         2.49127e-01       -4.59278e-02
DHRSX           0.530054      2050        -1.51523e-01       -4.52151e-01
Vmn2r122        0.500000      6513        -2.91221e-17       -3.20343e-16
CAAA01147332.1  0.505000      4893         6.76367e-01       -6.64274e-02
tomato-td       0.576875      1840         7.02671e-02       -2.27601e-01
               median.logFC.detected max.logFC.detected rank.logFC.detected
                           &lt;numeric&gt;          &lt;numeric&gt;           &lt;integer&gt;
Xkr4                       0.0000000        3.20343e-16                5560
Gm1992                     0.0000000        3.20343e-16                5560
Gm37381                    0.0000000        3.20343e-16                5560
Rp1                        0.0000000        3.20343e-16                5560
Sox17                     -0.0892673        1.51602e+00                 341
...                              ...                ...                 ...
AC149090.1                 0.0565835        9.55592e-01                2039
DHRSX                     -0.1810960        2.28269e-01                3943
Vmn2r122                   0.0000000        3.20343e-16                5560
CAAA01147332.1             0.6780719        1.00000e+00                 898
tomato-td                  0.0612002        4.63438e-01                3705</code></pre>
</div>
<p>Each column contains summary statistics for each gene in the given
cluster. These are usually the mean/median/min/max of statistics like
Cohen’s <em>d</em> and AUC when comparing this cluster (cluster 1 in
this case) to all other clusters. <code>mean.AUC</code> is usually the
most important to check. AUC is the probability that a randomly selected
cell in cluster <em>A</em> has a greater expression of gene <em>X</em>
than a randomly selected cell in cluster <em>B</em>. You can set
<code>full.stats=TRUE</code> if you’d like the marker data frames to
retain list columns containing each statistic for each pairwise
comparison.</p>
<p>We can then inspect the top marker genes for the first cluster using
the <code>plotExpression</code> function from the <a href="https://bioconductor.org/packages/scater" class="external-link">scater</a> package.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c1_markers</span> <span class="op">&lt;-</span> <span class="va">markers</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">c1_markers</span><span class="op">$</span><span class="va">mean.AUC</span>, </span>
<span>             decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top.markers</span> <span class="op">&lt;-</span> <span class="fu">head</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">c1_markers</span><span class="op">[</span><span class="va">ord</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotExpression</span><span class="op">(</span><span class="va">sce</span>, </span>
<span>               features <span class="op">=</span> <span class="va">top.markers</span>, </span>
<span>               x        <span class="op">=</span> <span class="st">"label"</span>, </span>
<span>               color_by <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-plot-markers-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Clearly, not every marker gene distinguishes cluster 1 from every
other cluster. However, with a combination of multiple marker genes it’s
possible to clearly identify gene patterns that are unique to cluster 1.
It’s sort of like the 20 questions game - with answers to the right
questions about a cell (e.g. “Do you highly express Ptn?”), you can
clearly identify what cluster it falls in.</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Looking at the last plot, what clusters are most difficult to
distinguish from cluster 1? Now re-run the UMAP plot from the previous
section. Do the difficult-to-distinguish clusters make sense?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>You can see that at least among the top markers, cluster 6 (pale
green) tends to have the least separation from cluster 1.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span>, color_by <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Looking at the UMAP again, we can see that the marker gene overlap of
clusters 1 and 6 makes sense. They’re right next to each other on the
UMAP. They’re probably closely related cell types, and a less granular
clustering would probably lump them together.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h2>
<hr class="half-width">
<p>The most challenging task in scRNA-seq data analysis is arguably the
interpretation of the results. Obtaining clusters of cells is fairly
straightforward, but it is more difficult to determine what biological
state is represented by each of those clusters. Doing so requires us to
bridge the gap between the current dataset and prior biological
knowledge, and the latter is not always available in a consistent and
quantitative manner. Indeed, even the concept of a “cell type” is <a href="https://doi.org/10.1016/j.cels.2017.03.006" class="external-link">not clearly
defined</a>, with most practitioners possessing a “I’ll know it when I
see it” intuition that is not amenable to computational analysis. As
such, interpretation of scRNA-seq data is often manual and a common
bottleneck in the analysis workflow.</p>
<p>To expedite this step, we can use various computational approaches
that exploit prior information to assign meaning to an uncharacterized
scRNA-seq dataset. The most obvious sources of prior information are the
curated gene sets associated with particular biological processes, e.g.,
from the Gene Ontology (GO) or the Kyoto Encyclopedia of Genes and
Genomes (KEGG) collections. Alternatively, we can directly compare our
expression profiles to published reference datasets where each sample or
cell has already been annotated with its putative biological state by
domain experts. Here, we will demonstrate both approaches on the
wild-type chimera dataset.</p>
<div class="section level3">
<h3 id="assigning-cell-labels-from-reference-data">Assigning cell labels from reference data<a class="anchor" aria-label="anchor" href="#assigning-cell-labels-from-reference-data"></a>
</h3>
<p>A conceptually straightforward annotation approach is to compare the
single-cell expression profiles with previously annotated reference
datasets. Labels can then be assigned to each cell in our
uncharacterized test dataset based on the most similar reference
sample(s), for some definition of “similar”. This is a standard
classification challenge that can be tackled by standard machine
learning techniques such as random forests and support vector machines.
Any published and labelled RNA-seq dataset (bulk or single-cell) can be
used as a reference, though its reliability depends greatly on the
expertise of the original authors who assigned the labels in the first
place.</p>
<p>In this section, we will demonstrate the use of the <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
method for cell type annotation <a href="https://www.nature.com/articles/s41590-018-0276-y" class="external-link">Aran et al.,
2019</a>. This method assigns labels to cells based on the reference
samples with the highest Spearman rank correlations, using only the
marker genes between pairs of labels to focus on the relevant
differences between cell types. It also performs a fine-tuning step for
each cell where the correlations are recomputed with just the marker
genes for the top-scoring labels. This aims to resolve any ambiguity
between those labels by removing noise from irrelevant markers for other
labels. Further details can be found in the <a href="https://bioconductor.org/books/release/SingleRBook" class="external-link"><em>SingleR</em>
book</a> from which most of the examples here are derived.</p>
<div id="callout1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout</h3>
<div class="callout-content">
<p>Remember, the quality of reference-based cell type annotation can
only be as good as the cell type assignments in the reference. Garbage
in, garbage out. In practice, it’s worthwhile to spend time carefully
assessing your to make sure the original assignments make sense and that
it’s compatible with the query dataset you’re trying to annotate.</p>
</div>
</div>
</div>
<p>Here we take a single sample from <code>EmbryoAtlasData</code> as our
reference dataset. In practice you would want to take more/all samples,
possibly with batch-effect correction (see the next episode).</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu">EmbryoAtlasData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">29</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ref</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29452 7569
metadata(0):
assays(1): counts
rownames(29452): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000096730 ENSMUSG00000095742
rowData names(2): ENSEMBL SYMBOL
colnames(7569): cell_95727 cell_95728 ... cell_103294 cell_103295
colData names(17): cell barcode ... colour sizeFactor
reducedDimNames(2): pca.corrected umap
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>In order to reduce the computational load, we subsample the dataset
to 1,000 cells.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">ncol</span><span class="op">(</span><span class="va">ref</span><span class="op">)</span>, <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">ref</span><span class="op">[</span>,<span class="va">ind</span><span class="op">]</span></span></code></pre>
</div>
<p>You can see we have an assortment of different cell types in the
reference (with varying frequency):</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu">sort</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">ref</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tab</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
  Forebrain/Midbrain/Hindbrain                     Erythroid3
                           131                             75
             Paraxial mesoderm                            NMP
                            69                             51
                  ExE mesoderm               Surface ectoderm
                            49                             47
                     Allantois                     Mesenchyme
                            46                             45
                   Spinal cord            Pharyngeal mesoderm
                            45                             41
                  ExE endoderm                   Neural crest
                            38                             35
                           Gut Haematoendothelial progenitors
                            30                             27
         Intermediate mesoderm                 Cardiomyocytes
                            27                             26
              Somitic mesoderm                    Endothelium
                            25                             23
                    Erythroid2                  Def. endoderm
                            11                              3
                    Erythroid1            Blood progenitors 1
                             2                              1
           Blood progenitors 2                Caudal Mesoderm
                             1                              1
                           PGC
                             1 </code></pre>
</div>
<p>We need the normalized log counts, so we add those on:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">ref</span><span class="op">)</span></span></code></pre>
</div>
<p>Some cleaning - remove cells of the reference dataset for which the
cell type annotation is missing:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nna</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">ref</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">ref</span><span class="op">[</span>,<span class="va">nna</span><span class="op">]</span></span></code></pre>
</div>
<p>Also remove cell types of very low abundance (here less than 10
cells) to remove noise prior to subsequent annotation tasks.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abu.ct</span> <span class="op">&lt;-</span> <span class="fu">names</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span><span class="op">[</span><span class="va">tab</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="va">ref</span><span class="op">$</span><span class="va">celltype</span> <span class="op">%in%</span> <span class="va">abu.ct</span></span>
<span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">ref</span><span class="op">[</span>,<span class="va">ind</span><span class="op">]</span> </span></code></pre>
</div>
<p>Restrict to genes shared between query and reference dataset.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">ref</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">ref</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span></span>
<span><span class="va">shared_genes</span> <span class="op">&lt;-</span> <span class="fu">intersect</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">ref</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="va">shared_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">ref</span><span class="op">[</span><span class="va">shared_genes</span>,<span class="op">]</span></span></code></pre>
</div>
<p>Convert sparse assay matrices to regular dense matrices for input to
SingleR:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.mat</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ref.mat</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">ref</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Finally, run SingleR with the query and reference datasets:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">SingleR</span><span class="op">(</span>test <span class="op">=</span> <span class="va">sce.mat</span>, </span>
<span>               ref <span class="op">=</span> <span class="va">ref.mat</span>,</span>
<span>               labels <span class="op">=</span> <span class="va">ref</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="va">res</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 1000 rows and 4 columns
                                   scores                 labels delta.next
                                 &lt;matrix&gt;            &lt;character&gt;  &lt;numeric&gt;
cell_11995 0.348586:0.335451:0.314515:... Forebrain/Midbrain/H..  0.1285110
cell_10294 0.273570:0.260013:0.298932:...             Erythroid3  0.1381951
cell_9963  0.328538:0.291288:0.475611:...            Endothelium  0.2193295
cell_11610 0.281161:0.269245:0.299961:...             Erythroid3  0.0359215
cell_10910 0.422454:0.346897:0.355947:...           ExE mesoderm  0.0984285
...                                   ...                    ...        ...
cell_11597 0.323805:0.292967:0.300485:...                    NMP  0.1663369
cell_9807  0.464466:0.374189:0.381698:...             Mesenchyme  0.0833019
cell_10095 0.341721:0.288215:0.485324:...            Endothelium  0.0889931
cell_11706 0.267487:0.240215:0.286012:...             Erythroid2  0.0350557
cell_11860 0.345786:0.343437:0.313994:... Forebrain/Midbrain/H..  0.0117001
                    pruned.labels
                      &lt;character&gt;
cell_11995 Forebrain/Midbrain/H..
cell_10294             Erythroid3
cell_9963             Endothelium
cell_11610             Erythroid3
cell_10910           ExE mesoderm
...                           ...
cell_11597                    NMP
cell_9807              Mesenchyme
cell_10095            Endothelium
cell_11706             Erythroid2
cell_11860 Forebrain/Midbrain/H..</code></pre>
</div>
<p>We inspect the results using a heatmap of the per-cell and label
scores. Ideally, each cell should exhibit a high score in one label
relative to all of the others, indicating that the assignment to that
label was unambiguous.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotScoreHeatmap</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-score-heat-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We obtained fairly unambiguous predictions for mesenchyme and
endothelial cells, whereas we see expectedly more ambiguity between the
two erythroid cell populations.</p>
<p>We can also compare the cell type assignments with the unsupervised
clustering results to determine the identity of each cluster. Here,
several cell type classes are nested within the same cluster, indicating
that these clusters are composed of several transcriptomically similar
cell populations. On the other hand, there are also instances where we
have several clusters for the same cell type, indicating that the
clustering represents finer subdivisions within these cell types.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu">table</span><span class="op">(</span>anno <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">pruned.labels</span>, cluster <span class="op">=</span> <span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu">log2</span><span class="op">(</span><span class="va">tab</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">101</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>As it so happens, we are in the fortunate position where our test
dataset also contains independently defined labels. We see strong
consistency between the two sets of labels, indicating that our
automatic annotation is comparable to that generated manually by domain
experts.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu">table</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pruned.labels</span>, <span class="va">sce</span><span class="op">$</span><span class="va">celltype.mapped</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="fu">log2</span><span class="op">(</span><span class="va">tab</span> <span class="op">+</span> <span class="fl">10</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">101</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-anno-vs-preanno-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Assign the SingleR annotations as a column in the colData for the
query object <code>sce</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">SingleR_label</span> <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">pruned.labels</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="assigning-cell-labels-from-gene-sets">Assigning cell labels from gene sets<a class="anchor" aria-label="anchor" href="#assigning-cell-labels-from-gene-sets"></a>
</h3>
<p>A related strategy is to explicitly identify sets of marker genes
that are highly expressed in each individual cell. This does not require
matching of individual cells to the expression values of the reference
dataset, which is faster and more convenient when only the identities of
the markers are available. We demonstrate this approach using cell type
markers derived from the mouse embryo atlas dataset.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wilcox.z</span> <span class="op">&lt;-</span> <span class="fu">pairwiseWilcox</span><span class="op">(</span><span class="va">ref</span>, <span class="va">ref</span><span class="op">$</span><span class="va">celltype</span>, lfc <span class="op">=</span> <span class="fl">1</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers.z</span> <span class="op">&lt;-</span> <span class="fu">getTopMarkers</span><span class="op">(</span><span class="va">wilcox.z</span><span class="op">$</span><span class="va">statistics</span>, <span class="va">wilcox.z</span><span class="op">$</span><span class="va">pairs</span>, </span>
<span>                           pairwise <span class="op">=</span> <span class="cn">FALSE</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">lengths</span><span class="op">(</span><span class="va">markers.z</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     Allantois                 Cardiomyocytes
                           106                            106
                   Endothelium                     Erythroid2
                           103                             54
                    Erythroid3                   ExE endoderm
                            84                            102
                  ExE mesoderm   Forebrain/Midbrain/Hindbrain
                            97                             97
                           Gut Haematoendothelial progenitors
                            90                             71
         Intermediate mesoderm                     Mesenchyme
                            70                            118
                  Neural crest                            NMP
                            66                             91
             Paraxial mesoderm            Pharyngeal mesoderm
                            88                             85
              Somitic mesoderm                    Spinal cord
                            86                             91
              Surface ectoderm
                            92 </code></pre>
</div>
<!---

This version with scoreMarkers() produces worse looking diagnostics, so let's leave it with the pairwise Wilcox version.

``` r
ref_markers <- scoreMarkers(ref, groups = ref$celltype, lfc = 1)

get_top_markers <- function(marker_df, n = 100) {
  ord <- order(marker_df$mean.AUC, decreasing = TRUE)

  rownames(marker_df[ord,])[1:n]
}

markers.z <- lapply(ref_markers, get_top_markers)
```

-->
<p>Our test dataset will be as before the wild-type chimera dataset.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29411 1000
metadata(0):
assays(2): counts logcounts
rownames(29411): Xkr4 Gm1992 ... Vmn2r122 CAAA01147332.1
rowData names(2): ENSEMBL SYMBOL
colnames(1000): cell_11995 cell_10294 ... cell_11706 cell_11860
colData names(14): cell barcode ... clust2 SingleR_label
reducedDimNames(4): pca.corrected.E7.5 pca.corrected.E8.5 PCA UMAP
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>We use the <em><a href="https://bioconductor.org/packages/3.19/AUCell" class="external-link">AUCell</a></em>
package to identify marker sets that are highly expressed in each cell.
This method ranks genes by their expression values within each cell and
constructs a response curve of the number of genes from each marker set
that are present with increasing rank. It then computes the area under
the curve (AUC) for each marker set, quantifying the enrichment of those
markers among the most highly expressed genes in that cell. This is
roughly similar to performing a Wilcoxon rank sum test between genes in
and outside of the set, but involving only the top ranking genes by
expression in each cell.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.sets</span> <span class="op">&lt;-</span> <span class="fu">lapply</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">markers.z</span><span class="op">)</span>, </span>
<span>                   <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">GeneSet</span><span class="op">(</span><span class="va">markers.z</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, setName <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all.sets</span> <span class="op">&lt;-</span> <span class="fu">GeneSetCollection</span><span class="op">(</span><span class="va">all.sets</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all.sets</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GeneSetCollection
  names: Allantois, Cardiomyocytes, ..., Surface ectoderm (19 total)
  unique identifiers: Phlda2, Spin2c, ..., Akr7a5 (991 total)
  types in collection:
    geneIdType: NullIdentifier (1 total)
    collectionType: NullCollection (1 total)</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rankings</span> <span class="op">&lt;-</span> <span class="fu">AUCell_buildRankings</span><span class="op">(</span><span class="fu">as.matrix</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                 plotStats <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cell.aucs</span> <span class="op">&lt;-</span> <span class="fu">AUCell_calcAUC</span><span class="op">(</span><span class="va">all.sets</span>, <span class="va">rankings</span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cell.aucs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            gene sets
cells        Allantois Cardiomyocytes Endothelium Erythroid2 Erythroid3
  cell_11995    0.0691         0.0536      0.0459     0.1056     0.0948
  cell_10294    0.0384         0.0370      0.0451     0.4910     0.5153
  cell_9963     0.2177         0.1011      0.4380     0.1070     0.1087
  cell_11610    0.0108         0.0428      0.0347     0.4687     0.4555
  cell_10910    0.1868         0.0766      0.0869     0.0766     0.0682
  cell_11021    0.0695         0.0596      0.0465     0.0991     0.0993
            gene sets
cells        ExE endoderm ExE mesoderm Forebrain/Midbrain/Hindbrain    Gut
  cell_11995       0.0342       0.1324                       0.3173 0.1011
  cell_10294       0.0680       0.0172                       0.0439 0.0458
  cell_9963        0.0686       0.1147                       0.1116 0.1237
  cell_11610       0.0573       0.0202                       0.0460 0.0268
  cell_10910       0.0764       0.3255                       0.1747 0.1816
  cell_11021       0.0680       0.2029                       0.2500 0.1277
            gene sets
cells        Haematoendothelial progenitors Intermediate mesoderm Mesenchyme
  cell_11995                         0.0396                0.1627     0.0869
  cell_10294                         0.0371                0.0387     0.0369
  cell_9963                          0.3906                0.1157     0.2338
  cell_11610                         0.0258                0.0444     0.0324
  cell_10910                         0.1477                0.2265     0.2042
  cell_11021                         0.0473                0.2016     0.0814
            gene sets
cells        Neural crest    NMP Paraxial mesoderm Pharyngeal mesoderm
  cell_11995        0.208 0.1805            0.1889              0.1844
  cell_10294        0.109 0.0445            0.0226              0.0263
  cell_9963         0.145 0.1045            0.1706              0.1561
  cell_11610        0.107 0.0640            0.0195              0.0294
  cell_10910        0.135 0.2025            0.1437              0.1686
  cell_11021        0.168 0.3432            0.1255              0.1515
            gene sets
cells        Somitic mesoderm Spinal cord Surface ectoderm
  cell_11995           0.1279      0.2592           0.1611
  cell_10294           0.0203      0.0633           0.0468
  cell_9963            0.1197      0.1014           0.0850
  cell_11610           0.0424      0.0692           0.0332
  cell_10910           0.1787      0.1521           0.1043
  cell_11021           0.2259      0.1974           0.1509</code></pre>
</div>
<p>We assign cell type identity to each cell in the test dataset by
taking the marker set with the top AUC as the label for that cell. Our
new labels mostly agree with the original annotation (and, thus, also
with the reference-based annotation). Instances where the original
annotation is divided into several new label groups typically points to
large overlaps in their marker sets. In the absence of prior annotation,
a more general diagnostic check is to compare the assigned labels to
cluster identities, under the expectation that most cells of a single
cluster would have the same label (or, if multiple labels are present,
they should at least represent closely related cell states). We only
print out the top-left corner of the table here, but you should try
looking at the whole thing:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new.labels</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">[</span><span class="fu">max.col</span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu">table</span><span class="op">(</span><span class="va">new.labels</span>, <span class="va">sce</span><span class="op">$</span><span class="va">celltype.mapped</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
new.labels       Allantois Blood progenitors 1 Blood progenitors 2
  Allantois             44                   0                   0
  Cardiomyocytes         0                   0                   0
  Endothelium            0                   3                   0
  Erythroid2             0                   1                   7

new.labels       Cardiomyocytes
  Allantois                   0
  Cardiomyocytes             32
  Endothelium                 0
  Erythroid2                  0</code></pre>
</div>
<p>As a diagnostic measure, we examine the distribution of AUCs across
cells for each label. In heterogeneous populations, the distribution for
each label should be bimodal with one high-scoring peak containing cells
of that cell type and a low-scoring peak containing cells of other
types. The gap between these two peaks can be used to derive a threshold
for whether a label is “active” for a particular cell. (In this case, we
simply take the single highest-scoring label per cell as the labels
should be mutually exclusive.) In populations where a particular cell
type is expected, lack of clear bimodality for the corresponding label
may indicate that its gene set is not sufficiently informative.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">AUCell_exploreThresholds</span><span class="op">(</span><span class="va">cell.aucs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, plotHist <span class="op">=</span> <span class="cn">TRUE</span>, assign <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-auc-dist-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Shown is the distribution of AUCs in the wild-type chimera dataset
for each label in the embryo atlas dataset. The blue curve represents
the density estimate, the red curve represents a fitted two-component
mixture of normals, the pink curve represents a fitted three-component
mixture, and the grey curve represents a fitted normal distribution.
Vertical lines represent threshold estimates corresponding to each
estimate of the distribution.</p>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Inspect the diagnostics for the next nine cell types. Do they look
okay?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">AUCell_exploreThresholds</span><span class="op">(</span><span class="va">cell.aucs</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">18</span><span class="op">]</span>, plotHist <span class="op">=</span> <span class="cn">TRUE</span>, assign <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-auc-dist2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<div id="exercise-1-clustering" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-clustering" class="callout-inner">
<h3 class="callout-title">Exercise 1: Clustering</h3>
<div class="callout-content">
<p>The <a href="https://www.nature.com/articles/s41598-019-41695-z" class="external-link">Leiden
algorithm</a> is similar to the Louvain algorithm, but it is faster and
has been shown to result in better connected communities. Modify the
above call to <code>clusterCells</code> to carry out the community
detection with the Leiden algorithm instead. Visualize the results in a
UMAP plot.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>The <code>NNGraphParam</code> constructor has an argument
<code>cluster.args</code>. This allows to specify arguments passed on to
the <code>cluster_leiden</code> function from the <a href="https://cran.r-project.org/web/packages/igraph/index.html" class="external-link">igraph</a>
package. Use the <code>cluster.args</code> argument to parameterize the
clustering to use modularity as the objective function and a resolution
parameter of 0.5.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arg_list</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>objective_function <span class="op">=</span> <span class="st">"modularity"</span>,</span>
<span>                 resolution_parameter <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">leiden_clust</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                               BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>, </span>
<span>                                                        cluster.args <span class="op">=</span> <span class="va">arg_list</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span>, color_by <span class="op">=</span> <span class="st">"leiden_clust"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="exercise-2-reference-marker-genes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-reference-marker-genes" class="callout-inner">
<h3 class="callout-title">Exercise 2: Reference marker genes</h3>
<div class="callout-content">
<p>Identify the marker genes in the reference single cell experiment,
using the <code>celltype</code> labels that come with the dataset as the
groups. Compare the top 100 marker genes of two cell types that are
close in UMAP space. Do they share similar marker sets?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">scoreMarkers</span><span class="op">(</span><span class="va">ref</span>, groups <span class="op">=</span> <span class="va">ref</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of length 19
names(19): Allantois Cardiomyocytes ... Spinal cord Surface ectoderm</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># It comes with UMAP precomputed too</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">ref</span>, dimred <span class="op">=</span> <span class="st">"umap"</span>, color_by <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/cell_type_annotation-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Repetitive work -&gt; write a function</span></span>
<span><span class="va">order_marker_df</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">m_df</span>, <span class="va">n</span> <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">m_df</span><span class="op">$</span><span class="va">mean.AUC</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu">rownames</span><span class="op">(</span><span class="va">m_df</span><span class="op">[</span><span class="va">ord</span>,<span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">order_marker_df</span><span class="op">(</span><span class="va">markers</span><span class="op">[[</span><span class="st">"Erythroid2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">order_marker_df</span><span class="op">(</span><span class="va">markers</span><span class="op">[[</span><span class="st">"Erythroid3"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fl">100</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.66</code></pre>
</div>
<p>Turns out there’s pretty substantial overlap between
<code>Erythroid2</code> and <code>Erythroid3</code>. It would also be
interesting to plot the expression of the set difference to confirm that
the remainder are the the genes used to distinguish these two types from
each other.</p>
</div>
</div>
</div>
</div>
<div id="extension-challenge-1-group-pair-comparisons" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-1-group-pair-comparisons" class="callout-inner">
<h3 class="callout-title">Extension Challenge 1: Group pair comparisons</h3>
<div class="callout-content">
<p>Why do you think marker genes are found by aggregating pairwise
comparisons rather than iteratively comparing each cluster to all other
clusters?</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<p>One important reason why is because averages over all other clusters
can be sensitive to the cell type composition. If a rare cell type shows
up in one sample, the most discriminative marker genes found in this way
could be very different from those found in another sample where the
rare cell type is absent.</p>
<p>Generally, it’s good to keep in mind that the concept of “everything
else” is not a stable basis for comparison. Read that sentence again,
because its a subtle but broadly applicable point. Think about it and
you can probably identify analogous issues in fields outside of
single-cell analysis. It frequently comes up when comparisons between
multiple categories are involved.</p>
</div>
</div>
</div>
</div>
<div id="extension-challenge-2-parallelizing-singler" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-2-parallelizing-singler" class="callout-inner">
<h3 class="callout-title">Extension Challenge 2: Parallelizing SingleR</h3>
<div class="callout-content">
<p>SingleR can be computationally expensive. How do you set it to run in
parallel?</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p>Use <code>BiocParallel</code> and the <code>BPPARAM</code> argument!
This example will set it to use four cores on your laptop, but you can
also configure BiocParallel to use cluster jobs.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BiocParallel</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_bpparam</span> <span class="op">&lt;-</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">SingleR</span><span class="op">(</span>test <span class="op">=</span> <span class="va">sce.mat</span>, </span>
<span>                ref <span class="op">=</span> <span class="va">ref.mat</span>,</span>
<span>                labels <span class="op">=</span> <span class="va">ref</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>                BPPARAM <span class="op">=</span> <span class="va">my_bpparam</span><span class="op">)</span></span></code></pre>
</div>
<p><code>BiocParallel</code> is the most common way to enable parallel
computation in Bioconductor packages, so you can expect to see it
elsewhere outside of SingleR.</p>
</div>
</div>
</div>
</div>
<div id="extension-challenge-3-critical-inspection-of-diagnostics" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-3-critical-inspection-of-diagnostics" class="callout-inner">
<h3 class="callout-title">Extension Challenge 3: Critical inspection of diagnostics</h3>
<div class="callout-content">
<p>The first set of AUCell diagnostics don’t look so good for some of
the examples here. Which ones? Why?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<p>The example that jumps out most strongly to the eye is ExE endoderm,
which doesn’t show clear separate modes. Simultaneously, Endothelium
seems to have three or four modes.</p>
<p>Remember, this is an exploratory diagnostic, not the final word! At
this point it’d be good to engage in some critical inspection of the
results. Maybe we don’t have enough / the best marker genes. In this
particular case, the fact that we subsetted the reference set to 1000
cells probably didn’t help.</p>
</div>
</div>
</div>
</div>
<div id="further-reading" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further Reading</h3>
<div class="callout-content">
<ul>
<li>OSCA book, <a href="https://bioconductor.org/books/release/OSCA.basic/clustering.html" class="external-link">Chapters
5-7</a>
</li>
<li>Assigning cell types with SingleR (<a href="https://bioconductor.org/books/release/SingleRBook/" class="external-link">the
book</a>).</li>
<li>The <a href="https://bioconductor.org/packages/AUCell" class="external-link">AUCell</a>
package vignette.</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The two main approaches for cell type annotation are 1) manual
annotation of clusters based on marker gene expression, and 2)
computational annotation based on annotation transfer from reference
datasets or marker gene set enrichment testing.</li>
<li>For manual annotation, cells are first clustered with unsupervised
methods such as graph-based clustering followed by community detection
algorithms such as Louvain or Leiden.</li>
<li>The <code>clusterCells</code> function from the <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em>
package provides different algorithms that are commonly used for the
clustering of scRNA-seq data.</li>
<li>Once clusters have been obtained, cell type labels are then manually
assigned to cell clusters by matching cluster-specific upregulated
marker genes with prior knowledge of cell-type markers.</li>
<li>The <code>scoreMarkers</code> function from the <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em>
package package can be used to find candidate marker genes for clusters
of cells by ranking differential expression between pairs of
clusters.</li>
<li>Computational annotation using published reference datasets or
curated gene sets provides a fast, automated, and reproducible
alternative to the manual annotation of cell clusters based on marker
gene expression.</li>
<li>The <em><a href="https://bioconductor.org/packages/3.19/SingleR" class="external-link">SingleR</a></em>
package is a popular choice for reference-based annotation and assigns
labels to cells based on the reference samples with the highest Spearman
rank correlations.</li>
<li>The <em><a href="https://bioconductor.org/packages/3.19/AUCell" class="external-link">AUCell</a></em>
package provides an enrichment test to identify curated marker sets that
are highly expressed in each cell.</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] GSEABase_1.66.0              graph_1.82.0
 [3] annotate_1.82.0              XML_3.99-0.18
 [5] AnnotationDbi_1.66.0         pheatmap_1.0.12
 [7] scran_1.32.0                 scater_1.32.1
 [9] ggplot2_3.5.1                scuttle_1.14.0
[11] bluster_1.14.0               SingleR_2.6.0
[13] MouseGastrulationData_1.18.0 SpatialExperiment_1.14.0
[15] SingleCellExperiment_1.26.0  SummarizedExperiment_1.34.0
[17] Biobase_2.64.0               GenomicRanges_1.56.2
[19] GenomeInfoDb_1.40.1          IRanges_2.38.1
[21] S4Vectors_0.42.1             BiocGenerics_0.50.0
[23] MatrixGenerics_1.16.0        matrixStats_1.5.0
[25] AUCell_1.26.0                BiocStyle_2.32.1

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3        jsonlite_1.8.9
  [3] magrittr_2.0.3            ggbeeswarm_0.7.2
  [5] magick_2.8.5              farver_2.1.2
  [7] rmarkdown_2.29            zlibbioc_1.50.0
  [9] vctrs_0.6.5               memoise_2.0.1
 [11] DelayedMatrixStats_1.26.0 htmltools_0.5.8.1
 [13] S4Arrays_1.4.1            AnnotationHub_3.12.0
 [15] curl_6.2.0                BiocNeighbors_1.22.0
 [17] SparseArray_1.4.8         htmlwidgets_1.6.4
 [19] plotly_4.10.4             cachem_1.1.0
 [21] igraph_2.1.4              mime_0.12
 [23] lifecycle_1.0.4           pkgconfig_2.0.3
 [25] rsvd_1.0.5                Matrix_1.7-2
 [27] R6_2.5.1                  fastmap_1.2.0
 [29] GenomeInfoDbData_1.2.12   digest_0.6.37
 [31] colorspace_2.1-1          dqrng_0.4.1
 [33] irlba_2.3.5.1             ExperimentHub_2.12.0
 [35] RSQLite_2.3.9             beachmat_2.20.0
 [37] labeling_0.4.3            filelock_1.0.3
 [39] httr_1.4.7                abind_1.4-8
 [41] compiler_4.4.2            bit64_4.6.0-1
 [43] withr_3.0.2               BiocParallel_1.38.0
 [45] viridis_0.6.5             DBI_1.2.3
 [47] R.utils_2.12.3            MASS_7.3-64
 [49] rappdirs_0.3.3            DelayedArray_0.30.1
 [51] rjson_0.2.23              tools_4.4.2
 [53] vipor_0.4.7               beeswarm_0.4.0
 [55] R.oo_1.27.0               glue_1.8.0
 [57] nlme_3.1-167              grid_4.4.2
 [59] cluster_2.1.8             generics_0.1.3
 [61] gtable_0.3.6              R.methodsS3_1.8.2
 [63] tidyr_1.3.1               data.table_1.16.4
 [65] BiocSingular_1.20.0       ScaledMatrix_1.12.0
 [67] metapod_1.12.0            XVector_0.44.0
 [69] ggrepel_0.9.6             BiocVersion_3.19.1
 [71] pillar_1.10.1             limma_3.60.6
 [73] BumpyMatrix_1.12.0        splines_4.4.2
 [75] dplyr_1.1.4               BiocFileCache_2.12.0
 [77] lattice_0.22-6            survival_3.8-3
 [79] renv_1.1.0                FNN_1.1.4.1
 [81] bit_4.5.0.1               tidyselect_1.2.1
 [83] locfit_1.5-9.11           Biostrings_2.72.1
 [85] knitr_1.49                gridExtra_2.3
 [87] edgeR_4.2.2               xfun_0.50
 [89] mixtools_2.0.0            statmod_1.5.0
 [91] UCSC.utils_1.0.0          lazyeval_0.2.2
 [93] yaml_2.3.10               evaluate_1.0.3
 [95] codetools_0.2-20          kernlab_0.9-33
 [97] tibble_3.2.1              BiocManager_1.30.25
 [99] cli_3.6.3                 uwot_0.2.2
[101] xtable_1.8-4              segmented_2.1-3
[103] munsell_0.5.1             Rcpp_1.0.14
[105] dbplyr_2.5.0              png_0.1-8
[107] parallel_4.4.2            blob_1.2.4
[109] sparseMatrixStats_1.16.0  viridisLite_0.4.2
[111] scales_1.3.0              purrr_1.0.2
[113] crayon_1.5.3              rlang_1.1.5
[115] formatR_1.14              cowplot_1.1.3
[117] KEGGREST_1.44.1          </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-multi-sample"><p>Content from <a href="multi-sample.html">Multi-sample analyses</a></p>
<hr>
<p>Last updated on 2025-02-07 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/multi-sample.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we integrate data from multiple batches, samples, and
studies?</li>
<li>How can we identify differentially expressed genes between
experimental conditions for each cell type?</li>
<li>How can we identify changes in cell type abundance between
experimental conditions?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Correct batch effects and diagnose potential problems such as
over-correction.</li>
<li>Perform differential expression comparisons between conditions based
on pseudo-bulk samples.</li>
<li>Perform differential abundance comparisons between conditions.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="setup-and-data-exploration">Setup and data exploration<a class="anchor" aria-label="anchor" href="#setup-and-data-exploration"></a>
</h2>
<hr class="half-width">
<p>As before, we will use the the wild-type data from the Tal1 chimera
experiment:</p>
<ul>
<li>Sample 5: E8.5 injected cells (tomato positive), pool 3</li>
<li>Sample 6: E8.5 host cells (tomato negative), pool 3</li>
<li>Sample 7: E8.5 injected cells (tomato positive), pool 4</li>
<li>Sample 8: E8.5 host cells (tomato negative), pool 4</li>
<li>Sample 9: E8.5 injected cells (tomato positive), pool 5</li>
<li>Sample 10: E8.5 host cells (tomato negative), pool 5</li>
</ul>
<p>Note that this is a paired design in which for each biological
replicate (pool 3, 4, and 5), we have both host and injected cells.</p>
<p>We start by loading the data and doing a quick exploratory analysis,
essentially applying the normalization and visualization techniques that
we have seen in the previous lectures to all samples. Note that this
time we’re selecting samples 5 to 10, not just 5 by itself. Also note
the <code>type = "processed"</code> argument: we are explicitly
selecting the version of the data that has already been QC
processed.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">edgeR</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 20935
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(20935): cell_9769 cell_9770 ... cell_30702 cell_30703
colData names(11): cell barcode ... doub.density sizeFactor
reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 20935 rows and 11 columns
                  cell          barcode    sample       stage    tomato
           &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;
cell_9769    cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE
cell_9770    cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE
cell_9771    cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE
cell_9772    cell_9772 AAACCTGCATACTCTT         5        E8.5      TRUE
cell_9773    cell_9773 AAACGGGTCAACACCA         5        E8.5      TRUE
...                ...              ...       ...         ...       ...
cell_30699  cell_30699 TTTGTCACAGCTCGCA        10        E8.5     FALSE
cell_30700  cell_30700 TTTGTCAGTCTAGTCA        10        E8.5     FALSE
cell_30701  cell_30701 TTTGTCATCATCGGAT        10        E8.5     FALSE
cell_30702  cell_30702 TTTGTCATCATTATCC        10        E8.5     FALSE
cell_30703  cell_30703 TTTGTCATCCCATTTA        10        E8.5     FALSE
                pool stage.mapped        celltype.mapped closest.cell
           &lt;integer&gt;  &lt;character&gt;            &lt;character&gt;  &lt;character&gt;
cell_9769          3        E8.25             Mesenchyme   cell_24159
cell_9770          3         E8.5            Endothelium   cell_96660
cell_9771          3         E8.5              Allantois  cell_134982
cell_9772          3         E8.5             Erythroid3  cell_133892
cell_9773          3        E8.25             Erythroid1   cell_76296
...              ...          ...                    ...          ...
cell_30699         5         E8.5             Erythroid3   cell_38810
cell_30700         5         E8.5       Surface ectoderm   cell_38588
cell_30701         5        E8.25 Forebrain/Midbrain/H..   cell_66082
cell_30702         5         E8.5             Erythroid3  cell_138114
cell_30703         5         E8.0                Doublet   cell_92644
           doub.density sizeFactor
              &lt;numeric&gt;  &lt;numeric&gt;
cell_9769    0.02985045    1.41243
cell_9770    0.00172753    1.22757
cell_9771    0.01338013    1.15439
cell_9772    0.00218402    1.28676
cell_9773    0.00211723    1.78719
...                 ...        ...
cell_30699   0.00146287   0.389311
cell_30700   0.00374155   0.588784
cell_30701   0.05651258   0.624455
cell_30702   0.00108837   0.550807
cell_30703   0.82369305   1.184919</code></pre>
</div>
<p>For the sake of making these examples run faster, we drop some
problematic types (stripped nuclei and doublets) and also randomly
select 50% cells per sample.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">drop</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"Doublet"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">drop</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">29482</span><span class="op">)</span></span>
<span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">tapply</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">perc</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">0.50</span> <span class="op">*</span> <span class="fu">length</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">sample</span><span class="op">(</span><span class="va">x</span>, <span class="va">perc</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">idx</span><span class="op">]</span></span></code></pre>
</div>
<p>We now normalize the data, run some dimensionality reduction steps,
and visualize them in a tSNE plot. In this case we happen to have a ton
of cell types to visualize, so we define a custom palette with a lot of
visually distinct colors (adapted from the <code>polychrome</code>
palette in the <a href="https://cran.r-project.org/web/packages/pals/vignettes/pals_examples.html" class="external-link"><code>pals</code>
package</a>).</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, block <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span></span>
<span><span class="va">chosen.hvgs</span> <span class="op">&lt;-</span> <span class="va">dec</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">chosen.hvgs</span>, ntop <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">color_vec</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#5A5156"</span>, <span class="st">"#E4E1E3"</span>, <span class="st">"#F6222E"</span>, <span class="st">"#FE00FA"</span>, <span class="st">"#16FF32"</span>, <span class="st">"#3283FE"</span>, </span>
<span>               <span class="st">"#FEAF16"</span>, <span class="st">"#B00068"</span>, <span class="st">"#1CFFCE"</span>, <span class="st">"#90AD1C"</span>, <span class="st">"#2ED9FF"</span>, <span class="st">"#DEA0FD"</span>, </span>
<span>               <span class="st">"#AA0DFE"</span>, <span class="st">"#F8A19F"</span>, <span class="st">"#325A9B"</span>, <span class="st">"#C4451C"</span>, <span class="st">"#1C8356"</span>, <span class="st">"#85660D"</span>, </span>
<span>               <span class="st">"#B10DA1"</span>, <span class="st">"#3B00FB"</span>, <span class="st">"#1CBE4F"</span>, <span class="st">"#FA0087"</span>, <span class="st">"#333333"</span>, <span class="st">"#F7E1A0"</span>, </span>
<span>               <span class="st">"#C075A6"</span>, <span class="st">"#782AB6"</span>, <span class="st">"#AAF400"</span>, <span class="st">"#BDCDFF"</span>, <span class="st">"#822E1C"</span>, <span class="st">"#B5EFB5"</span>, </span>
<span>               <span class="st">"#7ED7D1"</span>, <span class="st">"#1C7F93"</span>, <span class="st">"#D85FF7"</span>, <span class="st">"#683B79"</span>, <span class="st">"#66B0FF"</span>, <span class="st">"#FBE426"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_vec</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-2-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are evident sample effects. Depending on the analysis that you
want to perform you may want to remove or retain the sample effect. For
instance, if the goal is to identify cell types with a clustering
method, one may want to remove the sample effects with “batch effect”
correction methods.</p>
<p>For now, let’s assume that we want to remove this effect.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>It seems like samples 5 and 6 are separated off from the others in
gene expression space. Given the group of cells in each sample, why
might this make sense versus some other pair of samples? What is the
factor presumably leading to this difference?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Samples 5 and 6 were from the same “pool” of cells. Looking at the
documentation for the dataset under <code>?WTChimeraData</code> we see
that the pool variable is defined as: “Integer, embryo pool from which
cell derived; samples with same value are matched.” So samples 5 and 6
have an experimental factor in common which causes a shared, systematic
difference in gene expression profiles compared to the other samples.
That’s why you can see many of isolated blue/orange clusters on the
first TSNE plot. If you were developing single-cell library preparation
protocols you might want to preserve this effect to understand how
variation in pools leads to variation in expression, but for now, given
that we’re investigating other effects, we’ll want to remove this as
undesired technical variation.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="correcting-batch-effects">Correcting batch effects<a class="anchor" aria-label="anchor" href="#correcting-batch-effects"></a>
</h2>
<hr class="half-width">
<p>We “correct” the effect of samples with the
<code>correctExperiment</code> function in the <code>batchelor</code>
package and using the <code>sample</code> column as batch.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">10102</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu">correctExperiments</span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    batch <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>    subset.row <span class="op">=</span> <span class="va">chosen.hvgs</span>,</span>
<span>    PARAM <span class="op">=</span> <span class="fu">FastMnnParam</span><span class="op">(</span></span>
<span>        merge.order <span class="op">=</span> <span class="fu">list</span><span class="op">(</span></span>
<span>            <span class="fu">list</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="co"># WT (3 replicates)</span></span>
<span>            <span class="fu">list</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>  <span class="co"># td-Tomato (3 replicates)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">merged</span>, dimred <span class="op">=</span> <span class="st">"corrected"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">merged</span>, colour_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can also see that when coloring by cell type, the cell types are
now nicely confined to their own clusters for the most part:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">merged</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_vec</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Once we removed the sample batch effect, we can proceed with the
Differential Expression Analysis.</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>True or False: after batch correction, no batch-level information is
present in the corrected data.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>False. Batch-level data can be retained through confounding with
experimental factors or poor ability to distinguish experimental effects
from batch effects. Remember, the changes needed to correct the data are
empirically estimated, so they can carry along error.</p>
<p>While batch effect correction algorithms usually do a pretty good
job, it’s smart to do a sanity check for batch effects at the end of
your analysis. You always want to make sure that that effect you’re
resting your paper submission on isn’t driven by batch effects.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<hr class="half-width">
<p>In order to perform a differential expression analysis, we need to
identify groups of cells across samples/conditions (depending on the
experimental design and the final aim of the experiment).</p>
<p>As previously seen, we have two ways of grouping cells, cell
clustering and cell labeling. In our case we will focus on this second
aspect to group cells according to the already annotated cell types to
proceed with the computation of the pseudo-bulk samples.</p>
<div class="section level3">
<h3 id="pseudo-bulk-samples">Pseudo-bulk samples<a class="anchor" aria-label="anchor" href="#pseudo-bulk-samples"></a>
</h3>
<p>To compute differences between groups of cells, a possible way is to
compute pseudo-bulk samples, where we mediate the gene signal of all the
cells for each specific cell type. In this manner, we are then able to
detect differences between the same cell type across two different
conditions.</p>
<p>To compute pseudo-bulk samples, we use the
<code>aggregateAcrossCells</code> function in the <code>scuttle</code>
package, which takes as input not only a SingleCellExperiment, but also
the id to use for the identification of the group of cells. In our case,
we use as id not just the cell type, but also the sample, because we
want be able to discern between replicates and conditions during further
steps.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using 'label' and 'sample' as our two factors; each column of the output</span></span>
<span><span class="co"># corresponds to one unique combination of these two factors.</span></span>
<span></span>
<span><span class="va">summed</span> <span class="op">&lt;-</span> <span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="va">merged</span>, </span>
<span>    id <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"celltype.mapped"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">summed</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 13641 179
metadata(2): merge.info pca.info
assays(1): counts
rownames(13641): ENSMUSG00000051951 ENSMUSG00000025900 ...
  ENSMUSG00000096730 ENSMUSG00000095742
rowData names(3): rotation ENSEMBL SYMBOL
colnames: NULL
colData names(15): batch cell ... sample ncells
reducedDimNames(5): corrected pca.corrected.E7.5 pca.corrected.E8.5 PCA
  TSNE
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>The main advantage of using pseudo-bulk samples is the possibility to
use well-tested methods for differential analysis like
<code>edgeR</code> and <code>DESeq2</code>, we will focus on the former
for this analysis. <code>edgeR</code> and <code>DESeq2</code> both use
negative binomial models under the hood, but differ in their
normalization strategies and other implementation details.</p>
<p>First, let’s start with a specific cell type, for instance the
“Mesenchymal stem cells”, and look into differences between this cell
type across conditions. We put the counts table into a
<code>DGEList</code> container called <code>y</code>, along with the
corresponding metadata.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">current</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>, <span class="va">summed</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op">==</span> <span class="st">"Mesenchyme"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">DGEList</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">current</span><span class="op">)</span>, samples <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">current</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>An object of class "DGEList"
$counts
                   Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
ENSMUSG00000051951       2       0       0       0       1       0
ENSMUSG00000025900       0       0       0       0       0       0
ENSMUSG00000025902       4       0       2       0       3       6
ENSMUSG00000033845     765     130     508     213     781     305
ENSMUSG00000002459       2       0       1       0       0       0
13636 more rows ...

$samples
        group lib.size norm.factors batch cell barcode sample stage tomato pool
Sample1     1  2478901            1     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3
Sample2     1   548407            1     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3
Sample3     1  1260187            1     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4
Sample4     1   578699            1     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4
Sample5     1  2092329            1     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5
Sample6     1   904929            1    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5
        stage.mapped celltype.mapped closest.cell doub.density sizeFactor
Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
        celltype.mapped.1 sample.1 ncells
Sample1        Mesenchyme        5    151
Sample2        Mesenchyme        6     28
Sample3        Mesenchyme        7    127
Sample4        Mesenchyme        8     75
Sample5        Mesenchyme        9    239
Sample6        Mesenchyme       10    146</code></pre>
</div>
<p>A typical step is to discard low quality samples due to low sequenced
library size. We discard these samples because they can affect further
steps like normalization and/or DEGs analysis.</p>
<p>We can see that in our case we don’t have low quality samples and we
don’t need to filter out any of them.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discarded</span> <span class="op">&lt;-</span> <span class="va">current</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&lt;</span> <span class="fl">10</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span>,<span class="op">!</span><span class="va">discarded</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">discarded</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Mode   FALSE
logical       6 </code></pre>
</div>
<p>The same idea is typically applied to the genes, indeed we need to
discard low expressed genes to improve accuracy for the DEGs
modeling.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">y</span>, group <span class="op">=</span> <span class="va">current</span><span class="op">$</span><span class="va">tomato</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Mode   FALSE    TRUE
logical    9121    4520 </code></pre>
</div>
<p>We can now proceed to normalize the data. There are several
approaches for normalizing bulk, and hence pseudo-bulk data. Here, we
use the Trimmed Mean of M-values method, implemented in the
<code>edgeR</code> package within the <code>calcNormFactors</code>
function. Keep in mind that because we are going to normalize the
pseudo-bulk counts, we don’t need to normalize the data in “single cell
form”.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span><span class="op">$</span><span class="va">samples</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        group lib.size norm.factors batch cell barcode sample stage tomato pool
Sample1     1  2478901    1.0506857     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3
Sample2     1   548407    1.0399112     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3
Sample3     1  1260187    0.9700083     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4
Sample4     1   578699    0.9871129     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4
Sample5     1  2092329    0.9695559     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5
Sample6     1   904929    0.9858611    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5
        stage.mapped celltype.mapped closest.cell doub.density sizeFactor
Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
        celltype.mapped.1 sample.1 ncells
Sample1        Mesenchyme        5    151
Sample2        Mesenchyme        6     28
Sample3        Mesenchyme        7    127
Sample4        Mesenchyme        8     75
Sample5        Mesenchyme        9    239
Sample6        Mesenchyme       10    146</code></pre>
</div>
<p>To investigate the effect of our normalization, we use a
Mean-Difference (MD) plot for each sample in order to detect possible
normalization problems due to insufficient cells/reads/UMIs composing a
particular pseudo-bulk profile.</p>
<p>In our case, we verify that all these plots are centered in 0 (on
y-axis) and present a trumpet shape, as expected.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq_len</span><span class="op">(</span><span class="fu">ncol</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">plotMD</span><span class="op">(</span><span class="va">y</span>, column <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Furthermore, we want to check if the samples cluster together based
on their known factors (like the tomato injection in this case).</p>
<p>In this case, we’ll use the multidimensional scaling (MDS) plot.
Multidimensional scaling (which also goes by principal
<em>coordinate</em> analysis (PCoA)) is a dimensionality reduction
technique that’s conceptually similar to principal <em>component</em>
analysis (PCA).</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">limma</span><span class="fu">::</span><span class="fu">plotMDS</span><span class="op">(</span><span class="fu">cpm</span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>               col <span class="op">=</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">tomato</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then construct a design matrix by including both the pool and the
tomato as factors. This design indicates which samples belong to which
pool and condition, so we can use it in the next step of the
analysis.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span><span class="fu">factor</span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu">factor</span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">design</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        (Intercept) factor(pool)4 factor(pool)5 factor(tomato)TRUE
Sample1           1             0             0                  1
Sample2           1             0             0                  0
Sample3           1             1             0                  1
Sample4           1             1             0                  0
Sample5           1             0             1                  1
Sample6           1             0             1                  0
attr(,"assign")
[1] 0 1 1 2
attr(,"contrasts")
attr(,"contrasts")$`factor(pool)`
[1] "contr.treatment"

attr(,"contrasts")$`factor(tomato)`
[1] "contr.treatment"</code></pre>
</div>
<p>Now we can estimate the Negative Binomial (NB) overdispersion
parameter, to model the mean-variance trend.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">trended.dispersion</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max.
0.009325 0.016271 0.024233 0.021603 0.026868 0.027993 </code></pre>
</div>
<p>The BCV plot allows us to investigate the relation between the
Biological Coefficient of Variation and the Average log CPM for each
gene. Additionally, the Common and Trend BCV are shown in
<code>red</code> and <code>blue</code>.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotBCV</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then fit a Quasi-Likelihood (QL) negative binomial generalized
linear model for each gene. The <code>robust = TRUE</code> parameter
avoids distortions from highly variable clusters. The QL method includes
an additional dispersion parameter, useful to handle the uncertainty and
variability of the per-gene variance, which is not well estimated by the
NB dispersions, so the two dispersion types complement each other in the
final analysis.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">var.prior</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.2944  0.9342  1.0764  1.0030  1.1300  1.1923 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">df.prior</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.3042  8.7218  8.7218  8.6433  8.7218  8.7218 </code></pre>
</div>
<p>QL dispersion estimates for each gene as a function of abundance. Raw
estimates (black) are shrunk towards the trend (blue) to yield squeezed
estimates (red).</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotQLDisp</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then use an empirical Bayes quasi-likelihood F-test to test for
differential expression (due to tomato injection) per each gene at a
False Discovery Rate (FDR) of 5%. The low amount of DGEs highlights that
the tomato injection effect has a low influence on the mesenchyme
cells.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">glmQLFTest</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">decideTests</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       factor(tomato)TRUE
Down                    5
NotSig               4510
Up                      5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">topTags</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coefficient:  factor(tomato)TRUE
                        logFC   logCPM          F       PValue          FDR
ENSMUSG00000010760 -4.1526570 9.973704 1116.70486 3.464851e-12 1.566113e-08
ENSMUSG00000096768  1.9988567 8.844258  376.77267 1.070605e-09 2.419567e-06
ENSMUSG00000035299  1.7967382 6.904163  119.58114 3.781967e-07 5.698163e-04
ENSMUSG00000086503 -6.4743977 7.411257  225.86232 2.112730e-06 2.387385e-03
ENSMUSG00000101609  1.3775797 7.310009   80.03729 2.671320e-06 2.414873e-03
ENSMUSG00000019188 -1.0192069 7.545530   61.91544 8.939417e-06 6.734361e-03
ENSMUSG00000024423  0.9941377 7.391075   56.95653 1.312959e-05 8.477962e-03
ENSMUSG00000042607 -0.9516835 7.468203   45.46329 3.618639e-05 2.044531e-02
ENSMUSG00000036446 -0.8287663 9.401028   43.05154 4.740864e-05 2.317935e-02
ENSMUSG00000027520  1.5915670 6.952923   42.77910 5.128175e-05 2.317935e-02</code></pre>
</div>
<p>All the previous steps can be easily performed with the following
function for each cell type, thanks to the <code>pseudoBulkDGE</code>
function in the <code>scran</code> package.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summed.filt</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>,<span class="va">summed</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">de.results</span> <span class="op">&lt;-</span> <span class="fu">pseudoBulkDGE</span><span class="op">(</span></span>
<span>    <span class="va">summed.filt</span>, </span>
<span>    label <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">celltype.mapped</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span><span class="fu">factor</span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="va">tomato</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"tomatoTRUE"</span>,</span>
<span>    condition <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">tomato</span> </span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>The returned object is a list of <code>DataFrame</code>s each with
the results for a cell type. Each of these contains also the
intermediate results in <code>edgeR</code> format to perform any
intermediate plot or diagnostic.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cur.results</span> <span class="op">&lt;-</span> <span class="va">de.results</span><span class="op">[[</span><span class="st">"Allantois"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">cur.results</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">cur.results</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span>,<span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 13641 rows and 5 columns
                       logFC    logCPM         F      PValue         FDR
                   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
ENSMUSG00000037664 -7.993258  11.55290  3389.401 1.49770e-25 6.29183e-22
ENSMUSG00000010760 -2.574986  12.40592  1154.193 1.78995e-20 3.75980e-17
ENSMUSG00000086503 -7.015615   7.49749   728.110 2.59315e-18 3.63127e-15
ENSMUSG00000096768  1.828370   9.33239   304.252 2.37781e-14 2.49729e-11
ENSMUSG00000022464  0.970406  10.28302   119.624 2.36805e-10 1.98964e-07
...                      ...       ...       ...         ...         ...
ENSMUSG00000095247        NA        NA        NA          NA          NA
ENSMUSG00000096808        NA        NA        NA          NA          NA
ENSMUSG00000079808        NA        NA        NA          NA          NA
ENSMUSG00000096730        NA        NA        NA          NA          NA
ENSMUSG00000095742        NA        NA        NA          NA          NA</code></pre>
</div>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Clearly some of the results have low p-values. What about the effect
sizes? What does <code>logFC</code> stand for?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>“logFC” stands for log fold-change. <code>edgeR</code> uses a log2
convention. Rather than reporting e.g. a 5-fold increase, it’s better to
report a logFC of log2(5) = 2.32. Additive log scales are easier to work
with than multiplicative identity scales, once you get used to it.</p>
<p><code>ENSMUSG00000037664</code> seems to have an estimated logFC of
about -8. That’s a big difference if it’s real.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="differential-abundance">Differential Abundance<a class="anchor" aria-label="anchor" href="#differential-abundance"></a>
</h2>
<hr class="half-width">
<p>With DA we look for differences in cluster <em>abundance</em> across
conditions (the tomato injection in our case), rather than differences
in gene expression.</p>
<p>Our first steps are quantifying the number of cells per each cell
type and fitting a model to catch differences between the injected cells
and the background.</p>
<p>The process is very similar differential expression modeling, but
this time we start our analysis on the computed abundances and without
normalizing the data with TMM.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">table</span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">celltype.mapped</span>, <span class="va">merged</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">unclass</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">extra.info</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span><span class="fu">match</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span>, <span class="va">merged</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">y.ab</span> <span class="op">&lt;-</span> <span class="fu">DGEList</span><span class="op">(</span><span class="va">abundances</span>, samples <span class="op">=</span> <span class="va">extra.info</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span><span class="fu">factor</span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu">factor</span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>, <span class="va">y.ab</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y.ab</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">y.ab</span>, <span class="va">design</span>, trend <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit.ab</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">y.ab</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span>, abundance.trend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="background-on-compositional-effect">Background on compositional effect<a class="anchor" aria-label="anchor" href="#background-on-compositional-effect"></a>
</h3>
<p>As mentioned before, in DA we don’t normalize our data with
<code>calcNormFactors</code> function, because this approach considers
that most of the input features do not vary between conditions. This
cannot be applied to DA analysis because we have a small number of cell
populations that all can change due to the treatment. This means that
here we will normalize only for library depth, which in pseudo-bulk data
means by the total number of cells in each sample (cell type).</p>
<p>On the other hand, this can lead our data to be susceptible to
compositional effect. “Compositional” refers to the fact that the
cluster abundances in a sample are not independent of one another
because each cell type is effectively competing for space in the sample.
They behave like proportions in that they must sum to 1. If cell type A
abundance increases in a new condition, that means we’ll observe less of
everything else, even if everything else is unaffected by the new
condition.</p>
<p>Compositionality means that our conclusions can be biased by the
amount of cells present in each cell type. And this amount of cells can
be totally unbalanced between cell types. This is particularly
problematic for cell types that start at or end up near 0 or 100
percent.</p>
<p>For example, a specific cell type can be 40% of the total amount of
cells present in the experiment, while another just the 3%. The
differences in terms of abundance of these cell types are detected
between the different conditions, but our final interpretation could be
biased if we don’t consider this aspect.</p>
<p>We now look at different approaches for handling the compositional
effect.</p>
</div>
<div class="section level3">
<h3 id="assuming-most-labels-do-not-change">Assuming most labels do not change<a class="anchor" aria-label="anchor" href="#assuming-most-labels-do-not-change"></a>
</h3>
<p>We can use a similar approach used during the DEGs analysis, assuming
that most labels are not changing, in particular if we think about the
low number of DEGs resulted from the previous analysis.</p>
<p>To do so, we first normalize the data with
<code>calcNormFactors</code> and then we fit and estimate a QL-model for
our abundance data.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y.ab2</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">y.ab</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y.ab2</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">norm.factors</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.1029040 1.0228173 1.0695358 0.7686501 1.0402941 1.0365354</code></pre>
</div>
<p>We then use edgeR in a manner similar to what we ran before:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y.ab2</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">y.ab2</span>, <span class="va">design</span>, trend <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit.ab2</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">y.ab2</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span>, abundance.trend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">glmQLFTest</span><span class="op">(</span><span class="va">fit.ab2</span>, coef <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">decideTests</span><span class="op">(</span><span class="va">res2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       factor(tomato)TRUE
Down                    2
NotSig                 32
Up                      0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">topTags</span><span class="op">(</span><span class="va">res2</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coefficient:  factor(tomato)TRUE
                       logFC   logCPM         F       PValue          FDR
ExE ectoderm      -5.7462892 13.13490 40.650293 1.853481e-08 6.301836e-07
Parietal endoderm -6.9020425 12.36649 28.120078 1.338738e-06 2.275854e-05
Mesenchyme         0.9656630 16.32654  6.729443 1.160472e-02 1.111437e-01
Erythroid3        -0.9191728 17.34677  6.495750 1.307572e-02 1.111437e-01
Neural crest      -1.0201642 14.83912  5.769547 1.904354e-02 1.294961e-01
ExE endoderm      -3.9992208 10.75172  5.058004 2.775531e-02 1.572801e-01
Endothelium        0.8672530 14.12195  3.614431 6.151880e-02 2.988056e-01
Cardiomyocytes     0.6955522 14.93321  2.833474 9.690453e-02 4.118442e-01
Allantois          0.5992522 15.54924  2.277082 1.359313e-01 5.135182e-01
Erythroid2        -0.5185684 15.97357  1.773710 1.873705e-01 6.370598e-01</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="testing-against-a-log-fold-change-threshold">Testing against a log-fold change threshold<a class="anchor" aria-label="anchor" href="#testing-against-a-log-fold-change-threshold"></a>
</h3>
<p>A second approach assumes that the composition bias introduces a
spurious log2-fold change of no more than a quantity for a non-DA
label.</p>
<p>In other words, we interpret this as the maximum log-fold change in
the total number of cells given by DA in other labels. On the other
hand, when choosing , we should not consider fold-differences in the
totals due to differences in capture efficiency or the size of the
original cell population are not attributable to composition bias. We
then mitigate the effect of composition biases by testing each label for
changes in abundance beyond .</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.lfc</span> <span class="op">&lt;-</span> <span class="fu">glmTreat</span><span class="op">(</span><span class="va">fit.ab</span>, coef <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">design</span><span class="op">)</span>, lfc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">decideTests</span><span class="op">(</span><span class="va">res.lfc</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       factor(tomato)TRUE
Down                    2
NotSig                 32
Up                      0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">topTags</span><span class="op">(</span><span class="va">res.lfc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coefficient:  factor(tomato)TRUE
                         logFC unshrunk.logFC   logCPM       PValue
ExE ectoderm        -5.5199847     -5.9467998 13.06465 7.002934e-06
Parietal endoderm   -6.5913805    -27.4174596 12.30091 1.486452e-04
ExE endoderm        -3.9308146    -23.9393101 10.76159 8.050176e-02
Mesenchyme           1.1622187      1.1634540 16.35239 1.306006e-01
Endothelium          1.0599946      1.0656145 14.14422 2.114555e-01
Caudal neurectoderm -1.4559768     -1.6059830 11.09613 3.351750e-01
Cardiomyocytes       0.8474499      0.8498787 14.96579 3.707748e-01
Neural crest        -0.8411426     -0.8436838 14.83184 3.730129e-01
Def. endoderm        0.7372230      0.7504500 12.50001 4.213046e-01
Allantois            0.7569205      0.7581982 15.54528 4.713874e-01
                             FDR
ExE ectoderm        0.0002380998
Parietal endoderm   0.0025269677
ExE endoderm        0.9123532342
Mesenchyme          0.9859547701
Endothelium         0.9859547701
Caudal neurectoderm 0.9859547701
Cardiomyocytes      0.9859547701
Neural crest        0.9859547701
Def. endoderm       0.9859547701
Allantois           0.9859547701</code></pre>
</div>
<p>Addionally, the choice of can be guided by other external
experimental data, like a previous or a pilot experiment.</p>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<div id="exercise-1-heatmaps" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-heatmaps" class="callout-inner">
<h3 class="callout-title">Exercise 1: Heatmaps</h3>
<div class="callout-content">
<p>Use the <code>pheatmap</code> package to create a heatmap of the
abundances table. Does it comport with the model results?</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>You can simply hand <code>pheatmap()</code> a matrix as its only
argument. <code>pheatmap()</code> has a million options you can adjust,
but the defaults are usually pretty good. Try to overlay sample-level
information with the <code>annotation_col</code> argument for an extra
challenge.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">y.ab</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-24-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno_df</span> <span class="op">&lt;-</span> <span class="va">y.ab</span><span class="op">$</span><span class="va">samples</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"tomato"</span>, <span class="st">"pool"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">anno_df</span><span class="op">$</span><span class="va">pool</span> <span class="op">=</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">anno_df</span><span class="op">$</span><span class="va">pool</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_df</span><span class="op">$</span><span class="va">tomato</span> <span class="op">&lt;-</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">anno_df</span><span class="op">$</span><span class="va">tomato</span>,</span>
<span>                         <span class="st">"tomato+"</span>,</span>
<span>                         <span class="st">"tomato-"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">y.ab</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>         annotation_col <span class="op">=</span> <span class="va">anno_df</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-24-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The top DA result was a decrease in ExE ectoderm in the tomato
condition, which you can sort of see, especially if you
<code>log1p()</code> the counts or discard rows that show much higher
values. ExE ectoderm counts were much higher in samples 8 and 10
compared to 5, 7, and 9.</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-model-specification-and-comparison" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-model-specification-and-comparison" class="callout-inner">
<h3 class="callout-title">Exercise 2: Model specification and comparison</h3>
<div class="callout-content">
<p>Try re-running the pseudobulk DGE without the <code>pool</code>
factor in the design specification. Compare the logFC estimates and the
distribution of p-values for the <code>Erythroid3</code> cell type.</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<p>After running the second pseudobulk DGE, you can join the two
<code>DataFrame</code>s of <code>Erythroid3</code> statistics using the
<code>merge()</code> function. You will need to create a common key
column from the gene IDs.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de.results2</span> <span class="op">&lt;-</span> <span class="fu">pseudoBulkDGE</span><span class="op">(</span></span>
<span>    <span class="va">summed.filt</span>, </span>
<span>    label <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">celltype.mapped</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span><span class="va">tomato</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"tomatoTRUE"</span>,</span>
<span>    condition <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">tomato</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">eryth1</span> <span class="op">&lt;-</span> <span class="va">de.results</span><span class="op">$</span><span class="va">Erythroid3</span></span>
<span></span>
<span><span class="va">eryth2</span> <span class="op">&lt;-</span> <span class="va">de.results2</span><span class="op">$</span><span class="va">Erythroid3</span></span>
<span></span>
<span><span class="va">eryth1</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">eryth1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">eryth2</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">eryth2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">comp_df</span> <span class="op">&lt;-</span> <span class="fu">merge</span><span class="op">(</span><span class="va">eryth1</span>, <span class="va">eryth2</span>, by <span class="op">=</span> <span class="st">'gene'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">comp_df</span> <span class="op">&lt;-</span> <span class="va">comp_df</span><span class="op">[</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">comp_df</span><span class="op">$</span><span class="va">logFC.x</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">comp_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">logFC.x</span>, <span class="va">logFC.y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_abline</span><span class="op">(</span>lty <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-25-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reshape to long format for ggplot facets. This is 1000x times easier to do</span></span>
<span><span class="co"># with tidyverse packages:</span></span>
<span><span class="va">pval_df</span> <span class="op">&lt;-</span> <span class="fu">reshape</span><span class="op">(</span><span class="va">comp_df</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"PValue.x"</span>, <span class="st">"PValue.y"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                   direction <span class="op">=</span> <span class="st">"long"</span>, </span>
<span>                   v.names <span class="op">=</span> <span class="st">"Pvalue"</span>,</span>
<span>                   timevar <span class="op">=</span> <span class="st">"pool_factor"</span>,</span>
<span>                   times <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"with pool factor"</span>, <span class="st">"no pool factor"</span><span class="op">)</span>,</span>
<span>                   varying <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"PValue.x"</span>, <span class="st">"PValue.y"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pval_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">Pvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_histogram</span><span class="op">(</span>boundary <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                   bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">facet_wrap</span><span class="op">(</span><span class="st">"pool_factor"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/multi-sample-rendered-unnamed-chunk-25-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can see that in this case, the logFC estimates are strongly
consistent between the two models, which tells us that the inclusion of
the <code>pool</code> factor in the model doesn’t strongly influence the
estimate of the <code>tomato</code> coefficients in this case.</p>
<p>The p-value histograms both look alright here, with a largely flat
plateau over most of the 0 - 1 range and a spike near 0. This is
consistent with the hypothesis that most genes are unaffected by
<code>tomato</code> but there are a small handful that clearly are.</p>
<p>If there were large shifts in the logFC estimates or p-value
distributions, that’s a sign that the design specification change has a
large impact on how the model sees the data. If that happens, you’ll
need to think carefully and critically about what variables should and
should not be included in the model formula.</p>
</div>
</div>
</div>
</div>
<div id="extension-challenge-1-group-effects" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-1-group-effects" class="callout-inner">
<h3 class="callout-title">Extension challenge 1: Group effects</h3>
<div class="callout-content">
<p>Having multiple independent samples in each experimental group is
always helpful, but it’s particularly important when it comes to batch
effect correction. Why?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<p>It’s important to have multiple samples within each experimental
group because it helps the batch effect correction algorithm distinguish
differences due to batch effects (uninteresting) from differences due to
group/treatment/biology (interesting).</p>
<p>Imagine you had one sample that received a drug treatment and one
that did not, each with 10,000 cells. They differ substantially in
expression of gene X. Is that an important scientific finding? You can’t
tell for sure, because the effect of drug is indistinguishable from a
sample-wise batch effect. But if the difference in gene X holds up when
you have five treated samples and five untreated samples, now you can be
a bit more confident. Many batch effect correction methods will take
information on experimental factors as additional arguments, which they
can use to help remove batch effects while retaining experimental
differences.</p>
</div>
</div>
</div>
</div>
<div id="further-reading" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further Reading</h3>
<div class="callout-content">
<ul>
<li>OSCA book, Multi-sample analysis, <a href="https://bioconductor.org/books/release/OSCA.multisample" class="external-link">Chapters
1, 4, and 6</a>
</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Batch effects are systematic technical differences in the observed
expression in cells measured in different experimental batches.</li>
<li>Computational removal of batch-to-batch variation with the
<code>correctExperiment</code> function from the <em><a href="https://bioconductor.org/packages/3.19/batchelor" class="external-link">batchelor</a></em>
package allows us to combine data across multiple batches for a
consolidated downstream analysis.</li>
<li>Differential expression (DE) analysis of replicated multi-condition
scRNA-seq experiments is typically based on pseudo-bulk expression
profiles, generated by summing counts for all cells with the same
combination of label and sample.</li>
<li>The <code>aggregateAcrossCells</code> function from the <em><a href="https://bioconductor.org/packages/3.19/scater" class="external-link">scater</a></em>
package facilitates the creation of pseudo-bulk samples.<br>
</li>
<li>The <code>pseudoBulkDGE</code> function from the <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em>
package can be used to detect significant changes in expression between
conditions for pseudo-bulk samples consisting of cells of the same
type.</li>
<li>Differential abundance (DA) analysis aims at identifying significant
changes in cell type abundance across conditions.</li>
<li>DA analysis uses bulk DE methods such as <em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em> and
<em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>,
which provide suitable statistical models for count data in the presence
of limited replication - except that the counts are not of reads per
gene, but of cells per label.</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] pheatmap_1.0.12              scran_1.32.0
 [3] scater_1.32.1                ggplot2_3.5.1
 [5] scuttle_1.14.0               edgeR_4.2.2
 [7] limma_3.60.6                 batchelor_1.20.0
 [9] MouseGastrulationData_1.18.0 SpatialExperiment_1.14.0
[11] SingleCellExperiment_1.26.0  SummarizedExperiment_1.34.0
[13] Biobase_2.64.0               GenomicRanges_1.56.2
[15] GenomeInfoDb_1.40.1          IRanges_2.38.1
[17] S4Vectors_0.42.1             BiocGenerics_0.50.0
[19] MatrixGenerics_1.16.0        matrixStats_1.5.0
[21] BiocStyle_2.32.1

loaded via a namespace (and not attached):
  [1] DBI_1.2.3                 formatR_1.14
  [3] gridExtra_2.3             rlang_1.1.5
  [5] magrittr_2.0.3            compiler_4.4.2
  [7] RSQLite_2.3.9             DelayedMatrixStats_1.26.0
  [9] png_0.1-8                 vctrs_0.6.5
 [11] pkgconfig_2.0.3           crayon_1.5.3
 [13] fastmap_1.2.0             dbplyr_2.5.0
 [15] magick_2.8.5              XVector_0.44.0
 [17] labeling_0.4.3            rmarkdown_2.29
 [19] ggbeeswarm_0.7.2          UCSC.utils_1.0.0
 [21] purrr_1.0.2               bit_4.5.0.1
 [23] bluster_1.14.0            xfun_0.50
 [25] zlibbioc_1.50.0           cachem_1.1.0
 [27] beachmat_2.20.0           jsonlite_1.8.9
 [29] blob_1.2.4                DelayedArray_0.30.1
 [31] BiocParallel_1.38.0       cluster_2.1.8
 [33] irlba_2.3.5.1             parallel_4.4.2
 [35] R6_2.5.1                  RColorBrewer_1.1-3
 [37] Rcpp_1.0.14               knitr_1.49
 [39] splines_4.4.2             Matrix_1.7-2
 [41] igraph_2.1.4              tidyselect_1.2.1
 [43] viridis_0.6.5             abind_1.4-8
 [45] yaml_2.3.10               codetools_0.2-20
 [47] curl_6.2.0                lattice_0.22-6
 [49] tibble_3.2.1              withr_3.0.2
 [51] KEGGREST_1.44.1           BumpyMatrix_1.12.0
 [53] Rtsne_0.17                evaluate_1.0.3
 [55] BiocFileCache_2.12.0      ExperimentHub_2.12.0
 [57] Biostrings_2.72.1         pillar_1.10.1
 [59] BiocManager_1.30.25       filelock_1.0.3
 [61] renv_1.1.0                generics_0.1.3
 [63] BiocVersion_3.19.1        sparseMatrixStats_1.16.0
 [65] munsell_0.5.1             scales_1.3.0
 [67] glue_1.8.0                metapod_1.12.0
 [69] tools_4.4.2               AnnotationHub_3.12.0
 [71] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0
 [73] locfit_1.5-9.11           cowplot_1.1.3
 [75] grid_4.4.2                AnnotationDbi_1.66.0
 [77] colorspace_2.1-1          GenomeInfoDbData_1.2.12
 [79] beeswarm_0.4.0            BiocSingular_1.20.0
 [81] vipor_0.4.7               cli_3.6.3
 [83] rsvd_1.0.5                rappdirs_0.3.3
 [85] viridisLite_0.4.2         S4Arrays_1.4.1
 [87] dplyr_1.1.4               ResidualMatrix_1.14.1
 [89] gtable_0.3.6              digest_0.6.37
 [91] dqrng_0.4.1               ggrepel_0.9.6
 [93] SparseArray_1.4.8         farver_2.1.2
 [95] rjson_0.2.23              memoise_2.0.1
 [97] htmltools_0.5.8.1         lifecycle_1.0.4
 [99] httr_1.4.7                mime_0.12
[101] statmod_1.5.0             bit64_4.6.0-1            </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-large_data"><p>Content from <a href="large_data.html">Working with large data</a></p>
<hr>
<p>Last updated on 2025-02-07 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/large_data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we work with single-cell datasets that are too large to fit
in memory?</li>
<li>How do we speed up single-cell analysis workflows for large
datasets?</li>
<li>How do we convert between popular single-cell data formats?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Work with out-of-memory data representations such as HDF5.</li>
<li>Speed up single-cell analysis with parallel computation.</li>
<li>Invoke fast approximations for essential analysis steps.</li>
<li>Convert <code>SingleCellExperiment</code> objects to
<code>SeuratObject</code>s and <code>AnnData</code> objects.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<hr class="half-width">
<p>Advances in scRNA-seq technologies have increased the number of cells
that can be assayed in routine experiments. Public databases such as <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO</a> are continually
expanding with more scRNA-seq studies, while large-scale projects such
as the <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>
are expected to generate data for billions of cells. For effective data
analysis, the computational methods need to scale with the increasing
size of scRNA-seq data sets. This section discusses how we can use
various aspects of the Bioconductor ecosystem to tune our analysis
pipelines for greater speed and efficiency.</p>
</section><section><h2 class="section-heading" id="out-of-memory-representations">Out of memory representations<a class="anchor" aria-label="anchor" href="#out-of-memory-representations"></a>
</h2>
<hr class="half-width">
<p>The count matrix is the central structure around which our analyses
are based. In most of the previous chapters, this has been held fully in
memory as a dense <code>matrix</code> or as a sparse
<code>dgCMatrix</code>. Howevever, in-memory representations may not be
feasible for very large data sets, especially on machines with limited
memory. For example, the 1.3 million brain cell data set from 10X
Genomics (<a href="https://doi.org/10.1038/ncomms14049" class="external-link">Zheng et al.,
2017</a>) would require over 100 GB of RAM to hold as a
<code>matrix</code> and around 30 GB as a <code>dgCMatrix</code>. This
makes it challenging to explore the data on anything less than a HPC
system.</p>
<p>The obvious solution is to use a file-backed matrix representation
where the data are held on disk and subsets are retrieved into memory as
requested. While a number of implementations of file-backed matrices are
available (e.g., <a href="https://cran.r-project.org/web/packages/bigmemory/index.html" class="external-link">bigmemory</a>,
<a href="https://bioconductor.org/packages/matter" class="external-link">matter</a>), we will
be using the implementation from the <a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a>
package. This uses the popular HDF5 format as the underlying data store,
which provides a measure of standardization and portability across
systems. We demonstrate with a subset of 20,000 cells from the 1.3
million brain cell data set, as provided by the <a href="https://bioconductor.org/packages/TENxBrainData" class="external-link">TENxBrainData</a>
package.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">TENxBrainData</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce.brain</span> <span class="op">&lt;-</span> <span class="fu">TENxBrainData20k</span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">sce.brain</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 27998 20000
metadata(0):
assays(1): counts
rownames: NULL
rowData names(2): Ensembl Symbol
colnames: NULL
colData names(4): Barcode Sequence Library Mouse
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>Examination of the <code>SingleCellExperiment</code> object indicates
that the count matrix is a <code>HDF5Matrix</code>. From a comparison of
the memory usage, it is clear that this matrix object is simply a stub
that points to the much larger HDF5 file that actually contains the
data. This avoids the need for large RAM availability during
analyses.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;27998 x 20000&gt; HDF5Matrix object of type "integer":
             [,1]     [,2]     [,3]     [,4] ... [,19997] [,19998] [,19999]
    [1,]        0        0        0        0   .        0        0        0
    [2,]        0        0        0        0   .        0        0        0
    [3,]        0        0        0        0   .        0        0        0
    [4,]        0        0        0        0   .        0        0        0
    [5,]        0        0        0        0   .        0        0        0
     ...        .        .        .        .   .        .        .        .
[27994,]        0        0        0        0   .        0        0        0
[27995,]        0        0        0        1   .        0        2        0
[27996,]        0        0        0        0   .        0        1        0
[27997,]        0        0        0        0   .        0        0        0
[27998,]        0        0        0        0   .        0        0        0
         [,20000]
    [1,]        0
    [2,]        0
    [3,]        0
    [4,]        0
    [5,]        0
     ...        .
[27994,]        0
[27995,]        0
[27996,]        0
[27997,]        0
[27998,]        0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">object.size</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>2496 bytes</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">file.info</span><span class="op">(</span><span class="fu">path</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">size</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 76264332</code></pre>
</div>
<p>Manipulation of the count matrix will generally result in the
creation of a <code>DelayedArray</code> object from the <a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a>
package. This remembers the operations to be applied to the counts and
stores them in the object, to be executed when the modified matrix
values are realized for use in calculations. The use of delayed
operations avoids the need to write the modified values to a new file at
every operation, which would unnecessarily require time-consuming disk
I/O.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">log2</span><span class="op">(</span><span class="va">tmp</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tmp</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;27998 x 20000&gt; DelayedMatrix object of type "double":
             [,1]     [,2]     [,3] ... [,19999] [,20000]
    [1,]        0        0        0   .        0        0
    [2,]        0        0        0   .        0        0
    [3,]        0        0        0   .        0        0
    [4,]        0        0        0   .        0        0
    [5,]        0        0        0   .        0        0
     ...        .        .        .   .        .        .
[27994,]        0        0        0   .        0        0
[27995,]        0        0        0   .        0        0
[27996,]        0        0        0   .        0        0
[27997,]        0        0        0   .        0        0
[27998,]        0        0        0   .        0        0</code></pre>
</div>
<p>Many functions described in the previous workflows are capable of
accepting <code>HDF5Matrix</code> objects. This is powered by the
availability of common methods for all matrix representations (e.g.,
subsetting, combining, methods from <a href="https://bioconductor.org/packages/DelayedMatrixStats" class="external-link">DelayedMatrixStats</a>
as well as representation-agnostic C++ code using <a href="https://bioconductor.org/packages/beachmat" class="external-link">beachmat</a>. For
example, we compute QC metrics below with the same
<code>calculateQCMetrics()</code> function that we used in the other
workflows.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"^mt-"</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu">perCellQCMetrics</span><span class="op">(</span><span class="va">sce.brain</span>, subsets <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>Mt <span class="op">=</span> <span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qcstats</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 20000 rows and 6 columns
            sum  detected subsets_Mt_sum subsets_Mt_detected subsets_Mt_percent
      &lt;numeric&gt; &lt;numeric&gt;      &lt;numeric&gt;           &lt;numeric&gt;          &lt;numeric&gt;
1          3060      1546            123                  10            4.01961
2          3500      1694            118                  11            3.37143
3          3092      1613             58                   9            1.87581
4          4420      2050            131                  10            2.96380
5          3771      1813            100                   8            2.65182
...         ...       ...            ...                 ...                ...
19996      4431      2050            127                   9           2.866170
19997      6988      2704             60                   9           0.858615
19998      8749      2988            305                  11           3.486113
19999      3842      1711            129                   8           3.357626
20000      1775       945             26                   6           1.464789
          total
      &lt;numeric&gt;
1          3060
2          3500
3          3092
4          4420
5          3771
...         ...
19996      4431
19997      6988
19998      8749
19999      3842
20000      1775</code></pre>
</div>
<p>Needless to say, data access from file-backed representations is
slower than that from in-memory representations. The time spent
retrieving data from disk is an unavoidable cost of reducing memory
usage. Whether this is tolerable depends on the application. One example
usage pattern involves performing the heavy computing quickly with
in-memory representations on HPC systems with plentiful memory, and then
distributing file-backed counterparts to individual users for
exploration and visualization on their personal machines.</p>
</section><section><h2 class="section-heading" id="parallelization">Parallelization<a class="anchor" aria-label="anchor" href="#parallelization"></a>
</h2>
<hr class="half-width">
<p>Parallelization of calculations across genes or cells is an obvious
strategy for speeding up scRNA-seq analysis workflows.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.19/BiocParallel" class="external-link">BiocParallel</a></em>
package provides a common interface for parallel computing throughout
the Bioconductor ecosystem, manifesting as a <code>BPPARAM</code>
argument in compatible functions. We can also use
<code>BiocParallel</code> with more expressive functions directly
through the package’s interface.</p>
<div class="section level4">
<h4 id="basic-use">Basic use<a class="anchor" aria-label="anchor" href="#basic-use"></a>
</h4>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BiocParallel</span><span class="op">)</span></span></code></pre>
</div>
<p><code>BiocParallel</code> makes it quite easy to iterate over a
vector and distribute the computation across workers using the
<code>bplapply</code> function. Basic knowledge of <code>lapply</code>
is required.</p>
<p>In this example, we find the square root of a vector of numbers in
parallel by indicating the <code>BPPARAM</code> argument in
<code>bplapply</code>.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">bplapply</span><span class="op">(</span></span>
<span>    X <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">4</span>, <span class="fl">9</span>, <span class="fl">16</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    FUN <span class="op">=</span> <span class="va">sqrt</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="va">param</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]
[1] 2

[[2]]
[1] 3

[[3]]
[1] 4

[[4]]
[1] 5</code></pre>
</div>
<p><strong>Note</strong>. The number of workers is set to 1 due to
continuous testing resource limitations.</p>
<p>There exists a diverse set of parallelization backends depending on
available hardware and operating systems.</p>
<p>For example, we might use forking across two cores to parallelize the
variance calculations on a Unix system:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec.mc</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec.mc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 29453 rows and 6 columns
                          mean       total        tech         bio     p.value
                     &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
ENSMUSG00000051951 0.002800256 0.003504940 0.002856697 6.48243e-04 1.20905e-01
ENSMUSG00000089699 0.000000000 0.000000000 0.000000000 0.00000e+00         NaN
ENSMUSG00000102343 0.000000000 0.000000000 0.000000000 0.00000e+00         NaN
ENSMUSG00000025900 0.000794995 0.000863633 0.000811019 5.26143e-05 3.68953e-01
ENSMUSG00000025902 0.170777718 0.388633677 0.170891603 2.17742e-01 2.47893e-11
...                        ...         ...         ...         ...         ...
ENSMUSG00000095041  0.35571083  0.34572194  0.33640994  0.00931199    0.443233
ENSMUSG00000063897  0.49007956  0.41924282  0.44078158 -0.02153876    0.599499
ENSMUSG00000096730  0.00000000  0.00000000  0.00000000  0.00000000         NaN
ENSMUSG00000095742  0.00177158  0.00211619  0.00180729  0.00030890    0.188992
tomato-td           0.57257331  0.47487832  0.49719425 -0.02231593    0.591542
                           FDR
                     &lt;numeric&gt;
ENSMUSG00000051951 6.76255e-01
ENSMUSG00000089699         NaN
ENSMUSG00000102343         NaN
ENSMUSG00000025900 7.56202e-01
ENSMUSG00000025902 1.35508e-09
...                        ...
ENSMUSG00000095041    0.756202
ENSMUSG00000063897    0.756202
ENSMUSG00000096730         NaN
ENSMUSG00000095742    0.756202
tomato-td             0.756202</code></pre>
</div>
<p>Another approach would be to distribute jobs across a network of
computers, which yields the same result:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec.snow</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, BPPARAM <span class="op">=</span> <span class="fu">SnowParam</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>For high-performance computing (HPC) systems with a cluster of
compute nodes, we can distribute jobs via the job scheduler using the
<code>BatchtoolsParam</code> class. The example below assumes a SLURM
cluster, though the settings can be easily configured for a particular
system (see <a href="https://bioconductor.org/packages/3.19/BiocParallel/vignettes/BiocParallel_BatchtoolsParam.pdf" class="external-link">here</a>
for details).</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 2 hours, 8 GB, 1 CPU per task, for 10 tasks.</span></span>
<span><span class="va">rs</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">7200</span>, memory <span class="op">=</span> <span class="fl">8000</span>, ncpus <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bpp</span> <span class="op">&lt;-</span> <span class="fu">BatchtoolsParam</span><span class="op">(</span><span class="fl">10</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, resources <span class="op">=</span> <span class="va">rs</span><span class="op">)</span></span></code></pre>
</div>
<p>Parallelization is best suited for independent, CPU-intensive tasks
where the division of labor results in a concomitant reduction in
compute time. It is not suited for tasks that are bounded by other
compute resources, e.g., memory or file I/O (though the latter is less
of an issue on HPC systems with parallel read/write). In particular, R
itself is inherently single-core, so many of the parallelization
backends involve (i) setting up one or more separate R sessions, (ii)
loading the relevant packages and (iii) transmitting the data to that
session. Depending on the nature and size of the task, this overhead may
outweigh any benefit from parallel computing. While the default behavior
of the parallel job managers often works well for simple cases, it is
sometimes necessary to explicitly specify what data/libraries are sent
to / loaded on the parallel workers in order to avoid unnecessary
overhead.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>How do you turn on progress bars with parallel processing?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>From <code>?MulticoreParam</code> :</p>
<blockquote>
<p><code>progressbar</code> logical(1) Enable progress bar (based on
plyr:::progress_text). Enabling the progress bar changes the default
value of tasks to .Machine$integer.max, so that progress is reported for
each element of X.</p>
</blockquote>
<p>Progress bars are a helpful way to gauge whether that task is going
to take 5 minutes or 5 hours.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="fast-approximations">Fast approximations<a class="anchor" aria-label="anchor" href="#fast-approximations"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="nearest-neighbor-searching">Nearest neighbor searching<a class="anchor" aria-label="anchor" href="#nearest-neighbor-searching"></a>
</h3>
<p>Identification of neighbouring cells in PC or expression space is a
common procedure that is used in many functions, e.g.,
<code>buildSNNGraph()</code>, <code>doubletCells()</code>. The default
is to favour accuracy over speed by using an exact nearest neighbour
(NN) search, implemented with the <span class="math inline">\(k\)</span>-means for <span class="math inline">\(k\)</span>-nearest neighbours algorithm. However,
for large data sets, it may be preferable to use a faster approximate
approach.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.19/BiocNeighbors" class="external-link">BiocNeighbors</a></em>
framework makes it easy to switch between search options by simply
changing the <code>BNPARAM</code> argument in compatible functions. To
demonstrate, we will use the wild-type chimera data for which we had
applied graph-based clustering using the Louvain algorithm for community
detection:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                               BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>The above clusters on a nearest neighbor graph generated with an
exact neighbour search. We repeat this below using an approximate
search, implemented using the <a href="https://github.com/spotify/Annoy" class="external-link">Annoy</a> algorithm. This
involves constructing a <code>AnnoyParam</code> object to specify the
search algorithm and then passing it to the parameterization of the
<code>NNGraphParam()</code> function. The results from the exact and
approximate searches are consistent with most clusters from the former
re-appearing in the latter. This suggests that the inaccuracy from the
approximation can be largely ignored.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">clusterCells</span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                         BLUSPARAM <span class="op">=</span> <span class="fu">NNGraphParam</span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>                                                  BNPARAM <span class="op">=</span> <span class="fu">AnnoyParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">table</span><span class="op">(</span>exact <span class="op">=</span> <span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, approx <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     approx
exact   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15
   1   90   0   0   0   0   0   0   0   1   0   0   0   0   0   0
   2    0 143   0   0   0   0   0   0   0   0   0   0   0   0   1
   3    0   0  76   0   0   0   0   0   0   0   0   0   0   0   0
   4    0   0   0 341   0   0   0   0   0   0   0   0   0   0  56
   5    0   0   1   0 394   1   0   0   0   1   0   0   0   0   0
   6    0   0   0   0   0 209   0   0   0   0   0   0   0   0   0
   7    0   0   0   0   0   0 245   0   0   1   0   0   0   0   0
   8    0   0   0   0   0   0   0  95   0   0   0   0   0   0   0
   9    1   0   0   0   1   0   0   0 106   0   0   0   0   0   0
   10   0   0   0   0   1   0   0   0   0 113   4  10   0   0   0
   11   0   0   0   0   0   1   0   0   0   1 141   0   0   0   0
   12   0   0   0   0   1   0   0   0   0   0   0 204   0   0   0
   13   0   0   0   0   0   0   0   0   0   0   6   0 146   0   0
   14   0   0   0   0   0   0   0   0   0   0   0   0   0  20   0</code></pre>
</div>
<p>The similarity of the two clusterings can be quantified by
calculating the pairwise Rand index:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rand</span> <span class="op">&lt;-</span> <span class="fu">pairwiseRand</span><span class="op">(</span><span class="fu">colLabels</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">clusters</span>, mode <span class="op">=</span> <span class="st">"index"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">stopifnot</span><span class="op">(</span><span class="va">rand</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre>
</div>
<p>Note that Annoy writes the NN index to disk prior to performing the
search. Thus, it may not actually be faster than the default exact
algorithm for small datasets, depending on whether the overhead of disk
write is offset by the computational complexity of the search. It is
also not difficult to find situations where the approximation
deteriorates, especially at high dimensions, though this may not have an
appreciable impact on the biological conclusions.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y1</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">50000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">50000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">y1</span>, <span class="va">y2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">exact</span> <span class="op">&lt;-</span> <span class="fu">findKNN</span><span class="op">(</span><span class="va">Y</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="va">approx</span> <span class="op">&lt;-</span> <span class="fu">findKNN</span><span class="op">(</span><span class="va">Y</span>, k <span class="op">=</span> <span class="fl">20</span>, BNPARAM <span class="op">=</span> <span class="fu">AnnoyParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">exact</span><span class="op">$</span><span class="va">index</span> <span class="op">!=</span> <span class="va">approx</span><span class="op">$</span><span class="va">index</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.561925</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="singular-value-decomposition">Singular value decomposition<a class="anchor" aria-label="anchor" href="#singular-value-decomposition"></a>
</h3>
<p>The singular value decomposition (SVD) underlies the PCA used
throughout our analyses, e.g., in <code>denoisePCA()</code>,
<code>fastMNN()</code>, <code>doubletCells()</code>. (Briefly, the right
singular vectors are the eigenvectors of the gene-gene covariance
matrix, where each eigenvector represents the axis of maximum remaining
variation in the PCA.) The default <code>base::svd()</code> function
performs an exact SVD that is not performant for large datasets.
Instead, we use fast approximate methods from the <em><a href="https://CRAN.R-project.org/package=irlba" class="external-link">irlba</a></em> and
<em><a href="https://CRAN.R-project.org/package=rsvd" class="external-link">rsvd</a></em>
packages, conveniently wrapped into the <em><a href="https://bioconductor.org/packages/3.19/BiocSingular" class="external-link">BiocSingular</a></em>
package for ease of use and package development. Specifically, we can
change the SVD algorithm used in any of these functions by simply
specifying an alternative value for the <code>BSPARAM</code>
argument.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">BiocSingular</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># As the name suggests, it is random, so we need to set the seed.</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">101000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">r.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu">RandomParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> num [1:2411, 1:20] 14.79 5.79 13.07 -32.19 -26.45 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...
  ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...
 - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...
 - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...
 - attr(*, "rotation")= num [1:500, 1:20] -0.174 -0.173 -0.157 0.105 -0.132 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...
  .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">101001</span><span class="op">)</span></span>
<span></span>
<span><span class="va">i.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu">IrlbaParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">i.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> num [1:2411, 1:20] -14.79 -5.79 -13.07 32.19 26.45 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...
  ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...
 - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...
 - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...
 - attr(*, "rotation")= num [1:500, 1:20] 0.174 0.173 0.157 -0.105 0.132 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...
  .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</code></pre>
</div>
<p>Both IRLBA and randomized SVD (RSVD) are much faster than the exact
SVD and usually yield only a negligible loss of accuracy. This motivates
their default use in many <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em> and
<em><a href="https://bioconductor.org/packages/3.19/scater" class="external-link">scater</a></em>
functions, at the cost of requiring users to set the seed to guarantee
reproducibility. IRLBA can occasionally fail to converge and require
more iterations (passed via <code>maxit=</code> in
<code>IrlbaParam()</code>), while RSVD involves an explicit trade-off
between accuracy and speed based on its oversampling parameter
(<code>p=</code>) and number of power iterations (<code>q=</code>). We
tend to prefer IRLBA as its default behavior is more accurate, though
RSVD is much faster for file-backed matrices.</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>The uncertainty from approximation error is sometimes aggravating.
“Why can’t my computer just give me the right answer?” One way to
alleviate this feeling is to quantify the approximation error on a small
test set like the sce we have here. Using the <code>ExactParam()</code>
class, visualize the error in PC1 coordinates compared to the RSVD
results.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>This code block calculates the exact PCA coordinates. Another thing
to note: PC vectors are only identified up to a sign flip. We can see
that the RSVD PC1 vector points in the</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">e.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu">ExactParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">e.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> num [1:2411, 1:20] -14.79 -5.79 -13.07 32.19 26.45 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...
  ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...
 - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...
 - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...
 - attr(*, "rotation")= num [1:500, 1:20] 0.174 0.173 0.157 -0.105 0.132 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...
  .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">e.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 PC1       PC2        PC3
cell_9769 -14.793684 18.470324 -0.4893474
cell_9770  -5.789032 13.347277  5.0560761
cell_9771 -13.066503 16.803152 -0.5602737
cell_9772  32.185950  6.697517 -0.6945423
cell_9773  26.452390  3.083474 -0.2271916</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 PC1       PC2        PC3
cell_9769  14.793780 18.470111 -0.4888676
cell_9770   5.789148 13.348438  5.0702153
cell_9771  13.066327 16.803423 -0.5562241
cell_9772 -32.186341  6.698347 -0.6892421
cell_9773 -26.452373  3.083974 -0.2299814</code></pre>
</div>
<p>For the sake of visualizing the error we can just flip the PC1
coordinates:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span> <span class="op">=</span> <span class="op">-</span><span class="fl">1</span> <span class="op">*</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span></span></code></pre>
</div>
<p>From there we can visualize the error with a histogram:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">error</span> <span class="op">&lt;-</span> <span class="fu">reducedDim</span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"PC1"</span><span class="op">]</span> <span class="op">-</span> </span>
<span>         <span class="fu">reducedDim</span><span class="op">(</span><span class="va">e.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"PC1"</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">data.frame</span><span class="op">(</span>approx_error <span class="op">=</span> <span class="va">error</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">approx_error</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/large_data-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It’s almost never more than .001 in this case.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="interoperability-with-popular-single-cell-analysis-ecosytems">Interoperability with popular single-cell analysis ecosytems<a class="anchor" aria-label="anchor" href="#interoperability-with-popular-single-cell-analysis-ecosytems"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h3>
<p><a href="https://satijalab.org/seurat" class="external-link">Seurat</a> is an R package
designed for QC, analysis, and exploration of single-cell RNA-seq data.
Seurat can be used to identify and interpret sources of heterogeneity
from single-cell transcriptomic measurements, and to integrate diverse
types of single-cell data. Seurat is developed and maintained by the <a href="https://satijalab.org/seurat/authors.html" class="external-link">Satija lab</a> and is
released under the <a href="https://opensource.org/license/mit/" class="external-link">MIT
license</a>.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Seurat</span><span class="op">)</span></span></code></pre>
</div>
<p>Although the basic processing of single-cell data with Bioconductor
packages (described in the <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">OSCA book</a>) and
with Seurat is very similar and will produce overall roughly identical
results, there is also complementary functionality with regard to cell
type annotation, dataset integration, and downstream analysis. To make
the most of both ecosystems it is therefore beneficial to be able to
easily switch between a <code>SeuratObject</code> and a
<code>SingleCellExperiment</code>. See also the Seurat <a href="https://satijalab.org/seurat/articles/conversion_vignette.html" class="external-link">conversion
vignette</a> for conversion to/from other popular single cell formats
such as the AnnData format used by <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">scanpy</a>.</p>
<p>Here, we demonstrate converting the Seurat object produced in
Seurat’s <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">PBMC
tutorial</a> to a <code>SingleCellExperiment</code> for further analysis
with functionality from OSCA/Bioconductor. We therefore need to first
install the <a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> package,
which is available from GitHub only.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"satijalab/seurat-data"</span><span class="op">)</span></span></code></pre>
</div>
<p>We then proceed by loading all required packages and installing the
PBMC dataset:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SeuratData</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">InstallData</span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span></code></pre>
</div>
<p>We then load the dataset as an <code>SeuratObject</code> and convert
it to a <code>SingleCellExperiment</code>.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use PBMC3K from SeuratData</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">LoadData</span><span class="op">(</span>ds <span class="op">=</span> <span class="st">"pbmc3k"</span>, type <span class="op">=</span> <span class="st">"pbmc3k.final"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">UpdateSeuratObject</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span></span>
<span></span>
<span><span class="va">pbmc.sce</span> <span class="op">&lt;-</span> <span class="fu">as.SingleCellExperiment</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc.sce</span></span></code></pre>
</div>
<p>Seurat also allows conversion from <code>SingleCellExperiment</code>
objects to Seurat objects; we demonstrate this here on the wild-type
chimera mouse gastrulation dataset.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<p>After some processing of the dataset, the actual conversion is
carried out with the <code>as.Seurat</code> function.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sobj</span> <span class="op">&lt;-</span> <span class="fu">as.Seurat</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">sobj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.mapped"</span></span>
<span></span>
<span><span class="va">sobj</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="scanpy">Scanpy<a class="anchor" aria-label="anchor" href="#scanpy"></a>
</h3>
<p><a href="https://scanpy.readthedocs.io" class="external-link">Scanpy</a> is a scalable
toolkit for analyzing single-cell gene expression data built jointly
with <a href="https://anndata.readthedocs.io/" class="external-link">anndata</a>. It includes
preprocessing, visualization, clustering, trajectory inference and
differential expression testing. The Python-based implementation
efficiently deals with datasets of more than one million cells. Scanpy
is developed and maintained by the <a href="">Theis lab</a> and is
released under a <a href="https://github.com/scverse/scanpy/blob/master/LICENSE" class="external-link">BSD-3-Clause
license</a>. Scanpy is part of the <a href="https://scverse.org/" class="external-link">scverse</a>, a Python-based ecosystem for
single-cell omics data analysis.</p>
<p>At the core of scanpy’s single-cell functionality is the
<code>anndata</code> data structure, scanpy’s integrated single-cell
data container, which is conceptually very similar to Bioconductor’s
<code>SingleCellExperiment</code> class.</p>
<p>Bioconductor’s <em><a href="https://bioconductor.org/packages/3.19/zellkonverter" class="external-link">zellkonverter</a></em>
package provides a lightweight interface between the Bioconductor
<code>SingleCellExperiment</code> data structure and the Python
<code>AnnData</code>-based single-cell analysis environment. The idea is
to enable users and developers to easily move data between these
frameworks to construct a multi-language analysis pipeline across
R/Bioconductor and Python.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">zellkonverter</span><span class="op">)</span></span></code></pre>
</div>
<p>The <code>readH5AD()</code> function can be used to read a
<code>SingleCellExperiment</code> from an H5AD file. Here, we use an
example H5AD file contained in the <em><a href="https://bioconductor.org/packages/3.19/zellkonverter" class="external-link">zellkonverter</a></em>
package.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_h5ad</span> <span class="op">&lt;-</span> <span class="fu">system.file</span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"krumsiek11.h5ad"</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">"zellkonverter"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">readH5AD</span><span class="op">(</span><span class="va">example_h5ad</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 11 640
metadata(2): highlights iroot
assays(1): X
rownames(11): Gata2 Gata1 ... EgrNab Gfi1
rowData names(0):
colnames(640): 0 1 ... 158-3 159-3
colData names(1): cell_type
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>We can also write a <code>SingleCellExperiment</code> to an H5AD file
with the <code>writeH5AD()</code> function. This is demonstrated below
on the wild-type chimera mouse gastrulation dataset.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.file</span> <span class="op">&lt;-</span> <span class="fu">tempfile</span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">writeH5AD</span><span class="op">(</span><span class="va">sce</span>, file <span class="op">=</span> <span class="va">out.file</span><span class="op">)</span></span></code></pre>
</div>
<p>The resulting H5AD file can then be read into Python using scanpy’s
<a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_h5ad.html" class="external-link">read_h5ad</a>
function and then directly used in compatible Python-based analysis
frameworks.</p>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<div id="exercise-1-out-of-memory-representation" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-out-of-memory-representation" class="callout-inner">
<h3 class="callout-title">Exercise 1: Out of memory representation</h3>
<div class="callout-content">
<p>Write the counts matrix of the wild-type chimera mouse gastrulation
dataset to an HDF5 file. Create another counts matrix that reads the
data from the HDF5 file. Compare memory usage of holding the entire
matrix in memory as opposed to holding the data out of memory.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>See the <code>HDF5Array</code> function for reading from HDF5 and the
<code>writeHDF5Array</code> function for writing to HDF5 from the <em><a href="https://bioconductor.org/packages/3.19/HDF5Array" class="external-link">HDF5Array</a></em>
package.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wt_out</span> <span class="op">&lt;-</span> <span class="fu">tempfile</span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">wt_counts</span> <span class="op">&lt;-</span> <span class="fu">counts</span><span class="op">(</span><span class="fu">WTChimeraData</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">writeHDF5Array</span><span class="op">(</span><span class="va">wt_counts</span>,</span>
<span>               name <span class="op">=</span> <span class="st">"wt_counts"</span>,</span>
<span>               file <span class="op">=</span> <span class="va">wt_out</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;29453 x 30703&gt; sparse HDF5Matrix object of type "double":
                       cell_1     cell_2     cell_3 ... cell_30702 cell_30703
ENSMUSG00000051951          0          0          0   .          0          0
ENSMUSG00000089699          0          0          0   .          0          0
ENSMUSG00000102343          0          0          0   .          0          0
ENSMUSG00000025900          0          0          0   .          0          0
ENSMUSG00000025902          0          0          0   .          0          0
               ...          .          .          .   .          .          .
ENSMUSG00000095041          0          1          2   .          0          0
ENSMUSG00000063897          0          0          0   .          0          0
ENSMUSG00000096730          0          0          0   .          0          0
ENSMUSG00000095742          0          0          0   .          0          0
         tomato-td          1          0          1   .          0          0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oom_wt</span> <span class="op">&lt;-</span> <span class="fu">HDF5Array</span><span class="op">(</span><span class="va">wt_out</span>, <span class="st">"wt_counts"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">object.size</span><span class="op">(</span><span class="va">wt_counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>1520366960 bytes</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">object.size</span><span class="op">(</span><span class="va">oom_wt</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>2488 bytes</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2-parallelization" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-parallelization" class="callout-inner">
<h3 class="callout-title">Exercise 2: Parallelization</h3>
<div class="callout-content">
<p>Perform a PCA analysis of the wild-type chimera mouse gastrulation
dataset using a multicore backend for parallel computation. Compare the
runtime of performing the PCA either in serial execution mode, in
multicore execution mode with 2 workers, and in multicore execution mode
with 3 workers.</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" aria-labelledby="headingHint2" data-bs-parent="#accordionHint2">
<div class="accordion-body">
<p>Use the function <code>system.time</code> to obtain the runtime of
each job.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.brain</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">system.time</span><span class="op">(</span><span class="op">{</span><span class="va">i.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce.brain</span>, </span>
<span>                             ncomponents <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                             BSPARAM <span class="op">=</span> <span class="fu">ExactParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                             BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">system.time</span><span class="op">(</span><span class="op">{</span><span class="va">i.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce.brain</span>, </span>
<span>                             ncomponents <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                             BSPARAM <span class="op">=</span> <span class="fu">ExactParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                             BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">system.time</span><span class="op">(</span><span class="op">{</span><span class="va">i.out</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce.brain</span>, </span>
<span>                             ncomponents <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                             BSPARAM <span class="op">=</span> <span class="fu">ExactParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>                             BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="further-reading" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further Reading</h3>
<div class="callout-content">
<ul>
<li>OSCA book, <a href="https://bioconductor.org/books/release/OSCA.advanced/dealing-with-big-data.html" class="external-link">Chapter
14</a>: Dealing with big data</li>
<li>The <code>BiocParallel</code> <a href="https://bioconductor.org/packages/3.19/BiocParallel/vignettes/Introduction_To_BiocParallel.html" class="external-link">intro
vignette</a>.</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Out-of-memory representations can be used to work with single-cell
datasets that are too large to fit in memory.</li>
<li>Parallelization of calculations across genes or cells is an
effective strategy for speeding up analysis of large single-cell
datasets.</li>
<li>Fast approximations for nearest neighbor search and singular value
composition can speed up essential steps of single-cell analysis with
minimal loss of accuracy.</li>
<li>Converter functions between existing single-cell data formats enable
analysis workflows that leverage complementary functionality from
poplular single-cell analysis ecosystems.</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] zellkonverter_1.14.1         Seurat_5.2.1
 [3] SeuratObject_5.0.2           sp_2.2-0
 [5] BiocSingular_1.20.0          BiocNeighbors_1.22.0
 [7] bluster_1.14.0               scran_1.32.0
 [9] MouseGastrulationData_1.18.0 SpatialExperiment_1.14.0
[11] BiocParallel_1.38.0          scater_1.32.1
[13] ggplot2_3.5.1                scuttle_1.14.0
[15] TENxBrainData_1.24.0         HDF5Array_1.32.1
[17] rhdf5_2.48.0                 DelayedArray_0.30.1
[19] SparseArray_1.4.8            S4Arrays_1.4.1
[21] abind_1.4-8                  Matrix_1.7-2
[23] SingleCellExperiment_1.26.0  SummarizedExperiment_1.34.0
[25] Biobase_2.64.0               GenomicRanges_1.56.2
[27] GenomeInfoDb_1.40.1          IRanges_2.38.1
[29] S4Vectors_0.42.1             BiocGenerics_0.50.0
[31] MatrixGenerics_1.16.0        matrixStats_1.5.0
[33] BiocStyle_2.32.1

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22          splines_4.4.2
  [3] later_1.4.1               filelock_1.0.3
  [5] tibble_3.2.1              polyclip_1.10-7
  [7] basilisk.utils_1.16.0     fastDummies_1.7.5
  [9] lifecycle_1.0.4           edgeR_4.2.2
 [11] globals_0.16.3            lattice_0.22-6
 [13] MASS_7.3-64               magrittr_2.0.3
 [15] plotly_4.10.4             limma_3.60.6
 [17] rmarkdown_2.29            yaml_2.3.10
 [19] metapod_1.12.0            httpuv_1.6.15
 [21] sctransform_0.4.1         spam_2.11-1
 [23] spatstat.sparse_3.1-0     reticulate_1.40.0
 [25] pbapply_1.7-2             cowplot_1.1.3
 [27] DBI_1.2.3                 RColorBrewer_1.1-3
 [29] zlibbioc_1.50.0           Rtsne_0.17
 [31] purrr_1.0.2               BumpyMatrix_1.12.0
 [33] rappdirs_0.3.3            GenomeInfoDbData_1.2.12
 [35] ggrepel_0.9.6             irlba_2.3.5.1
 [37] spatstat.utils_3.1-2      listenv_0.9.1
 [39] goftest_1.2-3             RSpectra_0.16-2
 [41] spatstat.random_3.3-2     dqrng_0.4.1
 [43] fitdistrplus_1.2-2        parallelly_1.42.0
 [45] DelayedMatrixStats_1.26.0 codetools_0.2-20
 [47] tidyselect_1.2.1          UCSC.utils_1.0.0
 [49] farver_2.1.2              ScaledMatrix_1.12.0
 [51] viridis_0.6.5             spatstat.explore_3.3-4
 [53] BiocFileCache_2.12.0      jsonlite_1.8.9
 [55] progressr_0.15.1          ggridges_0.5.6
 [57] survival_3.8-3            tools_4.4.2
 [59] ica_1.0-3                 Rcpp_1.0.14
 [61] glue_1.8.0                gridExtra_2.3
 [63] xfun_0.50                 dplyr_1.1.4
 [65] withr_3.0.2               formatR_1.14
 [67] BiocManager_1.30.25       fastmap_1.2.0
 [69] basilisk_1.16.0           rhdf5filters_1.16.0
 [71] digest_0.6.37             rsvd_1.0.5
 [73] R6_2.5.1                  mime_0.12
 [75] colorspace_2.1-1          scattermore_1.2
 [77] tensor_1.5                spatstat.data_3.1-4
 [79] RSQLite_2.3.9             tidyr_1.3.1
 [81] generics_0.1.3            data.table_1.16.4
 [83] renv_1.1.0                htmlwidgets_1.6.4
 [85] httr_1.4.7                uwot_0.2.2
 [87] pkgconfig_2.0.3           gtable_0.3.6
 [89] blob_1.2.4                lmtest_0.9-40
 [91] XVector_0.44.0            htmltools_0.5.8.1
 [93] dotCall64_1.2             scales_1.3.0
 [95] png_0.1-8                 spatstat.univar_3.1-1
 [97] knitr_1.49                reshape2_1.4.4
 [99] rjson_0.2.23              nlme_3.1-167
[101] curl_6.2.0                cachem_1.1.0
[103] zoo_1.8-12                stringr_1.5.1
[105] BiocVersion_3.19.1        KernSmooth_2.23-26
[107] parallel_4.4.2            miniUI_0.1.1.1
[109] vipor_0.4.7               AnnotationDbi_1.66.0
[111] pillar_1.10.1             grid_4.4.2
[113] vctrs_0.6.5               RANN_2.6.2
[115] promises_1.3.2            dbplyr_2.5.0
[117] beachmat_2.20.0           xtable_1.8-4
[119] cluster_2.1.8             beeswarm_0.4.0
[121] evaluate_1.0.3            magick_2.8.5
[123] cli_3.6.3                 locfit_1.5-9.11
[125] compiler_4.4.2            rlang_1.1.5
[127] crayon_1.5.3              future.apply_1.11.3
[129] labeling_0.4.3            plyr_1.8.9
[131] ggbeeswarm_0.7.2          stringi_1.8.4
[133] deldir_2.0-4              viridisLite_0.4.2
[135] munsell_0.5.1             Biostrings_2.72.1
[137] lazyeval_0.2.2            spatstat.geom_3.3-5
[139] dir.expiry_1.12.0         ExperimentHub_2.12.0
[141] RcppHNSW_0.6.0            patchwork_1.3.0
[143] sparseMatrixStats_1.16.0  bit64_4.6.0-1
[145] future_1.34.0             Rhdf5lib_1.26.0
[147] KEGGREST_1.44.1           statmod_1.5.0
[149] shiny_1.10.0              AnnotationHub_3.12.0
[151] ROCR_1.0-11               igraph_2.1.4
[153] memoise_2.0.1             bit_4.5.0.1              </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-hca"><p>Content from <a href="hca.html">Accessing data from the Human Cell Atlas (HCA)</a></p>
<hr>
<p>Last updated on 2025-02-07 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/hca.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to obtain single-cell reference maps from the Human Cell
Atlas?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn about different resources for public single-cell RNA-seq
data.</li>
<li>Access data from the Human Cell Atlas using the
<code>CuratedAtlasQueryR</code> package.</li>
<li>Query for cells of interest and download them into a
<code>SingleCellExperiment</code> object.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="single-cell-data-sources">Single Cell data sources<a class="anchor" aria-label="anchor" href="#single-cell-data-sources"></a>
</h1>
<div class="section level2">
<h2 id="hca-project">HCA Project<a class="anchor" aria-label="anchor" href="#hca-project"></a>
</h2>
<p>The Human Cell Atlas (HCA) is a large project that aims to learn from
and map every cell type in the human body. The project extracts spatial
and molecular characteristics in order to understand cellular function
and networks. It is an international collaborative that charts healthy
cells in the human body at all ages. There are about 37.2 trillion cells
in the human body. To read more about the project, head over to their
website at <a href="https://www.humancellatlas.org" class="external-link uri">https://www.humancellatlas.org</a>.</p>
</div>
<div class="section level2">
<h2 id="cellxgene">CELLxGENE<a class="anchor" aria-label="anchor" href="#cellxgene"></a>
</h2>
<p>CELLxGENE is a database and a suite of tools that help scientists to
find, download, explore, analyze, annotate, and publish single cell
data. It includes several analytic and visualization tools to help you
to discover single cell data patterns. To see the list of tools, browse
to <a href="https://cellxgene.cziscience.com/" class="external-link uri">https://cellxgene.cziscience.com/</a>.</p>
</div>
<div class="section level2">
<h2 id="cellxgene-census">CELLxGENE | Census<a class="anchor" aria-label="anchor" href="#cellxgene-census"></a>
</h2>
<p>The Census provides efficient computational tooling to access, query,
and analyze all single-cell RNA data from CZ CELLxGENE Discover. Using a
new access paradigm of cell-based slicing and querying, you can interact
with the data through TileDB-SOMA, or get slices in AnnData or Seurat
objects, thus accelerating your research by significantly minimizing
data harmonization at <a href="https://chanzuckerberg.github.io/cellxgene-census/" class="external-link uri">https://chanzuckerberg.github.io/cellxgene-census/</a>.</p>
</div>
<div class="section level2">
<h2 id="the-curatedatlasqueryr-project">The CuratedAtlasQueryR Project<a class="anchor" aria-label="anchor" href="#the-curatedatlasqueryr-project"></a>
</h2>
<p>The <code>CuratedAtlasQueryR</code> is an alternative package that
can also be used to access the CELLxGENE data from R through a tidy API.
The data has also been harmonized, curated, and re-annotated across
studies.</p>
<p><code>CuratedAtlasQueryR</code> supports data access and programmatic
exploration of the harmonized atlas. Cells of interest can be selected
based on ontology, tissue of origin, demographics, and disease. For
example, the user can select CD4 T helper cells across healthy and
diseased lymphoid tissue. The data for the selected cells can be
downloaded locally into SingleCellExperiment objects. Pseudo bulk counts
are also available to facilitate large-scale, summary analyses of
transcriptional profiles.</p>
<figure><img src="https://raw.githubusercontent.com/carpentries-incubator/bioc-scrnaseq/main/episodes/figures/curatedAtlasQuery.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level2">
<h2 id="data-sources-in-r-bioconductor">Data Sources in R / Bioconductor<a class="anchor" aria-label="anchor" href="#data-sources-in-r-bioconductor"></a>
</h2>
<p>There are a few options to access single cell data with R /
Bioconductor.</p>
<table class="table">
<colgroup>
<col width="29%">
<col width="41%">
<col width="29%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Target</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://bioconductor.org/packages/hca" class="external-link">hca</a></td>
<td><a href="https://www.humancellatlas.org/data-portal/" class="external-link">HCA Data
Portal API</a></td>
<td>Project, Sample, and File level HCA data</td>
</tr>
<tr class="even">
<td><a href="https://bioconductor.org/packages/cellxgenedp" class="external-link">cellxgenedp</a></td>
<td><a href="https://cellxgene.cziscience.com/" class="external-link">CellxGene</a></td>
<td>Human and mouse SC data including HCA</td>
</tr>
<tr class="odd">
<td><a href="https://stemangiola.github.io/CuratedAtlasQueryR/" class="external-link">CuratedAtlasQueryR</a></td>
<td><a href="https://cellxgene.cziscience.com/" class="external-link">CellxGene</a></td>
<td>fine-grained query capable CELLxGENE data including HCA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"CuratedAtlasQueryR"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="package-load">Package load<a class="anchor" aria-label="anchor" href="#package-load"></a>
</h2>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">CuratedAtlasQueryR</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="hca-metadata">HCA Metadata<a class="anchor" aria-label="anchor" href="#hca-metadata"></a>
</h2>
<p>The metadata allows the user to get a lay of the land of what is
available via the package. In this example, we are using the sample
database URL which allows us to get a small and quick subset of the
available metadata.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">get_metadata</span><span class="op">(</span>remote_url <span class="op">=</span> <span class="fu">CuratedAtlasQueryR</span><span class="fu">::</span><span class="va">SAMPLE_DATABASE_URL</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">collect</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>Get a view of the first 10 columns in the metadata with
<code>glimpse()</code></p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">glimpse</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: ??
Columns: 10
Database: DuckDB v0.10.2 [unknown@Linux 6.5.0-1021-azure:R 4.4.0/:memory:]
$ cell_                             &lt;chr&gt; "TTATGCTAGGGTGTTG_12", "GCTTGAACATGG…
$ sample_                           &lt;chr&gt; "039c558ca1c43dc74c563b58fe0d6289", …
$ cell_type                         &lt;chr&gt; "mature NK T cell", "mature NK T cel…
$ cell_type_harmonised              &lt;chr&gt; "immune_unclassified", "cd8 tem", "i…
$ confidence_class                  &lt;dbl&gt; 5, 3, 5, 5, 5, 5, 5, 5, 5, 5, 5, 1, …
$ cell_annotation_azimuth_l2        &lt;chr&gt; "gdt", "cd8 tem", "cd8 tem", "cd8 te…
$ cell_annotation_blueprint_singler &lt;chr&gt; "cd4 tem", "cd8 tem", "cd8 tcm", "cl…
$ cell_annotation_monaco_singler    &lt;chr&gt; "natural killer", "effector memory c…
$ sample_id_db                      &lt;chr&gt; "0c1d320a7d0cbbc281a535912722d272", …
$ `_sample_name`                    &lt;chr&gt; "BPH340PrSF_Via___transition zone of…</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="a-tangent-on-the-pipe-operator">A tangent on the pipe operator<a class="anchor" aria-label="anchor" href="#a-tangent-on-the-pipe-operator"></a>
</h2>
<p>The vignette materials provided by <code>CuratedAtlasQueryR</code>
show the use of the ‘native’ R pipe (implemented after R version
<code>4.1.0</code>). For those not familiar with the pipe operator
(<code>|&gt;</code>), it allows you to chain functions by passing the
left-hand side as the first argument to the function on the right-hand
side. It is used extensively in the <code>tidyverse</code> dialect of R,
especially within the <a href="https://dplyr.tidyverse.org/" class="external-link"><code>dplyr</code> package</a>.</p>
<p>The pipe operator can be read as “and then”. Thankfully, R doesn’t
care about whitespace, so it’s common to start a new line after a pipe.
Together these points enable users to “chain” complex sequences of
commands into readable blocks.</p>
<p>In this example, we start with the built-in <code>mtcars</code>
dataset and then filter to rows where <code>cyl</code> is not equal to
4, and then compute the mean <code>disp</code> value by each unique
<code>cyl</code> value.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtcars</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">cyl</span> <span class="op">!=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">summarise</span><span class="op">(</span>avg_disp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">disp</span><span class="op">)</span>,</span>
<span>            .by <span class="op">=</span> <span class="va">cyl</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  cyl avg_disp
1   6 183.3143
2   8 353.1000</code></pre>
</div>
<p>This command is equivalent to the following:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summarise</span><span class="op">(</span><span class="fu">filter</span><span class="op">(</span><span class="va">mtcars</span>, <span class="va">cyl</span> <span class="op">!=</span> <span class="fl">4</span><span class="op">)</span>, mean_disp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">disp</span><span class="op">)</span>, .by <span class="op">=</span> <span class="va">cyl</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="exploring-the-metadata">Exploring the metadata<a class="anchor" aria-label="anchor" href="#exploring-the-metadata"></a>
</h2>
<p>Let’s examine the metadata to understand what information it
contains.</p>
<p>We can tally the tissue types across datasets to see what tissues the
experimental data come from:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">distinct</span><span class="op">(</span><span class="va">tissue</span>, <span class="va">dataset_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 33 × 2
   tissue                   n
   &lt;chr&gt;                &lt;int&gt;
 1 blood                   17
 2 kidney                   8
 3 cortex of kidney         7
 4 heart left ventricle     7
 5 renal medulla            6
 6 respiratory airway       6
 7 bone marrow              4
 8 kidney blood vessel      4
 9 lung                     4
10 renal pelvis             4
# ℹ 23 more rows</code></pre>
</div>
<p>We can do the same for the assay types:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">distinct</span><span class="op">(</span><span class="va">assay</span>, <span class="va">dataset_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">assay</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 12 × 2
   assay                              n
   &lt;chr&gt;                          &lt;int&gt;
 1 10x 3' v1                          1
 2 10x 3' v2                         27
 3 10x 3' v3                         21
 4 10x 5' v1                          7
 5 10x 5' v2                          2
 6 Drop-seq                           1
 7 Seq-Well                           2
 8 Slide-seq                          4
 9 Smart-seq2                         1
10 Visium Spatial Gene Expression     7
11 scRNA-seq                          4
12 sci-RNA-seq                        1</code></pre>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Look through the full list of metadata column names. Do any other
metadata columns jump out as interesting to you for your work?</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="downloading-single-cell-data">Downloading single cell data<a class="anchor" aria-label="anchor" href="#downloading-single-cell-data"></a>
</h2>
<p>The data can be provided as either “counts” or counts per million
“cpm” as given by the <code>assays</code> argument in the
<code>get_single_cell_experiment()</code> function. By default, the
<code>SingleCellExperiment</code> provided will contain only the
‘counts’ data.</p>
<p>For the sake of demonstration, we’ll focus this small subset of
samples. We use the <code>filter()</code> function from the
<code>dplyr</code> package to identify cells meeting the following
criteria:</p>
<ul>
<li>African ethnicity</li>
<li>10x assay</li>
<li>lung parenchyma tissue</li>
<li>CD4 cells</li>
</ul>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_subset</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span></span>
<span>        <span class="va">ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">grepl</span><span class="op">(</span><span class="st">"10x"</span>, <span class="va">assay</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">grepl</span><span class="op">(</span><span class="st">"CD4"</span>, <span class="va">cell_type</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre>
</div>
<p>Out of the 111355 cells in the sample database, 1571 cells meet this
criteria.</p>
<p>Now we can use <code>get_single_cell_experiment()</code>:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">&lt;-</span> <span class="va">sample_subset</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">get_single_cell_experiment</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 36229 1571
metadata(0):
assays(1): counts
rownames(36229): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
rowData names(0):
colnames(1571): ACACCAAAGCCACCTG_SC18_1 TCAGCTCCAGACAAGC_SC18_1 ...
  CAGCATAAGCTAACAA_F02607_1 AAGGAGCGTATAATGG_F02607_1
colData names(56): sample_ cell_type ... updated_at_y original_cell_id
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>You can provide different arguments to
<code>get_single_cell_experiment()</code> to get different formats or
subsets of the data, like data scaled to counts per million:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_subset</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">get_single_cell_experiment</span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 36229 1571
metadata(0):
assays(1): cpm
rownames(36229): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
rowData names(0):
colnames(1571): ACACCAAAGCCACCTG_SC18_1 TCAGCTCCAGACAAGC_SC18_1 ...
  CAGCATAAGCTAACAA_F02607_1 AAGGAGCGTATAATGG_F02607_1
colData names(56): sample_ cell_type ... updated_at_y original_cell_id
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>or data on only specific genes:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">&lt;-</span> <span class="va">sample_subset</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">get_single_cell_experiment</span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span>, features <span class="op">=</span> <span class="st">"PUM1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 1 1571
metadata(0):
assays(1): cpm
rownames(1): PUM1
rowData names(0):
colnames(1571): ACACCAAAGCCACCTG_SC18_1 TCAGCTCCAGACAAGC_SC18_1 ...
  CAGCATAAGCTAACAA_F02607_1 AAGGAGCGTATAATGG_F02607_1
colData names(56): sample_ cell_type ... updated_at_y original_cell_id
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>Or if needed, the H5 <code>SingleCellExperiment</code> can be
returned a Seurat object (note that this may take a long time and use a
lot of memory depending on how many cells you are requesting).</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">&lt;-</span> <span class="va">sample_subset</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">get_seurat</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="save-your-singlecellexperiment">Save your <code>SingleCellExperiment</code>
<a class="anchor" aria-label="anchor" href="#save-your-singlecellexperiment"></a>
</h2>
<p>Once you have a dataset you’re happy with, you’ll probably want to
save it. The recommended way of saving these
<code>SingleCellExperiment</code> objects is to use
<code>saveHDF5SummarizedExperiment</code> from the
<code>HDF5Array</code> package.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> <span class="fu">saveHDF5SummarizedExperiment</span><span class="op">(</span><span class="st">"single_cell_counts"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<div id="exercise-1-basic-counting-piping" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-basic-counting-piping" class="callout-inner">
<h3 class="callout-title">Exercise 1: Basic counting + piping</h3>
<div class="callout-content">
<p>Use <code>count</code> and <code>arrange</code> to get the number of
cells per tissue in descending order.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 33 × 2
   tissue                          n
   &lt;chr&gt;                       &lt;int&gt;
 1 cortex of kidney            36940
 2 kidney                      23549
 3 lung parenchyma             16719
 4 renal medulla                7729
 5 respiratory airway           7153
 6 blood                        4248
 7 bone marrow                  4113
 8 heart left ventricle         1454
 9 transition zone of prostate  1140
10 lung                         1137
# ℹ 23 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2-tissue-type-counting" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-tissue-type-counting" class="callout-inner">
<h3 class="callout-title">Exercise 2: Tissue &amp; type counting</h3>
<div class="callout-content">
<p><code>count()</code> can group by multiple factors by simply adding
another grouping column as an additional argument. Get a tally of the
highest number of cell types per tissue combination. What tissue has the
most numerous type of cells?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">tissue</span>, <span class="va">cell_type</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">head</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1 × 3
  tissue           cell_type                              n
  &lt;chr&gt;            &lt;chr&gt;                              &lt;int&gt;
1 cortex of kidney epithelial cell of proximal tubule 29986</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-3-comparing-metadata-categories" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3-comparing-metadata-categories" class="callout-inner">
<h3 class="callout-title">Exercise 3: Comparing metadata categories</h3>
<div class="callout-content">
<p>Spot some differences between the <code>tissue</code> and
<code>tissue_harmonised</code> columns. Use <code>count</code> to
summarise.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 33 × 2
   tissue                          n
   &lt;chr&gt;                       &lt;int&gt;
 1 cortex of kidney            36940
 2 kidney                      23549
 3 lung parenchyma             16719
 4 renal medulla                7729
 5 respiratory airway           7153
 6 blood                        4248
 7 bone marrow                  4113
 8 heart left ventricle         1454
 9 transition zone of prostate  1140
10 lung                         1137
# ℹ 23 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">tissue_harmonised</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 19 × 2
   tissue_harmonised     n
   &lt;chr&gt;             &lt;int&gt;
 1 kidney            68851
 2 lung              25737
 3 blood              4248
 4 bone               4113
 5 heart              1454
 6 lymph node         1210
 7 prostate           1156
 8 intestine large     816
 9 liver               793
10 thymus              753
11 intestine small     530
12 eye                 437
13 intestine           360
14 esophagus           334
15 nose                290
16 vasculature         143
17 brain                97
18 adrenal gland        20
19 axilla               13</code></pre>
</div>
<p>For example you can see that <code>tissue_harmonised</code> merges
the <code>cortex of kidney</code> and <code>kidney</code> groups in
<code>tissue</code>.</p>
<p>To see the full list of curated columns in the metadata, see the
Details section in the <code>?get_metadata</code> documentation
page.</p>
</div>
</div>
</div>
</div>
<div id="exercise-4-highly-specific-cell-groups" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4-highly-specific-cell-groups" class="callout-inner">
<h3 class="callout-title">Exercise 4: Highly specific cell groups</h3>
<div class="callout-content">
<p>Now that we are a little familiar with navigating the metadata, let’s
obtain a <code>SingleCellExperiment</code> of 10X scRNA-seq counts of
<code>cd8 tem</code> <code>lung</code> cells for females older than
<code>80</code> with <code>COVID-19</code>. Note: Use the harmonized
columns, where possible.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">filter</span><span class="op">(</span></span>
<span>        <span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span> <span class="op">&amp;</span></span>
<span>        <span class="va">age_days</span> <span class="op">&gt;</span> <span class="fl">80</span> <span class="op">*</span> <span class="fl">365</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">grepl</span><span class="op">(</span><span class="st">"10x"</span>, <span class="va">assay</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">disease</span> <span class="op">==</span> <span class="st">"COVID-19"</span> <span class="op">&amp;</span>  </span>
<span>        <span class="va">tissue_harmonised</span> <span class="op">==</span> <span class="st">"lung"</span> <span class="op">&amp;</span> </span>
<span>        <span class="va">cell_type_harmonised</span> <span class="op">==</span> <span class="st">"cd8 tem"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">get_single_cell_experiment</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 36229 12
metadata(0):
assays(1): counts
rownames(36229): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
rowData names(0):
colnames(12): TCATCATCATAACCCA_1 TATCTGTCAGAACCGA_1 ...
  CCCTTAGCATGACTTG_1 CAGTTCCGTAGCGTAG_1
colData names(56): sample_ cell_type ... updated_at_y original_cell_id
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<p>You can see we don’t get very many cells given the strict set of
conditions we used.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The <code>CuratedAtlasQueryR</code> package provides programmatic
access to single-cell reference maps from the Human Cell Atlas.</li>
<li>The package provides functionality to query for cells of interest
and to download them into a <code>SingleCellExperiment</code>
object.</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] dplyr_1.1.4                  CuratedAtlasQueryR_1.2.0
 [3] scDblFinder_1.18.0           scran_1.32.0
 [5] scater_1.32.0                ggplot2_3.5.1
 [7] scuttle_1.14.0               EnsDb.Mmusculus.v79_2.99.0
 [9] ensembldb_2.28.0             AnnotationFilter_1.28.0
[11] GenomicFeatures_1.56.0       AnnotationDbi_1.66.0
[13] DropletUtils_1.24.0          MouseGastrulationData_1.18.0
[15] SpatialExperiment_1.14.0     SingleCellExperiment_1.26.0
[17] SummarizedExperiment_1.34.0  Biobase_2.64.0
[19] GenomicRanges_1.56.0         GenomeInfoDb_1.40.1
[21] IRanges_2.38.0               S4Vectors_0.42.0
[23] BiocGenerics_0.50.0          MatrixGenerics_1.16.0
[25] matrixStats_1.3.0            BiocStyle_2.32.0

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.0-3     ProtGenerics_1.36.0
  [3] bitops_1.0-7              httr_1.4.7
  [5] RColorBrewer_1.1-3        tools_4.4.1
  [7] sctransform_0.4.1         utf8_1.2.4
  [9] R6_2.5.1                  HDF5Array_1.32.0
 [11] uwot_0.2.2                lazyeval_0.2.2
 [13] rhdf5filters_1.16.0       withr_3.0.0
 [15] sp_2.1-4                  gridExtra_2.3
 [17] progressr_0.14.0          cli_3.6.2
 [19] formatR_1.14              spatstat.explore_3.2-7
 [21] fastDummies_1.7.3         Seurat_5.1.0
 [23] spatstat.data_3.0-4       ggridges_0.5.6
 [25] pbapply_1.7-2             Rsamtools_2.20.0
 [27] R.utils_2.12.3            parallelly_1.37.1
 [29] limma_3.60.2              RSQLite_2.3.7
 [31] generics_0.1.3            BiocIO_1.14.0
 [33] spatstat.random_3.2-3     ica_1.0-3
 [35] Matrix_1.7-0              ggbeeswarm_0.7.2
 [37] fansi_1.0.6               abind_1.4-5
 [39] R.methodsS3_1.8.2         lifecycle_1.0.4
 [41] yaml_2.3.8                edgeR_4.2.0
 [43] rhdf5_2.48.0              SparseArray_1.4.8
 [45] BiocFileCache_2.12.0      Rtsne_0.17
 [47] grid_4.4.1                blob_1.2.4
 [49] promises_1.3.0            dqrng_0.4.1
 [51] ExperimentHub_2.12.0      crayon_1.5.2
 [53] miniUI_0.1.1.1            lattice_0.22-6
 [55] beachmat_2.20.0           cowplot_1.1.3
 [57] KEGGREST_1.44.0           magick_2.8.3
 [59] pillar_1.9.0              knitr_1.47
 [61] metapod_1.12.0            rjson_0.2.21
 [63] xgboost_1.7.7.1           future.apply_1.11.2
 [65] codetools_0.2-20          leiden_0.4.3.1
 [67] glue_1.7.0                data.table_1.15.4
 [69] vctrs_0.6.5               png_0.1-8
 [71] spam_2.10-0               gtable_0.3.5
 [73] assertthat_0.2.1          cachem_1.1.0
 [75] xfun_0.44                 S4Arrays_1.4.1
 [77] mime_0.12                 survival_3.6-4
 [79] statmod_1.5.0             bluster_1.14.0
 [81] fitdistrplus_1.1-11       ROCR_1.0-11
 [83] nlme_3.1-164              bit64_4.0.5
 [85] filelock_1.0.3            RcppAnnoy_0.0.22
 [87] BumpyMatrix_1.12.0        irlba_2.3.5.1
 [89] vipor_0.4.7               KernSmooth_2.23-24
 [91] colorspace_2.1-0          DBI_1.2.3
 [93] duckdb_0.10.2             tidyselect_1.2.1
 [95] bit_4.0.5                 compiler_4.4.1
 [97] curl_5.2.1                BiocNeighbors_1.22.0
 [99] DelayedArray_0.30.1       plotly_4.10.4
[101] rtracklayer_1.64.0        scales_1.3.0
[103] lmtest_0.9-40             rappdirs_0.3.3
[105] goftest_1.2-3             stringr_1.5.1
[107] digest_0.6.35             spatstat.utils_3.0-4
[109] rmarkdown_2.27            XVector_0.44.0
[111] htmltools_0.5.8.1         pkgconfig_2.0.3
[113] sparseMatrixStats_1.16.0  highr_0.11
[115] dbplyr_2.5.0              fastmap_1.2.0
[117] rlang_1.1.3               htmlwidgets_1.6.4
[119] UCSC.utils_1.0.0          shiny_1.8.1.1
[121] DelayedMatrixStats_1.26.0 zoo_1.8-12
[123] jsonlite_1.8.8            BiocParallel_1.38.0
[125] R.oo_1.26.0               BiocSingular_1.20.0
[127] RCurl_1.98-1.14           magrittr_2.0.3
[129] GenomeInfoDbData_1.2.12   dotCall64_1.1-1
[131] patchwork_1.2.0           Rhdf5lib_1.26.0
[133] munsell_0.5.1             Rcpp_1.0.12
[135] viridis_0.6.5             reticulate_1.37.0
[137] stringi_1.8.4             zlibbioc_1.50.0
[139] MASS_7.3-60.2             AnnotationHub_3.12.0
[141] plyr_1.8.9                parallel_4.4.1
[143] listenv_0.9.1             ggrepel_0.9.5
[145] deldir_2.0-4              Biostrings_2.72.1
[147] splines_4.4.1             tensor_1.5
[149] locfit_1.5-9.9            igraph_2.0.3
[151] spatstat.geom_3.2-9       RcppHNSW_0.6.0
[153] reshape2_1.4.4            ScaledMatrix_1.12.0
[155] BiocVersion_3.19.1        XML_3.99-0.16.1
[157] evaluate_0.23             SeuratObject_5.0.2
[159] renv_1.0.11               BiocManager_1.30.23
[161] httpuv_1.6.15             polyclip_1.10-6
[163] RANN_2.6.1                tidyr_1.3.1
[165] purrr_1.0.2               future_1.33.2
[167] scattermore_1.2           rsvd_1.0.5
[169] xtable_1.8-4              restfulr_0.0.15
[171] RSpectra_0.16-1           later_1.3.2
[173] viridisLite_0.4.2         tibble_3.2.1
[175] memoise_2.0.1             beeswarm_0.4.0
[177] GenomicAlignments_1.40.0  cluster_2.1.6
[179] globals_0.16.3           </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/ccb-hms/osca-carpentries/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/ccb-hms/osca-carpentries/" class="external-link">Source</a></p>
				<p><a href="https://github.com/ccb-hms/osca-carpentries/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:christopher_magnano@hms.harvard.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.7" class="external-link">pegboard (0.7.7)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://ccb-hms.github.io/osca-carpentries/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "single-cell RNA-seq, The Carpentries, lesson, genomics, bioconductor",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://ccb-hms.github.io/osca-carpentries/aio.html",
  "identifier": "https://ccb-hms.github.io/osca-carpentries/aio.html",
  "dateCreated": "2024-01-10",
  "dateModified": "2025-02-18",
  "datePublished": "2025-02-18"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

