---
permalink: index.html
site: sandpaper::sandpaper_site
---

In the last few years, the profiling of a large number of genome-wide features
in individual cells has become routine. Consequently, a plethora of tools for
the analysis of single-cell data has been developed, making it hard to understand
the critical steps in the analysis workflow and the best methods for each objective of oneâ€™s study.

This tutorial aims to provide a solid foundation in using [Bioconductor](https://bioconductor.org)
tools for single-cell RNA-seq (scRNA-seq) analysis by walking through various steps of
typical workflows using example datasets.

This tutorial is based on the the online book "Orchestrating Single-Cell
Analysis with Bioconductor" ([OSCA](https://bioconductor.org/books/release/OSCA/)),
started in 2018 and continuously updated by many contributors from the Bioconductor community.
Like the book, this tutorial strives to be of interest to the experimental biologists
wanting to analyze their data and to the bioinformaticians approaching single-cell data.

This is a new lesson built with [The Carpentries Workbench][workbench]. 


[workbench]: https://carpentries.github.io/sandpaper-docs

::::::::::::::::::::::::::::::::::::::::::  prereq

## Prerequisites

- Familiarity with R/Bioconductor, such as the
[Introduction to data analysis with R and Bioconductor](https://carpentries-incubator.github.io/bioc-intro/)
lesson.
- Familiarity with the biology of gene expression and scRNA-seq, such as the review article 
[A practical guide to single-cell RNA-sequencing](https://doi.org/10.1186/s13073-017-0467-4) by Haque et.al.
  

::::::::::::::::::::::::::::::::::::::::::::::::::


