<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Orchestrating Large-Scale Single-Cell Analysis with Bioconductor: Multi-sample analyses</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../multi-sample.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Orchestrating Large-Scale Single-Cell Analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 58%" class="percentage">
    58%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 58%" aria-valuenow="58" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../multi-sample.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="intro-sce.html">1. Introduction to Bioconductor and the SingleCellExperiment class</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="eda_qc.html">2. Exploratory data analysis and quality control</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="cell_type_annotation.html">3. Cell type annotation</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Multi-sample analyses
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#setup-and-data-exploration">Setup and data exploration</a></li>
<li><a href="#correcting-batch-effects">Correcting batch effects</a></li>
<li><a href="#differential-expression">Differential Expression</a></li>
<li><a href="#differential-abundance">Differential Abundance</a></li>
<li><a href="#exercises">Exercises</a></li>
<li><a href="#session-info">Session Info</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="large_data.html">5. Working with large data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="hca.html">6. Accessing data from the Human Cell Atlas (HCA)</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/cell_type_annotation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/large_data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/cell_type_annotation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Cell type annotation
        </a>
        <a class="chapter-link float-end" href="../instructor/large_data.html" rel="next">
          Next: Working with large...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Multi-sample analyses</h1>
        <p>Last updated on 2025-03-14 |

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/multi-sample.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 45 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we integrate data from multiple batches, samples, and
studies?</li>
<li>How can we identify differentially expressed genes between
experimental conditions for each cell type?</li>
<li>How can we identify changes in cell type abundance between
experimental conditions?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Correct batch effects and diagnose potential problems such as
over-correction.</li>
<li>Perform differential expression comparisons between conditions based
on pseudo-bulk samples.</li>
<li>Perform differential abundance comparisons between conditions.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="setup-and-data-exploration">Setup and data exploration<a class="anchor" aria-label="anchor" href="#setup-and-data-exploration"></a></h2>
<hr class="half-width"><p>As before, we will use the the wild-type data from the Tal1 chimera
experiment:</p>
<ul><li>Sample 5: E8.5 injected cells (tomato positive), pool 3</li>
<li>Sample 6: E8.5 host cells (tomato negative), pool 3</li>
<li>Sample 7: E8.5 injected cells (tomato positive), pool 4</li>
<li>Sample 8: E8.5 host cells (tomato negative), pool 4</li>
<li>Sample 9: E8.5 injected cells (tomato positive), pool 5</li>
<li>Sample 10: E8.5 host cells (tomato negative), pool 5</li>
</ul><p>Note that this is a paired design in which for each biological
replicate (pool 3, 4, and 5), we have both host and injected cells.</p>
<p>We start by loading the data and doing a quick exploratory analysis,
essentially applying the normalization and visualization techniques that
we have seen in the previous lectures to all samples. Note that this
time we’re selecting samples 5 to 10, not just 5 by itself. Also note
the <code>type = "processed"</code> argument: we are explicitly
selecting the version of the data that has already been QC
processed.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">MouseGastrulationData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">edgeR</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scater</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">WTChimeraData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 29453 20935
metadata(0):
assays(1): counts
rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...
  ENSMUSG00000095742 tomato-td
rowData names(2): ENSEMBL SYMBOL
colnames(20935): cell_9769 cell_9770 ... cell_30702 cell_30703
colData names(11): cell barcode ... doub.density sizeFactor
reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 20935 rows and 11 columns
                  cell          barcode    sample       stage    tomato
           &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;
cell_9769    cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE
cell_9770    cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE
cell_9771    cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE
cell_9772    cell_9772 AAACCTGCATACTCTT         5        E8.5      TRUE
cell_9773    cell_9773 AAACGGGTCAACACCA         5        E8.5      TRUE
...                ...              ...       ...         ...       ...
cell_30699  cell_30699 TTTGTCACAGCTCGCA        10        E8.5     FALSE
cell_30700  cell_30700 TTTGTCAGTCTAGTCA        10        E8.5     FALSE
cell_30701  cell_30701 TTTGTCATCATCGGAT        10        E8.5     FALSE
cell_30702  cell_30702 TTTGTCATCATTATCC        10        E8.5     FALSE
cell_30703  cell_30703 TTTGTCATCCCATTTA        10        E8.5     FALSE
                pool stage.mapped        celltype.mapped closest.cell
           &lt;integer&gt;  &lt;character&gt;            &lt;character&gt;  &lt;character&gt;
cell_9769          3        E8.25             Mesenchyme   cell_24159
cell_9770          3         E8.5            Endothelium   cell_96660
cell_9771          3         E8.5              Allantois  cell_134982
cell_9772          3         E8.5             Erythroid3  cell_133892
cell_9773          3        E8.25             Erythroid1   cell_76296
...              ...          ...                    ...          ...
cell_30699         5         E8.5             Erythroid3   cell_38810
cell_30700         5         E8.5       Surface ectoderm   cell_38588
cell_30701         5        E8.25 Forebrain/Midbrain/H..   cell_66082
cell_30702         5         E8.5             Erythroid3  cell_138114
cell_30703         5         E8.0                Doublet   cell_92644
           doub.density sizeFactor
              &lt;numeric&gt;  &lt;numeric&gt;
cell_9769    0.02985045    1.41243
cell_9770    0.00172753    1.22757
cell_9771    0.01338013    1.15439
cell_9772    0.00218402    1.28676
cell_9773    0.00211723    1.78719
...                 ...        ...
cell_30699   0.00146287   0.389311
cell_30700   0.00374155   0.588784
cell_30701   0.05651258   0.624455
cell_30702   0.00108837   0.550807
cell_30703   0.82369305   1.184919</code></pre>
</div>
<p>For the sake of making these examples run faster, we drop some
problematic types (stripped nuclei and doublets) and also randomly
select 50% cells per sample.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">drop</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"Doublet"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">drop</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">29482</span><span class="op">)</span></span>
<span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">tapply</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">perc</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">0.50</span> <span class="op">*</span> <span class="fu">length</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">sample</span><span class="op">(</span><span class="va">x</span>, <span class="va">perc</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">idx</span><span class="op">]</span></span></code></pre>
</div>
<p>We now normalize the data, run some dimensionality reduction steps,
and visualize them in a tSNE plot. In this case we happen to have a ton
of cell types to visualize, so we define a custom palette with a lot of
visually distinct colors (adapted from the <code>polychrome</code>
palette in the <a href="https://cran.r-project.org/web/packages/pals/vignettes/pals_examples.html" class="external-link"><code>pals</code>
package</a>).</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu">modelGeneVar</span><span class="op">(</span><span class="va">sce</span>, block <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span></span>
<span><span class="va">chosen.hvgs</span> <span class="op">&lt;-</span> <span class="va">dec</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">chosen.hvgs</span>, ntop <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu">as.factor</span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">color_vec</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"#5A5156"</span>, <span class="st">"#E4E1E3"</span>, <span class="st">"#F6222E"</span>, <span class="st">"#FE00FA"</span>, <span class="st">"#16FF32"</span>, <span class="st">"#3283FE"</span>, </span>
<span>               <span class="st">"#FEAF16"</span>, <span class="st">"#B00068"</span>, <span class="st">"#1CFFCE"</span>, <span class="st">"#90AD1C"</span>, <span class="st">"#2ED9FF"</span>, <span class="st">"#DEA0FD"</span>, </span>
<span>               <span class="st">"#AA0DFE"</span>, <span class="st">"#F8A19F"</span>, <span class="st">"#325A9B"</span>, <span class="st">"#C4451C"</span>, <span class="st">"#1C8356"</span>, <span class="st">"#85660D"</span>, </span>
<span>               <span class="st">"#B10DA1"</span>, <span class="st">"#3B00FB"</span>, <span class="st">"#1CBE4F"</span>, <span class="st">"#FA0087"</span>, <span class="st">"#333333"</span>, <span class="st">"#F7E1A0"</span>, </span>
<span>               <span class="st">"#C075A6"</span>, <span class="st">"#782AB6"</span>, <span class="st">"#AAF400"</span>, <span class="st">"#BDCDFF"</span>, <span class="st">"#822E1C"</span>, <span class="st">"#B5EFB5"</span>, </span>
<span>               <span class="st">"#7ED7D1"</span>, <span class="st">"#1C7F93"</span>, <span class="st">"#D85FF7"</span>, <span class="st">"#683B79"</span>, <span class="st">"#66B0FF"</span>, <span class="st">"#FBE426"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_vec</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-2-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are evident sample effects. Depending on the analysis that you
want to perform you may want to remove or retain the sample effect. For
instance, if the goal is to identify cell types with a clustering
method, one may want to remove the sample effects with “batch effect”
correction methods.</p>
<p>For now, let’s assume that we want to remove this effect.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>It seems like samples 5 and 6 are separated off from the others in
gene expression space. Given the group of cells in each sample, why
might this make sense versus some other pair of samples? What is the
factor presumably leading to this difference?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Samples 5 and 6 were from the same “pool” of cells. Looking at the
documentation for the dataset under <code>?WTChimeraData</code> we see
that the pool variable is defined as: “Integer, embryo pool from which
cell derived; samples with same value are matched.” So samples 5 and 6
have an experimental factor in common which causes a shared, systematic
difference in gene expression profiles compared to the other samples.
That’s why you can see many of isolated blue/orange clusters on the
first TSNE plot. If you were developing single-cell library preparation
protocols you might want to preserve this effect to understand how
variation in pools leads to variation in expression, but for now, given
that we’re investigating other effects, we’ll want to remove this as
undesired technical variation.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="correcting-batch-effects">Correcting batch effects<a class="anchor" aria-label="anchor" href="#correcting-batch-effects"></a></h2>
<hr class="half-width"><p>We “correct” the effect of samples with the
<code>correctExperiment</code> function in the <code>batchelor</code>
package and using the <code>sample</code> column as batch.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">10102</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu">correctExperiments</span><span class="op">(</span></span>
<span>    <span class="va">sce</span>, </span>
<span>    batch <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>    subset.row <span class="op">=</span> <span class="va">chosen.hvgs</span>,</span>
<span>    PARAM <span class="op">=</span> <span class="fu">FastMnnParam</span><span class="op">(</span></span>
<span>        merge.order <span class="op">=</span> <span class="fu">list</span><span class="op">(</span></span>
<span>            <span class="fu">list</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="co"># WT (3 replicates)</span></span>
<span>            <span class="fu">list</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>  <span class="co"># td-Tomato (3 replicates)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">merged</span>, dimred <span class="op">=</span> <span class="st">"corrected"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">merged</span>, colour_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can also see that when coloring by cell type, the cell types are
now nicely confined to their own clusters for the most part:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotTSNE</span><span class="op">(</span><span class="va">merged</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">color_vec</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Once we removed the sample batch effect, we can proceed with the
Differential Expression Analysis.</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>True or False: after batch correction, no batch-level information is
present in the corrected data.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>False. Batch-level data can be retained through confounding with
experimental factors or poor ability to distinguish experimental effects
from batch effects. Remember, the changes needed to correct the data are
empirically estimated, so they can carry along error.</p>
<p>While batch effect correction algorithms usually do a pretty good
job, it’s smart to do a sanity check for batch effects at the end of
your analysis. You always want to make sure that that effect you’re
resting your paper submission on isn’t driven by batch effects.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a></h2>
<hr class="half-width"><p>In order to perform a differential expression analysis, we need to
identify groups of cells across samples/conditions (depending on the
experimental design and the final aim of the experiment).</p>
<p>As previously seen, we have two ways of grouping cells, cell
clustering and cell labeling. In our case we will focus on this second
aspect to group cells according to the already annotated cell types to
proceed with the computation of the pseudo-bulk samples.</p>
<div class="section level3">
<h3 id="pseudo-bulk-samples">Pseudo-bulk samples<a class="anchor" aria-label="anchor" href="#pseudo-bulk-samples"></a></h3>
<p>To compute differences between groups of cells, a possible way is to
compute pseudo-bulk samples, where we mediate the gene signal of all the
cells for each specific cell type. In this manner, we are then able to
detect differences between the same cell type across two different
conditions.</p>
<p>To compute pseudo-bulk samples, we use the
<code>aggregateAcrossCells</code> function in the <code>scuttle</code>
package, which takes as input not only a SingleCellExperiment, but also
the id to use for the identification of the group of cells. In our case,
we use as id not just the cell type, but also the sample, because we
want be able to discern between replicates and conditions during further
steps.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using 'label' and 'sample' as our two factors; each column of the output</span></span>
<span><span class="co"># corresponds to one unique combination of these two factors.</span></span>
<span></span>
<span><span class="va">summed</span> <span class="op">&lt;-</span> <span class="fu">aggregateAcrossCells</span><span class="op">(</span></span>
<span>    <span class="va">merged</span>, </span>
<span>    id <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"celltype.mapped"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">summed</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SingleCellExperiment
dim: 13641 179
metadata(2): merge.info pca.info
assays(1): counts
rownames(13641): ENSMUSG00000051951 ENSMUSG00000025900 ...
  ENSMUSG00000096730 ENSMUSG00000095742
rowData names(3): rotation ENSEMBL SYMBOL
colnames: NULL
colData names(15): batch cell ... sample ncells
reducedDimNames(5): corrected pca.corrected.E7.5 pca.corrected.E8.5 PCA
  TSNE
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a></h3>
<p>The main advantage of using pseudo-bulk samples is the possibility to
use well-tested methods for differential analysis like
<code>edgeR</code> and <code>DESeq2</code>, we will focus on the former
for this analysis. <code>edgeR</code> and <code>DESeq2</code> both use
negative binomial models under the hood, but differ in their
normalization strategies and other implementation details.</p>
<p>First, let’s start with a specific cell type, for instance the
“Mesenchymal stem cells”, and look into differences between this cell
type across conditions. We put the counts table into a
<code>DGEList</code> container called <code>y</code>, along with the
corresponding metadata.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">current</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>, <span class="va">summed</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op">==</span> <span class="st">"Mesenchyme"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">DGEList</span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">current</span><span class="op">)</span>, samples <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">current</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>An object of class "DGEList"
$counts
                   Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
ENSMUSG00000051951       2       0       0       0       1       0
ENSMUSG00000025900       0       0       0       0       0       0
ENSMUSG00000025902       4       0       2       0       3       6
ENSMUSG00000033845     765     130     508     213     781     305
ENSMUSG00000002459       2       0       1       0       0       0
13636 more rows ...

$samples
        group lib.size norm.factors batch cell barcode sample stage tomato pool
Sample1     1  2478901            1     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3
Sample2     1   548407            1     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3
Sample3     1  1260187            1     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4
Sample4     1   578699            1     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4
Sample5     1  2092329            1     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5
Sample6     1   904929            1    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5
        stage.mapped celltype.mapped closest.cell doub.density sizeFactor
Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
        celltype.mapped.1 sample.1 ncells
Sample1        Mesenchyme        5    151
Sample2        Mesenchyme        6     28
Sample3        Mesenchyme        7    127
Sample4        Mesenchyme        8     75
Sample5        Mesenchyme        9    239
Sample6        Mesenchyme       10    146</code></pre>
</div>
<p>A typical step is to discard low quality samples due to low sequenced
library size. We discard these samples because they can affect further
steps like normalization and/or DEGs analysis.</p>
<p>We can see that in our case we don’t have low quality samples and we
don’t need to filter out any of them.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discarded</span> <span class="op">&lt;-</span> <span class="va">current</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&lt;</span> <span class="fl">10</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span>,<span class="op">!</span><span class="va">discarded</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">discarded</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Mode   FALSE
logical       6 </code></pre>
</div>
<p>The same idea is typically applied to the genes, indeed we need to
discard low expressed genes to improve accuracy for the DEGs
modeling.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">filterByExpr</span><span class="op">(</span><span class="va">y</span>, group <span class="op">=</span> <span class="va">current</span><span class="op">$</span><span class="va">tomato</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Mode   FALSE    TRUE
logical    9121    4520 </code></pre>
</div>
<p>We can now proceed to normalize the data. There are several
approaches for normalizing bulk, and hence pseudo-bulk data. Here, we
use the Trimmed Mean of M-values method, implemented in the
<code>edgeR</code> package within the <code>calcNormFactors</code>
function. Keep in mind that because we are going to normalize the
pseudo-bulk counts, we don’t need to normalize the data in “single cell
form”.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span><span class="op">$</span><span class="va">samples</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        group lib.size norm.factors batch cell barcode sample stage tomato pool
Sample1     1  2478901    1.0506857     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3
Sample2     1   548407    1.0399112     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3
Sample3     1  1260187    0.9700083     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4
Sample4     1   578699    0.9871129     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4
Sample5     1  2092329    0.9695559     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5
Sample6     1   904929    0.9858611    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5
        stage.mapped celltype.mapped closest.cell doub.density sizeFactor
Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA
        celltype.mapped.1 sample.1 ncells
Sample1        Mesenchyme        5    151
Sample2        Mesenchyme        6     28
Sample3        Mesenchyme        7    127
Sample4        Mesenchyme        8     75
Sample5        Mesenchyme        9    239
Sample6        Mesenchyme       10    146</code></pre>
</div>
<p>To investigate the effect of our normalization, we use a
Mean-Difference (MD) plot for each sample in order to detect possible
normalization problems due to insufficient cells/reads/UMIs composing a
particular pseudo-bulk profile.</p>
<p>In our case, we verify that all these plots are centered in 0 (on
y-axis) and present a trumpet shape, as expected.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq_len</span><span class="op">(</span><span class="fu">ncol</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">plotMD</span><span class="op">(</span><span class="va">y</span>, column <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Furthermore, we want to check if the samples cluster together based
on their known factors (like the tomato injection in this case).</p>
<p>In this case, we’ll use the multidimensional scaling (MDS) plot.
Multidimensional scaling (which also goes by principal
<em>coordinate</em> analysis (PCoA)) is a dimensionality reduction
technique that’s conceptually similar to principal <em>component</em>
analysis (PCA).</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">limma</span><span class="fu">::</span><span class="fu">plotMDS</span><span class="op">(</span><span class="fu">cpm</span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>               col <span class="op">=</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">tomato</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then construct a design matrix by including both the pool and the
tomato as factors. This design indicates which samples belong to which
pool and condition, so we can use it in the next step of the
analysis.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span><span class="fu">factor</span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu">factor</span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>,</span>
<span>                       data <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">design</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        (Intercept) factor(pool)4 factor(pool)5 factor(tomato)TRUE
Sample1           1             0             0                  1
Sample2           1             0             0                  0
Sample3           1             1             0                  1
Sample4           1             1             0                  0
Sample5           1             0             1                  1
Sample6           1             0             1                  0
attr(,"assign")
[1] 0 1 1 2
attr(,"contrasts")
attr(,"contrasts")$`factor(pool)`
[1] "contr.treatment"

attr(,"contrasts")$`factor(tomato)`
[1] "contr.treatment"</code></pre>
</div>
<p>Now we can estimate the Negative Binomial (NB) overdispersion
parameter, to model the mean-variance trend.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">trended.dispersion</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max.
0.009325 0.016271 0.024233 0.021603 0.026868 0.027993 </code></pre>
</div>
<p>The BCV plot allows us to investigate the relation between the
Biological Coefficient of Variation and the Average log CPM for each
gene. Additionally, the Common and Trend BCV are shown in
<code>red</code> and <code>blue</code>.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotBCV</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then fit a Quasi-Likelihood (QL) negative binomial generalized
linear model for each gene. The <code>robust = TRUE</code> parameter
avoids distortions from highly variable clusters. The QL method includes
an additional dispersion parameter, useful to handle the uncertainty and
variability of the per-gene variance, which is not well estimated by the
NB dispersions, so the two dispersion types complement each other in the
final analysis.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">var.prior</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.2944  0.9342  1.0764  1.0030  1.1300  1.1923 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">df.prior</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.3042  8.7218  8.7218  8.6433  8.7218  8.7218 </code></pre>
</div>
<p>QL dispersion estimates for each gene as a function of abundance. Raw
estimates (black) are shrunk towards the trend (blue) to yield squeezed
estimates (red).</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotQLDisp</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We then use an empirical Bayes quasi-likelihood F-test to test for
differential expression (due to tomato injection) per each gene at a
False Discovery Rate (FDR) of 5%. The low amount of DGEs highlights that
the tomato injection effect has a low influence on the mesenchyme
cells.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">glmQLFTest</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">decideTests</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       factor(tomato)TRUE
Down                    5
NotSig               4510
Up                      5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">topTags</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coefficient:  factor(tomato)TRUE
                        logFC   logCPM          F       PValue          FDR
ENSMUSG00000010760 -4.1526570 9.973704 1116.70486 3.464851e-12 1.566113e-08
ENSMUSG00000096768  1.9988567 8.844258  376.77267 1.070605e-09 2.419567e-06
ENSMUSG00000035299  1.7967382 6.904163  119.58114 3.781967e-07 5.698163e-04
ENSMUSG00000086503 -6.4743977 7.411257  225.86232 2.112730e-06 2.387385e-03
ENSMUSG00000101609  1.3775797 7.310009   80.03729 2.671320e-06 2.414873e-03
ENSMUSG00000019188 -1.0192069 7.545530   61.91544 8.939417e-06 6.734361e-03
ENSMUSG00000024423  0.9941377 7.391075   56.95653 1.312959e-05 8.477962e-03
ENSMUSG00000042607 -0.9516835 7.468203   45.46329 3.618639e-05 2.044531e-02
ENSMUSG00000036446 -0.8287663 9.401028   43.05154 4.740864e-05 2.317935e-02
ENSMUSG00000027520  1.5915670 6.952923   42.77910 5.128175e-05 2.317935e-02</code></pre>
</div>
<p>All the previous steps can be easily performed with the following
function for each cell type, thanks to the <code>pseudoBulkDGE</code>
function in the <code>scran</code> package.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summed.filt</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>,<span class="va">summed</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">de.results</span> <span class="op">&lt;-</span> <span class="fu">pseudoBulkDGE</span><span class="op">(</span></span>
<span>    <span class="va">summed.filt</span>, </span>
<span>    label <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">celltype.mapped</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span><span class="fu">factor</span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="va">tomato</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"tomatoTRUE"</span>,</span>
<span>    condition <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">tomato</span> </span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>The returned object is a list of <code>DataFrame</code>s each with
the results for a cell type. Each of these contains also the
intermediate results in <code>edgeR</code> format to perform any
intermediate plot or diagnostic.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cur.results</span> <span class="op">&lt;-</span> <span class="va">de.results</span><span class="op">[[</span><span class="st">"Allantois"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">cur.results</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">cur.results</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span>,<span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 13641 rows and 5 columns
                       logFC    logCPM         F      PValue         FDR
                   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
ENSMUSG00000037664 -7.993258  11.55290  3389.401 1.49770e-25 6.29183e-22
ENSMUSG00000010760 -2.574986  12.40592  1154.193 1.78995e-20 3.75980e-17
ENSMUSG00000086503 -7.015615   7.49749   728.110 2.59315e-18 3.63127e-15
ENSMUSG00000096768  1.828370   9.33239   304.252 2.37781e-14 2.49729e-11
ENSMUSG00000022464  0.970406  10.28302   119.624 2.36805e-10 1.98964e-07
...                      ...       ...       ...         ...         ...
ENSMUSG00000095247        NA        NA        NA          NA          NA
ENSMUSG00000096808        NA        NA        NA          NA          NA
ENSMUSG00000079808        NA        NA        NA          NA          NA
ENSMUSG00000096730        NA        NA        NA          NA          NA
ENSMUSG00000095742        NA        NA        NA          NA          NA</code></pre>
</div>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Clearly some of the results have low p-values. What about the effect
sizes? What does <code>logFC</code> stand for?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>“logFC” stands for log fold-change. <code>edgeR</code> uses a log2
convention. Rather than reporting e.g. a 5-fold increase, it’s better to
report a logFC of log2(5) = 2.32. Additive log scales are easier to work
with than multiplicative identity scales, once you get used to it.</p>
<p><code>ENSMUSG00000037664</code> seems to have an estimated logFC of
about -8. That’s a big difference if it’s real.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="differential-abundance">Differential Abundance<a class="anchor" aria-label="anchor" href="#differential-abundance"></a></h2>
<hr class="half-width"><p>With DA we look for differences in cluster <em>abundance</em> across
conditions (the tomato injection in our case), rather than differences
in gene expression.</p>
<p>Our first steps are quantifying the number of cells per each cell
type and fitting a model to catch differences between the injected cells
and the background.</p>
<p>The process is very similar differential expression modeling, but
this time we start our analysis on the computed abundances and without
normalizing the data with TMM.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">table</span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">celltype.mapped</span>, <span class="va">merged</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu">unclass</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">extra.info</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span><span class="fu">match</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span>, <span class="va">merged</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">y.ab</span> <span class="op">&lt;-</span> <span class="fu">DGEList</span><span class="op">(</span><span class="va">abundances</span>, samples <span class="op">=</span> <span class="va">extra.info</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span><span class="fu">factor</span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu">factor</span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>, <span class="va">y.ab</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y.ab</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">y.ab</span>, <span class="va">design</span>, trend <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit.ab</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">y.ab</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span>, abundance.trend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="background-on-compositional-effect">Background on compositional effect<a class="anchor" aria-label="anchor" href="#background-on-compositional-effect"></a></h3>
<p>As mentioned before, in DA we don’t normalize our data with
<code>calcNormFactors</code> function, because this approach considers
that most of the input features do not vary between conditions. This
cannot be applied to DA analysis because we have a small number of cell
populations that all can change due to the treatment. This means that
here we will normalize only for library depth, which in pseudo-bulk data
means by the total number of cells in each sample (cell type).</p>
<p>On the other hand, this can lead our data to be susceptible to
compositional effect. “Compositional” refers to the fact that the
cluster abundances in a sample are not independent of one another
because each cell type is effectively competing for space in the sample.
They behave like proportions in that they must sum to 1. If cell type A
abundance increases in a new condition, that means we’ll observe less of
everything else, even if everything else is unaffected by the new
condition.</p>
<p>Compositionality means that our conclusions can be biased by the
amount of cells present in each cell type. And this amount of cells can
be totally unbalanced between cell types. This is particularly
problematic for cell types that start at or end up near 0 or 100
percent.</p>
<p>For example, a specific cell type can be 40% of the total amount of
cells present in the experiment, while another just the 3%. The
differences in terms of abundance of these cell types are detected
between the different conditions, but our final interpretation could be
biased if we don’t consider this aspect.</p>
<p>We now look at different approaches for handling the compositional
effect.</p>
</div>
<div class="section level3">
<h3 id="assuming-most-labels-do-not-change">Assuming most labels do not change<a class="anchor" aria-label="anchor" href="#assuming-most-labels-do-not-change"></a></h3>
<p>We can use a similar approach used during the DEGs analysis, assuming
that most labels are not changing, in particular if we think about the
low number of DEGs resulted from the previous analysis.</p>
<p>To do so, we first normalize the data with
<code>calcNormFactors</code> and then we fit and estimate a QL-model for
our abundance data.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y.ab2</span> <span class="op">&lt;-</span> <span class="fu">calcNormFactors</span><span class="op">(</span><span class="va">y.ab</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y.ab2</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">norm.factors</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.1029040 1.0228173 1.0695358 0.7686501 1.0402941 1.0365354</code></pre>
</div>
<p>We then use edgeR in a manner similar to what we ran before:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y.ab2</span> <span class="op">&lt;-</span> <span class="fu">estimateDisp</span><span class="op">(</span><span class="va">y.ab2</span>, <span class="va">design</span>, trend <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit.ab2</span> <span class="op">&lt;-</span> <span class="fu">glmQLFit</span><span class="op">(</span><span class="va">y.ab2</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span>, abundance.trend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">glmQLFTest</span><span class="op">(</span><span class="va">fit.ab2</span>, coef <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">decideTests</span><span class="op">(</span><span class="va">res2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       factor(tomato)TRUE
Down                    2
NotSig                 32
Up                      0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">topTags</span><span class="op">(</span><span class="va">res2</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coefficient:  factor(tomato)TRUE
                       logFC   logCPM         F       PValue          FDR
ExE ectoderm      -5.7462892 13.13490 40.650293 1.853481e-08 6.301836e-07
Parietal endoderm -6.9020425 12.36649 28.120078 1.338738e-06 2.275854e-05
Mesenchyme         0.9656630 16.32654  6.729443 1.160472e-02 1.111437e-01
Erythroid3        -0.9191728 17.34677  6.495750 1.307572e-02 1.111437e-01
Neural crest      -1.0201642 14.83912  5.769547 1.904354e-02 1.294961e-01
ExE endoderm      -3.9992208 10.75172  5.058004 2.775531e-02 1.572801e-01
Endothelium        0.8672530 14.12195  3.614431 6.151880e-02 2.988056e-01
Cardiomyocytes     0.6955522 14.93321  2.833474 9.690453e-02 4.118442e-01
Allantois          0.5992522 15.54924  2.277082 1.359313e-01 5.135182e-01
Erythroid2        -0.5185684 15.97357  1.773710 1.873705e-01 6.370598e-01</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="testing-against-a-log-fold-change-threshold">Testing against a log-fold change threshold<a class="anchor" aria-label="anchor" href="#testing-against-a-log-fold-change-threshold"></a></h3>
<p>A second approach assumes that the composition bias introduces a
spurious log2-fold change of no more than a quantity for a non-DA
label.</p>
<p>In other words, we interpret this as the maximum log-fold change in
the total number of cells given by DA in other labels. On the other
hand, when choosing , we should not consider fold-differences in the
totals due to differences in capture efficiency or the size of the
original cell population are not attributable to composition bias. We
then mitigate the effect of composition biases by testing each label for
changes in abundance beyond .</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.lfc</span> <span class="op">&lt;-</span> <span class="fu">glmTreat</span><span class="op">(</span><span class="va">fit.ab</span>, coef <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">design</span><span class="op">)</span>, lfc <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="fu">decideTests</span><span class="op">(</span><span class="va">res.lfc</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       factor(tomato)TRUE
Down                    2
NotSig                 32
Up                      0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">topTags</span><span class="op">(</span><span class="va">res.lfc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Coefficient:  factor(tomato)TRUE
                         logFC unshrunk.logFC   logCPM       PValue
ExE ectoderm        -5.5199847     -5.9467998 13.06465 7.002934e-06
Parietal endoderm   -6.5913805    -27.4174596 12.30091 1.486452e-04
ExE endoderm        -3.9308146    -23.9393101 10.76159 8.050176e-02
Mesenchyme           1.1622187      1.1634540 16.35239 1.306006e-01
Endothelium          1.0599946      1.0656145 14.14422 2.114555e-01
Caudal neurectoderm -1.4559768     -1.6059830 11.09613 3.351750e-01
Cardiomyocytes       0.8474499      0.8498787 14.96579 3.707748e-01
Neural crest        -0.8411426     -0.8436838 14.83184 3.730129e-01
Def. endoderm        0.7372230      0.7504500 12.50001 4.213046e-01
Allantois            0.7569205      0.7581982 15.54528 4.713874e-01
                             FDR
ExE ectoderm        0.0002380998
Parietal endoderm   0.0025269677
ExE endoderm        0.9123532342
Mesenchyme          0.9859547701
Endothelium         0.9859547701
Caudal neurectoderm 0.9859547701
Cardiomyocytes      0.9859547701
Neural crest        0.9859547701
Def. endoderm       0.9859547701
Allantois           0.9859547701</code></pre>
</div>
<p>Addionally, the choice of can be guided by other external
experimental data, like a previous or a pilot experiment.</p>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a></h2>
<hr class="half-width"><div id="exercise-1-heatmaps" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1-heatmaps" class="callout-inner">
<h3 class="callout-title">Exercise 1: Heatmaps</h3>
<div class="callout-content">
<p>Use the <code>pheatmap</code> package to create a heatmap of the
abundances table. Does it comport with the model results?</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p>You can simply hand <code>pheatmap()</code> a matrix as its only
argument. <code>pheatmap()</code> has a million options you can adjust,
but the defaults are usually pretty good. Try to overlay sample-level
information with the <code>annotation_col</code> argument for an extra
challenge.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">y.ab</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-24-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno_df</span> <span class="op">&lt;-</span> <span class="va">y.ab</span><span class="op">$</span><span class="va">samples</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"tomato"</span>, <span class="st">"pool"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">anno_df</span><span class="op">$</span><span class="va">pool</span> <span class="op">=</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">anno_df</span><span class="op">$</span><span class="va">pool</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_df</span><span class="op">$</span><span class="va">tomato</span> <span class="op">&lt;-</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">anno_df</span><span class="op">$</span><span class="va">tomato</span>,</span>
<span>                         <span class="st">"tomato+"</span>,</span>
<span>                         <span class="st">"tomato-"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">y.ab</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>         annotation_col <span class="op">=</span> <span class="va">anno_df</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-24-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The top DA result was a decrease in ExE ectoderm in the tomato
condition, which you can sort of see, especially if you
<code>log1p()</code> the counts or discard rows that show much higher
values. ExE ectoderm counts were much higher in samples 8 and 10
compared to 5, 7, and 9.</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-model-specification-and-comparison" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2-model-specification-and-comparison" class="callout-inner">
<h3 class="callout-title">Exercise 2: Model specification and comparison</h3>
<div class="callout-content">
<p>Try re-running the pseudobulk DGE without the <code>pool</code>
factor in the design specification. Compare the logFC estimates and the
distribution of p-values for the <code>Erythroid3</code> cell type.</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" aria-labelledby="headingHint2" data-bs-parent="#accordionHint2">
<div class="accordion-body">
<p>After running the second pseudobulk DGE, you can join the two
<code>DataFrame</code>s of <code>Erythroid3</code> statistics using the
<code>merge()</code> function. You will need to create a common key
column from the gene IDs.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de.results2</span> <span class="op">&lt;-</span> <span class="fu">pseudoBulkDGE</span><span class="op">(</span></span>
<span>    <span class="va">summed.filt</span>, </span>
<span>    label <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">celltype.mapped</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span><span class="va">tomato</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"tomatoTRUE"</span>,</span>
<span>    condition <span class="op">=</span> <span class="va">summed.filt</span><span class="op">$</span><span class="va">tomato</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">eryth1</span> <span class="op">&lt;-</span> <span class="va">de.results</span><span class="op">$</span><span class="va">Erythroid3</span></span>
<span></span>
<span><span class="va">eryth2</span> <span class="op">&lt;-</span> <span class="va">de.results2</span><span class="op">$</span><span class="va">Erythroid3</span></span>
<span></span>
<span><span class="va">eryth1</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">eryth1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">eryth2</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">eryth2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">comp_df</span> <span class="op">&lt;-</span> <span class="fu">merge</span><span class="op">(</span><span class="va">eryth1</span>, <span class="va">eryth2</span>, by <span class="op">=</span> <span class="st">'gene'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">comp_df</span> <span class="op">&lt;-</span> <span class="va">comp_df</span><span class="op">[</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">comp_df</span><span class="op">$</span><span class="va">logFC.x</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">comp_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">logFC.x</span>, <span class="va">logFC.y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_abline</span><span class="op">(</span>lty <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> </span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-25-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reshape to long format for ggplot facets. This is 1000x times easier to do</span></span>
<span><span class="co"># with tidyverse packages:</span></span>
<span><span class="va">pval_df</span> <span class="op">&lt;-</span> <span class="fu">reshape</span><span class="op">(</span><span class="va">comp_df</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"PValue.x"</span>, <span class="st">"PValue.y"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                   direction <span class="op">=</span> <span class="st">"long"</span>, </span>
<span>                   v.names <span class="op">=</span> <span class="st">"Pvalue"</span>,</span>
<span>                   timevar <span class="op">=</span> <span class="st">"pool_factor"</span>,</span>
<span>                   times <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"with pool factor"</span>, <span class="st">"no pool factor"</span><span class="op">)</span>,</span>
<span>                   varying <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"PValue.x"</span>, <span class="st">"PValue.y"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pval_df</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">Pvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_histogram</span><span class="op">(</span>boundary <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                   bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">facet_wrap</span><span class="op">(</span><span class="st">"pool_factor"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/multi-sample-rendered-unnamed-chunk-25-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can see that in this case, the logFC estimates are strongly
consistent between the two models, which tells us that the inclusion of
the <code>pool</code> factor in the model doesn’t strongly influence the
estimate of the <code>tomato</code> coefficients in this case.</p>
<p>The p-value histograms both look alright here, with a largely flat
plateau over most of the 0 - 1 range and a spike near 0. This is
consistent with the hypothesis that most genes are unaffected by
<code>tomato</code> but there are a small handful that clearly are.</p>
<p>If there were large shifts in the logFC estimates or p-value
distributions, that’s a sign that the design specification change has a
large impact on how the model sees the data. If that happens, you’ll
need to think carefully and critically about what variables should and
should not be included in the model formula.</p>
</div>
</div>
</div>
</div>
<div id="extension-challenge-1-group-effects" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="extension-challenge-1-group-effects" class="callout-inner">
<h3 class="callout-title">Extension challenge 1: Group effects</h3>
<div class="callout-content">
<p>Having multiple independent samples in each experimental group is
always helpful, but it’s particularly important when it comes to batch
effect correction. Why?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p>It’s important to have multiple samples within each experimental
group because it helps the batch effect correction algorithm distinguish
differences due to batch effects (uninteresting) from differences due to
group/treatment/biology (interesting).</p>
<p>Imagine you had one sample that received a drug treatment and one
that did not, each with 10,000 cells. They differ substantially in
expression of gene X. Is that an important scientific finding? You can’t
tell for sure, because the effect of drug is indistinguishable from a
sample-wise batch effect. But if the difference in gene X holds up when
you have five treated samples and five untreated samples, now you can be
a bit more confident. Many batch effect correction methods will take
information on experimental factors as additional arguments, which they
can use to help remove batch effects while retaining experimental
differences.</p>
</div>
</div>
</div>
</div>
<div id="further-reading" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further Reading</h3>
<div class="callout-content">
<ul><li>OSCA book, Multi-sample analysis, <a href="https://bioconductor.org/books/release/OSCA.multisample" class="external-link">Chapters
1, 4, and 6</a>
</li>
</ul></div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Batch effects are systematic technical differences in the observed
expression in cells measured in different experimental batches.</li>
<li>Computational removal of batch-to-batch variation with the
<code>correctExperiment</code> function from the <em><a href="https://bioconductor.org/packages/3.19/batchelor" class="external-link">batchelor</a></em>
package allows us to combine data across multiple batches for a
consolidated downstream analysis.</li>
<li>Differential expression (DE) analysis of replicated multi-condition
scRNA-seq experiments is typically based on pseudo-bulk expression
profiles, generated by summing counts for all cells with the same
combination of label and sample.</li>
<li>The <code>aggregateAcrossCells</code> function from the <em><a href="https://bioconductor.org/packages/3.19/scater" class="external-link">scater</a></em>
package facilitates the creation of pseudo-bulk samples.<br></li>
<li>The <code>pseudoBulkDGE</code> function from the <em><a href="https://bioconductor.org/packages/3.19/scran" class="external-link">scran</a></em>
package can be used to detect significant changes in expression between
conditions for pseudo-bulk samples consisting of cells of the same
type.</li>
<li>Differential abundance (DA) analysis aims at identifying significant
changes in cell type abundance across conditions.</li>
<li>DA analysis uses bulk DE methods such as <em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em> and
<em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>,
which provide suitable statistical models for count data in the presence
of limited replication - except that the counts are not of reads per
gene, but of cells per label.</li>
</ul></div>
</div>
</div>
</section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a></h2>
<hr class="half-width"><div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.4.3 (2025-02-28)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] pheatmap_1.0.12              scran_1.32.0
 [3] scater_1.32.1                ggplot2_3.5.1
 [5] scuttle_1.14.0               edgeR_4.2.2
 [7] limma_3.60.6                 batchelor_1.20.0
 [9] MouseGastrulationData_1.18.0 SpatialExperiment_1.14.0
[11] SingleCellExperiment_1.26.0  SummarizedExperiment_1.34.0
[13] Biobase_2.64.0               GenomicRanges_1.56.2
[15] GenomeInfoDb_1.40.1          IRanges_2.38.1
[17] S4Vectors_0.42.1             BiocGenerics_0.50.0
[19] MatrixGenerics_1.16.0        matrixStats_1.5.0
[21] BiocStyle_2.32.1

loaded via a namespace (and not attached):
  [1] DBI_1.2.3                 formatR_1.14
  [3] gridExtra_2.3             rlang_1.1.5
  [5] magrittr_2.0.3            compiler_4.4.3
  [7] RSQLite_2.3.9             DelayedMatrixStats_1.26.0
  [9] png_0.1-8                 vctrs_0.6.5
 [11] pkgconfig_2.0.3           crayon_1.5.3
 [13] fastmap_1.2.0             dbplyr_2.5.0
 [15] magick_2.8.5              XVector_0.44.0
 [17] labeling_0.4.3            rmarkdown_2.29
 [19] ggbeeswarm_0.7.2          UCSC.utils_1.0.0
 [21] purrr_1.0.2               bit_4.5.0.1
 [23] bluster_1.14.0            xfun_0.50
 [25] zlibbioc_1.50.0           cachem_1.1.0
 [27] beachmat_2.20.0           jsonlite_1.8.9
 [29] blob_1.2.4                DelayedArray_0.30.1
 [31] BiocParallel_1.38.0       cluster_2.1.8
 [33] irlba_2.3.5.1             parallel_4.4.3
 [35] R6_2.5.1                  RColorBrewer_1.1-3
 [37] Rcpp_1.0.14               knitr_1.49
 [39] splines_4.4.3             Matrix_1.7-2
 [41] igraph_2.1.4              tidyselect_1.2.1
 [43] viridis_0.6.5             abind_1.4-8
 [45] yaml_2.3.10               codetools_0.2-20
 [47] curl_6.2.0                lattice_0.22-6
 [49] tibble_3.2.1              withr_3.0.2
 [51] KEGGREST_1.44.1           BumpyMatrix_1.12.0
 [53] Rtsne_0.17                evaluate_1.0.3
 [55] BiocFileCache_2.12.0      ExperimentHub_2.12.0
 [57] Biostrings_2.72.1         pillar_1.10.1
 [59] BiocManager_1.30.25       filelock_1.0.3
 [61] renv_1.1.2                generics_0.1.3
 [63] BiocVersion_3.19.1        sparseMatrixStats_1.16.0
 [65] munsell_0.5.1             scales_1.3.0
 [67] glue_1.8.0                metapod_1.12.0
 [69] tools_4.4.3               AnnotationHub_3.12.0
 [71] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0
 [73] locfit_1.5-9.11           cowplot_1.1.3
 [75] grid_4.4.3                AnnotationDbi_1.66.0
 [77] colorspace_2.1-1          GenomeInfoDbData_1.2.12
 [79] beeswarm_0.4.0            BiocSingular_1.20.0
 [81] vipor_0.4.7               cli_3.6.3
 [83] rsvd_1.0.5                rappdirs_0.3.3
 [85] viridisLite_0.4.2         S4Arrays_1.4.1
 [87] dplyr_1.1.4               ResidualMatrix_1.14.1
 [89] gtable_0.3.6              digest_0.6.37
 [91] dqrng_0.4.1               ggrepel_0.9.6
 [93] SparseArray_1.4.8         farver_2.1.2
 [95] rjson_0.2.23              memoise_2.0.1
 [97] htmltools_0.5.8.1         lifecycle_1.0.4
 [99] httr_1.4.7                mime_0.12
[101] statmod_1.5.0             bit64_4.6.0-1            </code></pre>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/cell_type_annotation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/large_data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/cell_type_annotation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Cell type annotation
        </a>
        <a class="chapter-link float-end" href="../instructor/large_data.html" rel="next">
          Next: Working with large...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/ccb-hms/osca-carpentries/edit/main/episodes/multi-sample.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/ccb-hms/osca-carpentries/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/ccb-hms/osca-carpentries/" class="external-link">Source</a></p>
				<p><a href="https://github.com/ccb-hms/osca-carpentries/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:christopher_magnano@hms.harvard.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://ccb-hms.github.io/osca-carpentries/instructor/multi-sample.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "single-cell RNA-seq, The Carpentries, lesson, genomics, bioconductor",
  "name": "Multi-sample analyses",
  "creativeWorkStatus": "active",
  "url": "https://ccb-hms.github.io/osca-carpentries/instructor/multi-sample.html",
  "identifier": "https://ccb-hms.github.io/osca-carpentries/instructor/multi-sample.html",
  "dateCreated": "2024-01-10",
  "dateModified": "2025-03-14",
  "datePublished": "2025-03-14"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

