---
title: Introduction to Bioconductor and the SingleCellExperiment class
teaching: 20 # Minutes of teaching in the lesson
exercises: 10 # Minutes of exercises in the lesson
---

:::::::::::::::::::::::::::::::::::::: questions 

- What is Bioconductor?
- How is single-cell data stored in the Bioconductor ecosystem?
- What is a `SingleCellExperiment` object?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Install and update Bioconductor packages. 
- Load data generated with common single-cell technologies as `SingleCellExperiment` objects. 
- Inspect and manipulate `SingleCellExperiment` objects. 

::::::::::::::::::::::::::::::::::::::::::::::::


```{r chunk-opts, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(BiocStyle)
options(digits = 3)
```

## Bioconductor

### Overview 

Within the [R](https://www.r-project.org/) ecosystem, the [Bioconductor](https://bioconductor.org/) project provides tools for the analysis and comprehension of high-throughput genomics data.
The scope of the project covers microarray data, various forms of sequencing (RNA-seq, ChIP-seq, bisulfite, genotyping, etc.), proteomics, flow cytometry and more.
One of Bioconductor's main selling points is the use of common data structures to promote interoperability between packages, allowing code written by different people (from different organizations, in different countries) to work together seamlessly in complex analyses. 

### Installing Bioconductor Packages

The default repository for R packages is the [Comprehensive R Archive Network](https://cran.r-project.org/mirrors.html) (CRAN), which is home to over 13,000 different R packages. 
We can easily install packages from CRAN - say, the popular `r CRANpkg("ggplot2")` package for data visualization - by opening up R and typing in:

```{r, eval=FALSE}
install.packages("ggplot2")
```

In our case, however, we want to install Bioconductor packages.
These packages are located in a separate repository hosted by Bioconductor, so we first install the `r CRANpkg("BiocManager")` package to easily connect to the Bioconductor servers.

```{r, eval=FALSE}
install.packages("BiocManager")
```

After that, we can use `r CRANpkg("BiocManager")`'s `install()` function to install any package from Bioconductor.
For example, the code chunk below uses this approach to install the `r Biocpkg("SingleCellExperiment")` package.

```{r, eval=FALSE}
BiocManager::install("SingleCellExperiment")
```

Should we forget, the same instructions are present on the landing page of any Bioconductor package.
For example, looking at the [`scater`](https://bioconductor.org/packages/release/bioc/html/scater.html) package page on Bioconductor, we can see the following copy-pasteable instructions:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("scater")
```

Packages only need to be installed once, and then they are available for all subsequent uses of a particular R installation.
There is no need to repeat the installation every time we start R.

### Finding relevant packages

To find relevant Bioconductor packages, one useful resource is the [BiocViews](https://bioconductor.org/packages/release/BiocViews.html) page.
This provides a hierarchically organized view of annotations associated with each Bioconductor package.
For example, under the ["Software"](https://bioconductor.org/packages/release/BiocViews.html#___Software) label, we might be interested in a particular ["Technology"](https://bioconductor.org/packages/release/BiocViews.html#___Technology) such as... say, ["SingleCell"](https://bioconductor.org/packages/release/BiocViews.html#___SingleCell).
This gives us a listing of all Bioconductor packages that might be useful for our single-cell data analyses. 
CRAN uses the similar concept of ["Task views"](https://cran.r-project.org/web/views/), though this is understandably more general than genomics.
For example, the [Cluster task view page](https://cran.r-project.org/web/views/Cluster.html) lists an assortment of packages that are relevant to cluster analyses.

### Staying up to date

Updating all R/Bioconductor packages is as simple as running `BiocManager::install()` without any arguments.
This will check for more recent versions of each package (within a Bioconductor release) and prompt the user to update if any are available.

```{r, eval=FALSE}
BiocManager::install()
```

This might take some time if many packages need to be updated, but is typically
recommended to avoid issues resulting from outdated package versions. 

## The `SingleCellExperiment` class

### Setup

We start by loading some libraries we'll be using:

```{r setup}
library(SingleCellExperiment)
library(MouseGastrulationData)
```

It is normal to see lot of startup messages when loading these packages. 

### Motivation and overview

One of the main strengths of the Bioconductor project lies in the use of a common data infrastructure that powers interoperability across packages. 

Users should be able to analyze their data using functions from different Bioconductor packages without the need to convert between formats. To this end, the `SingleCellExperiment` class (from the _SingleCellExperiment_ package) serves as the common currency for data exchange across 70+ single-cell-related Bioconductor packages.

This class implements a data structure that stores all aspects of our single-cell data - gene-by-cell expression data, cell-wise metadata, and gene-wise annotation - and lets us manipulate them in an organized manner.

```{r, echo=FALSE}
knitr::include_graphics("http://bioconductor.org/books/release/OSCA.intro/images/SingleCellExperiment.png")
```

:::: spoiler

### Benefits of using the integrated `SingleCellExperiment` data container

The complexity of the `SingleCellExperiment` container might be a little bit intimidating in the beginning. 
One might be tempted to use a simpler approach by just keeping all of these components in separate objects,
e.g. a `matrix` of counts, a `data.frame` of sample metadata, a `data.frame` of gene annotations, and so on.

There are two main disadvantages to this "from-scratch" approach:

1. It requires a substantial amount of manual bookkeeping to keep the different data components in sync. If you performed a QC step that removed dead cells from the count matrix, you also had to remember to remove that same set of cells from the cell-wise metadata. Did you filtered out genes that did not display sufficient expression levels to be retained for further analysis? Then you would need to make sure to not forget to filter the gene metadata table too. 
2. All the downstream steps had to be "from scratch" as well. All the data munging, analysis, and visualization code had to be customized to the idiosyncrasies of a given input set.

::::

Let's look at an example dataset. `WTChimeraData` comes from a study on mouse development. We can assign one sample to a `SingleCellExperiment` object named `sce` like so:

```{r, message = FALSE, warning=FALSE}
sce <- WTChimeraData(samples = 5)

sce
```

We can think of this (and other) class as a _container_, that contains several different pieces of data in so-called _slots_. SingleCellExperiment objects come with dedicated methods for _getting_ and _setting_ the data in their slots. 

Depending on the object, slots can contain different types of data (e.g., numeric matrices, lists, etc.). Here we'll review the main slots of the SingleCellExperiment class as well as their getter/setter methods.

:::: challenge

Get the data for a different sample from  `WTChimeraData` (other than the fifth one).

::: solution

Here we obtain the sixth sample and assign it to `sce6`:

```{r, message = FALSE, warning=FALSE, eval=FALSE}
sce6 <- WTChimeraData(samples = 6)

sce6
```

:::
  
::::

### `assays`

This is arguably the most fundamental part of the object that contains the count matrix, and potentially other matrices with transformed data. We can access the _list_ of matrices with the `assays` function and individual matrices with the `assay` function. If one of these matrices is called "counts", we can use the special `counts` getter (likewise for `logcounts`).

```{r}
names(assays(sce))

counts(sce)[1:3, 1:3]
```

You will notice that in this case we have a sparse matrix of class `dgTMatrix`
inside the object. More generally, any "matrix-like" object can be used, e.g.,
dense matrices or HDF5-backed matrices (as we will explore later in the
[Working with large data](https://ccb-hms.github.io/osca-workbench/large_data.html)
episode).

### `colData` and `rowData`

Conceptually, these are two data frames that annotate the columns and the rows of your assay, respectively.

One can interact with them as usual, e.g., by extracting columns or adding additional variables as columns.

```{r}
colData(sce)[1:3, 1:4]

rowData(sce)[1:3, 1:2]
```

You can access columns of the colData with the `$` accessor to quickly add cell-wise metadata to the `colData`.

```{r}
sce$my_sum <- colSums(counts(sce))

colData(sce)[1:3,]
```

:::: challenge 

Add a column of gene-wise metadata to the `rowData`.

::: solution

Here, we add a column named `conservation` that could represent an evolutionary conservation score.

```{r}
rowData(sce)$conservation <- rnorm(nrow(sce))
```

This is just an example for demonstration purposes, but in practice it is convenient
and simplifies data management to store any sort of gene-wise information in the columns of the `rowData`.

::: 

::::

### The `reducedDims`

Everything that we have described so far (except for the `counts` getter) is part of the `SummarizedExperiment` class that `SingleCellExperiment` extends. You can find a complete lesson on the `SummarizedExperiment` class in [Introduction to data analysis with R and Bioconductor](https://carpentries-incubator.github.io/bioc-intro/60-next-steps.html) course.

One peculiarity of `SingleCellExperiment` is its ability to store reduced dimension matrices within the object. These may include PCA, t-SNE, UMAP, etc.

```{r}
reducedDims(sce)
```

As for the other slots, we have the usual setter/getter, but it is somewhat rare to interact directly with these functions.

It is more common for other functions to _store_ this information in the object, e.g., the `runPCA` function from the `scater` package.

Here, we use `scater`'s `plotReducedDim` function as an example of how to extract this information _indirectly_ from the objects. Note that one could obtain the same results (somewhat less efficiently) by extracting the corresponding `reducedDim` matrix and `ggplot`.

```{r, message = FALSE, warning = FALSE}
library(scater)

plotReducedDim(sce, "pca.corrected.E8.5", colour_by = "celltype.mapped")
```

:::::::::::::::::::::::::::::::::: challenge

#### Exercise 1

Create a `SingleCellExperiment` object "from scratch". That means: start from a `matrix` (either randomly generated or with some fake data in it) and add one or more columns as `colData`.

:::::::::::::: hint

The `SingleCellExperiment` constructor function can be used to create a new `SingleCellExperiment` object.

:::::::::::::::::::::::


::: solution

```{r}
mat <- matrix(runif(30), ncol = 5)

my_sce <- SingleCellExperiment(assays = list(logcounts = mat))

my_sce$my_col_info = runif(5)

my_sce
```

:::

:::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::: challenge

#### Exercise 2

Combine two `SingleCellExperiment` objects. The `MouseGastrulationData` package contains several datasets. Download sample 6 of the chimera experiment. Use the `cbind` function to combine the new data with the `sce` object created before. 

::: solution

```{r message = FALSE, warning=FALSE}
sce  <- WTChimeraData(samples = 5)

sce6 <- WTChimeraData(samples = 6)

combined_sce <- cbind(sce, sce6)

combined_sce
```

:::


:::::::::::::::::::::::::::::::::::::::::::::


:::::::::::::: checklist

## Further Reading

* OSCA book, [Introduction](https://bioconductor.org/books/release/OSCA.intro)

::::::::::::::

::::::::::::::::::::::::::::::::::::: keypoints 

- The Bioconductor project provides open-source software packages for the comprehension of high-throughput biological data.
- A `SingleCellExperiment` object is an extension of the `SummarizedExperiment` object.
- `SingleCellExperiment` objects contain specialized data fields for storing data unique to single-cell analyses, such as the `reducedDims` field. 

::::::::::::::::::::::::::::::::::::::::::::::::


